{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9e61aca2",
   "metadata": {},
   "source": [
    "Assess the accuracy of BEAN estimates by # of co-occurring variants within the editing window of a gRNA."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9386dae5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/pinello/SHARED_SOFTWARE/anaconda_latest/envs/jy_18loci/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy import stats\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import bean as be"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e4c66ca9",
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.style.use('jr')\n",
    "plt.rcParams[\"pdf.use14corefonts\"] = True\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42 \n",
    "matplotlib.rcParams[\"axes.unicode_minus\"] = False\n",
    "sns.set_theme(style=\"ticks\", rc={\"axes.facecolor\": (0, 0, 0, 0)})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "6a143479",
   "metadata": {},
   "outputs": [],
   "source": [
    "bdata = be.read_h5ad(\"../../../results/filtered_annotated/LDLRCDS/bean_count_LDLRCDS_annotated_spacer0.1_0.3.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "456d3163",
   "metadata": {},
   "outputs": [],
   "source": [
    "score_tbl = pd.read_csv(\"../../Fig5c_hits/TableSX_tiling_screen_UKBmerged.20230817.alpha_fixed.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c1b67774",
   "metadata": {},
   "outputs": [],
   "source": [
    "score_tbl_single = score_tbl.loc[score_tbl.n_guides == 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "0a5158e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_tbl = bdata.uns[\"sig_allele_counts_spacer_0_19_A.G_translated_prop0.1_0.3\"].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5f1d9be2",
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_rates_tbl = bdata.get_normalized_allele_counts(allele_tbl)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "721586d1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>guide</th>\n",
       "      <th>aa_allele</th>\n",
       "      <th>rep5_top</th>\n",
       "      <th>rep5_high</th>\n",
       "      <th>rep5_bulk</th>\n",
       "      <th>rep5_low</th>\n",
       "      <th>rep5_bot</th>\n",
       "      <th>rep6_top</th>\n",
       "      <th>rep6_high</th>\n",
       "      <th>rep6_bulk</th>\n",
       "      <th>...</th>\n",
       "      <th>rep8_bot</th>\n",
       "      <th>rep9_top</th>\n",
       "      <th>rep9_high</th>\n",
       "      <th>rep9_bulk</th>\n",
       "      <th>rep9_low</th>\n",
       "      <th>rep9_bot</th>\n",
       "      <th>bulk_mean</th>\n",
       "      <th>nt_edits</th>\n",
       "      <th>aa_edits</th>\n",
       "      <th>edits</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10_2511_neg</td>\n",
       "      <td>|11224168:20:-:A&gt;G</td>\n",
       "      <td>0.009091</td>\n",
       "      <td>0.144186</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.150000</td>\n",
       "      <td>0.006369</td>\n",
       "      <td>0.209503</td>\n",
       "      <td>0.067568</td>\n",
       "      <td>0.124260</td>\n",
       "      <td>...</td>\n",
       "      <td>0.146341</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.011834</td>\n",
       "      <td>0.146341</td>\n",
       "      <td>0.086022</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.103728</td>\n",
       "      <td>[11224168:T&gt;C]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[11224168:T&gt;C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10_2511_neg</td>\n",
       "      <td>|11224175:13:-:A&gt;G,11224178:10:-:A&gt;G</td>\n",
       "      <td>0.490909</td>\n",
       "      <td>0.362791</td>\n",
       "      <td>0.443396</td>\n",
       "      <td>0.212500</td>\n",
       "      <td>0.050955</td>\n",
       "      <td>0.222462</td>\n",
       "      <td>0.319820</td>\n",
       "      <td>0.278107</td>\n",
       "      <td>...</td>\n",
       "      <td>0.414634</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.213018</td>\n",
       "      <td>0.121951</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.357741</td>\n",
       "      <td>[11224175:T&gt;C, 11224178:T&gt;C]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[11224175:T&gt;C, 11224178:T&gt;C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>10_2511_neg</td>\n",
       "      <td>|11224178:10:-:A&gt;G</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>0.148837</td>\n",
       "      <td>0.405660</td>\n",
       "      <td>0.250000</td>\n",
       "      <td>0.305732</td>\n",
       "      <td>0.127430</td>\n",
       "      <td>0.279279</td>\n",
       "      <td>0.171598</td>\n",
       "      <td>...</td>\n",
       "      <td>0.067073</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.360947</td>\n",
       "      <td>0.268293</td>\n",
       "      <td>0.204301</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.220288</td>\n",
       "      <td>[11224178:T&gt;C]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[11224178:T&gt;C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>10_2514_neg</td>\n",
       "      <td>|11224178:13:-:A&gt;G</td>\n",
       "      <td>0.190722</td>\n",
       "      <td>0.251553</td>\n",
       "      <td>0.337748</td>\n",
       "      <td>0.280255</td>\n",
       "      <td>0.296552</td>\n",
       "      <td>0.314894</td>\n",
       "      <td>0.260753</td>\n",
       "      <td>0.430693</td>\n",
       "      <td>...</td>\n",
       "      <td>0.195767</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.191781</td>\n",
       "      <td>0.133333</td>\n",
       "      <td>0.205674</td>\n",
       "      <td>0.583333</td>\n",
       "      <td>0.299068</td>\n",
       "      <td>[11224178:T&gt;C]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[11224178:T&gt;C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10_2514_neg</td>\n",
       "      <td>|11224184:7:-:A&gt;G</td>\n",
       "      <td>0.254296</td>\n",
       "      <td>0.478261</td>\n",
       "      <td>0.271523</td>\n",
       "      <td>0.337580</td>\n",
       "      <td>0.344828</td>\n",
       "      <td>0.591489</td>\n",
       "      <td>0.502688</td>\n",
       "      <td>0.227723</td>\n",
       "      <td>...</td>\n",
       "      <td>0.222222</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.150685</td>\n",
       "      <td>0.300000</td>\n",
       "      <td>0.262411</td>\n",
       "      <td>0.416667</td>\n",
       "      <td>0.236297</td>\n",
       "      <td>[11224184:T&gt;C]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[11224184:T&gt;C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11706</th>\n",
       "      <td>Intron 1 DNaseHS 2_5704_pos</td>\n",
       "      <td>|11203065:6:+:A&gt;G,11203076:17:+:A&gt;G,11203077:1...</td>\n",
       "      <td>0.666667</td>\n",
       "      <td>0.867925</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.883117</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.301887</td>\n",
       "      <td>0.431818</td>\n",
       "      <td>0.939394</td>\n",
       "      <td>...</td>\n",
       "      <td>0.650000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.705882</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.409091</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.623585</td>\n",
       "      <td>[11203076:A&gt;G, 11203079:A&gt;G, 11203065:A&gt;G, 112...</td>\n",
       "      <td>[]</td>\n",
       "      <td>[11203076:A&gt;G, 11203079:A&gt;G, 11203065:A&gt;G, 112...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11707</th>\n",
       "      <td>Intron 1 DNaseHS 2_5707_pos</td>\n",
       "      <td>|11203077:15:+:A&gt;G,11203079:17:+:A&gt;G</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.049383</td>\n",
       "      <td>0.107527</td>\n",
       "      <td>0.151515</td>\n",
       "      <td>0.260563</td>\n",
       "      <td>0.035417</td>\n",
       "      <td>0.134078</td>\n",
       "      <td>0.195980</td>\n",
       "      <td>...</td>\n",
       "      <td>0.371429</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.184971</td>\n",
       "      <td>0.235294</td>\n",
       "      <td>0.077519</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.217760</td>\n",
       "      <td>[11203077:A&gt;G, 11203079:A&gt;G]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[11203077:A&gt;G, 11203079:A&gt;G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11708</th>\n",
       "      <td>Intron 1 DNaseHS 2_5707_pos</td>\n",
       "      <td>|11203079:17:+:A&gt;G</td>\n",
       "      <td>0.487685</td>\n",
       "      <td>0.370370</td>\n",
       "      <td>0.870968</td>\n",
       "      <td>0.745455</td>\n",
       "      <td>0.521127</td>\n",
       "      <td>0.393750</td>\n",
       "      <td>0.430168</td>\n",
       "      <td>0.628141</td>\n",
       "      <td>...</td>\n",
       "      <td>0.342857</td>\n",
       "      <td>0.266667</td>\n",
       "      <td>0.630058</td>\n",
       "      <td>0.647059</td>\n",
       "      <td>0.635659</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.631733</td>\n",
       "      <td>[11203079:A&gt;G]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[11203079:A&gt;G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11709</th>\n",
       "      <td>Intron 1 DNaseHS 2_5710_pos</td>\n",
       "      <td>|11203079:14:+:A&gt;G</td>\n",
       "      <td>0.772727</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.400000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.882353</td>\n",
       "      <td>...</td>\n",
       "      <td>0.920000</td>\n",
       "      <td>0.192308</td>\n",
       "      <td>0.333333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.777778</td>\n",
       "      <td>0.428571</td>\n",
       "      <td>0.727451</td>\n",
       "      <td>[11203079:A&gt;G]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[11203079:A&gt;G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11710</th>\n",
       "      <td>Intron 1 DNaseHS 2_5713_pos</td>\n",
       "      <td>|11203075:7:+:A&gt;G,11203079:11:+:A&gt;G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.107143</td>\n",
       "      <td>[11203079:A&gt;G, 11203075:A&gt;G]</td>\n",
       "      <td>[]</td>\n",
       "      <td>[11203079:A&gt;G, 11203075:A&gt;G]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11711 rows Ã— 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             guide  \\\n",
       "0                      10_2511_neg   \n",
       "1                      10_2511_neg   \n",
       "2                      10_2511_neg   \n",
       "3                      10_2514_neg   \n",
       "4                      10_2514_neg   \n",
       "...                            ...   \n",
       "11706  Intron 1 DNaseHS 2_5704_pos   \n",
       "11707  Intron 1 DNaseHS 2_5707_pos   \n",
       "11708  Intron 1 DNaseHS 2_5707_pos   \n",
       "11709  Intron 1 DNaseHS 2_5710_pos   \n",
       "11710  Intron 1 DNaseHS 2_5713_pos   \n",
       "\n",
       "                                               aa_allele  rep5_top  rep5_high  \\\n",
       "0                                     |11224168:20:-:A>G  0.009091   0.144186   \n",
       "1                   |11224175:13:-:A>G,11224178:10:-:A>G  0.490909   0.362791   \n",
       "2                                     |11224178:10:-:A>G  0.300000   0.148837   \n",
       "3                                     |11224178:13:-:A>G  0.190722   0.251553   \n",
       "4                                      |11224184:7:-:A>G  0.254296   0.478261   \n",
       "...                                                  ...       ...        ...   \n",
       "11706  |11203065:6:+:A>G,11203076:17:+:A>G,11203077:1...  0.666667   0.867925   \n",
       "11707               |11203077:15:+:A>G,11203079:17:+:A>G  0.000000   0.049383   \n",
       "11708                                 |11203079:17:+:A>G  0.487685   0.370370   \n",
       "11709                                 |11203079:14:+:A>G  0.772727   0.000000   \n",
       "11710                |11203075:7:+:A>G,11203079:11:+:A>G       NaN        NaN   \n",
       "\n",
       "       rep5_bulk  rep5_low  rep5_bot  rep6_top  rep6_high  rep6_bulk  ...  \\\n",
       "0       0.000000  0.150000  0.006369  0.209503   0.067568   0.124260  ...   \n",
       "1       0.443396  0.212500  0.050955  0.222462   0.319820   0.278107  ...   \n",
       "2       0.405660  0.250000  0.305732  0.127430   0.279279   0.171598  ...   \n",
       "3       0.337748  0.280255  0.296552  0.314894   0.260753   0.430693  ...   \n",
       "4       0.271523  0.337580  0.344828  0.591489   0.502688   0.227723  ...   \n",
       "...          ...       ...       ...       ...        ...        ...  ...   \n",
       "11706   0.500000  0.883117       NaN  0.301887   0.431818   0.939394  ...   \n",
       "11707   0.107527  0.151515  0.260563  0.035417   0.134078   0.195980  ...   \n",
       "11708   0.870968  0.745455  0.521127  0.393750   0.430168   0.628141  ...   \n",
       "11709   0.400000       NaN       NaN       NaN        NaN   0.882353  ...   \n",
       "11710   0.107143       NaN       NaN       NaN        NaN        NaN  ...   \n",
       "\n",
       "       rep8_bot  rep9_top  rep9_high  rep9_bulk  rep9_low  rep9_bot  \\\n",
       "0      0.146341  0.000000   0.011834   0.146341  0.086022       NaN   \n",
       "1      0.414634  0.000000   0.213018   0.121951  0.333333       NaN   \n",
       "2      0.067073  0.000000   0.360947   0.268293  0.204301       NaN   \n",
       "3      0.195767  0.000000   0.191781   0.133333  0.205674  0.583333   \n",
       "4      0.222222  0.000000   0.150685   0.300000  0.262411  0.416667   \n",
       "...         ...       ...        ...        ...       ...       ...   \n",
       "11706  0.650000  0.000000   0.705882        NaN  0.409091       NaN   \n",
       "11707  0.371429  0.000000   0.184971   0.235294  0.077519  1.000000   \n",
       "11708  0.342857  0.266667   0.630058   0.647059  0.635659  0.000000   \n",
       "11709  0.920000  0.192308   0.333333        NaN  0.777778  0.428571   \n",
       "11710       NaN       NaN        NaN        NaN       NaN       NaN   \n",
       "\n",
       "       bulk_mean                                           nt_edits  aa_edits  \\\n",
       "0       0.103728                                     [11224168:T>C]        []   \n",
       "1       0.357741                       [11224175:T>C, 11224178:T>C]        []   \n",
       "2       0.220288                                     [11224178:T>C]        []   \n",
       "3       0.299068                                     [11224178:T>C]        []   \n",
       "4       0.236297                                     [11224184:T>C]        []   \n",
       "...          ...                                                ...       ...   \n",
       "11706   0.623585  [11203076:A>G, 11203079:A>G, 11203065:A>G, 112...        []   \n",
       "11707   0.217760                       [11203077:A>G, 11203079:A>G]        []   \n",
       "11708   0.631733                                     [11203079:A>G]        []   \n",
       "11709   0.727451                                     [11203079:A>G]        []   \n",
       "11710   0.107143                       [11203079:A>G, 11203075:A>G]        []   \n",
       "\n",
       "                                                   edits  \n",
       "0                                         [11224168:T>C]  \n",
       "1                           [11224175:T>C, 11224178:T>C]  \n",
       "2                                         [11224178:T>C]  \n",
       "3                                         [11224178:T>C]  \n",
       "4                                         [11224184:T>C]  \n",
       "...                                                  ...  \n",
       "11706  [11203076:A>G, 11203079:A>G, 11203065:A>G, 112...  \n",
       "11707                       [11203077:A>G, 11203079:A>G]  \n",
       "11708                                     [11203079:A>G]  \n",
       "11709                                     [11203079:A>G]  \n",
       "11710                       [11203079:A>G, 11203075:A>G]  \n",
       "\n",
       "[11711 rows x 31 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele_rates_tbl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "28426b1a",
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_rates_tbl[\"bulk_mean\"] = allele_rates_tbl.loc[:,allele_rates_tbl.columns.map(lambda s: \"bulk\" in s)].mean(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "29c90b0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_rates_tbl[\"nt_edits\"] = allele_rates_tbl.aa_allele.map(lambda cnallele: [e.get_abs_edit() for e in cnallele.nt_allele.edits])\n",
    "allele_rates_tbl[\"aa_edits\"] = allele_rates_tbl.aa_allele.map(lambda cnallele: [e for e in cnallele.aa_allele.edits])\n",
    "allele_rates_tbl[\"edits\"] = allele_rates_tbl[[\"guide\",\"nt_edits\"]].apply(lambda row: [f\"{row.guide}!{s}\" for s in row.nt_edits] if \"CONTROL\" in row.guide else [str(s) for s in row.nt_edits], axis=1) + allele_rates_tbl[\"aa_edits\"].map(lambda a: [\"A\"+str(s) for s in a])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "df2b758f",
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_rates_tbl = allele_rates_tbl.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "17319d6c",
   "metadata": {},
   "outputs": [],
   "source": [
    "score_tbl_single = score_tbl_single.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "96688f7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_allele_rate(e: be.Edit):\n",
    "    n_vars = []\n",
    "    cooccurring_edits = set()\n",
    "    mean_rates = []\n",
    "    for i, edits_list in enumerate(allele_rates_tbl.edits):\n",
    "        if e in edits_list:\n",
    "            n_vars.append(len(edits_list))\n",
    "            cooccurring_edits.update(edits_list)\n",
    "            mean_rates.append(allele_rates_tbl.iloc[i][\"bulk_mean\"])\n",
    "    return n_vars, mean_rates, cooccurring_edits\n",
    "n_vars, variant_mean_rates, coocc = zip(*score_tbl.edit.map(get_allele_rate))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "0c641578",
   "metadata": {},
   "outputs": [],
   "source": [
    "score_tbl[\"n_vars\"] = n_vars\n",
    "score_tbl[\"allele_rate\"] = variant_mean_rates\n",
    "score_tbl['cooccurring_edits'] = coocc\n",
    "score_tbl['n_coo'] = score_tbl['cooccurring_edits'].map(len) - 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "393c098e",
   "metadata": {},
   "outputs": [],
   "source": [
    "score_tbl_with_ukb = score_tbl.loc[~score_tbl.mean_LDL.isnull()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "933a1d07",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_40952/1029637229.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  score_tbl_with_ukb[\"eer_qbin\"] = pd.qcut(score_tbl_with_ukb.effective_edit_rate, q=nbins)\n",
      "/tmp/ipykernel_40952/1029637229.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  score_tbl_with_ukb[\"nco_qbin\"] = pd.qcut(score_tbl_with_ukb.n_coo, q=nbins)\n",
      "/tmp/ipykernel_40952/1029637229.py:4: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  score_tbl_with_ukb[\"ng_qbin\"] = pd.qcut(score_tbl_with_ukb.n_guides, q=nbins)\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'DataFrame' object has no attribute 'n_coo_div_nguides'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_40952/1029637229.py\u001b[0m in \u001b[0;36m?\u001b[0;34m()\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0mnbins\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;36m4\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0mscore_tbl_with_ukb\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"eer_qbin\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mqcut\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mscore_tbl_with_ukb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0meffective_edit_rate\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mq\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnbins\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mscore_tbl_with_ukb\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"nco_qbin\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mqcut\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mscore_tbl_with_ukb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mn_coo\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mq\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnbins\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mscore_tbl_with_ukb\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"ng_qbin\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mqcut\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mscore_tbl_with_ukb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mn_guides\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mq\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnbins\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mscore_tbl_with_ukb\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"cdg_qbin\"\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpd\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mqcut\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mscore_tbl_with_ukb\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mn_coo_div_nguides\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mq\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnbins\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m/data/pinello/SHARED_SOFTWARE/anaconda_latest/envs/jy_18loci/lib/python3.10/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m?\u001b[0;34m(self, name)\u001b[0m\n\u001b[1;32m   6200\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mname\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_accessors\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6201\u001b[0m             \u001b[0;32mand\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_info_axis\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_hold_identifiers_and_holds_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6202\u001b[0m         ):\n\u001b[1;32m   6203\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6204\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mobject\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__getattribute__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m: 'DataFrame' object has no attribute 'n_coo_div_nguides'"
     ]
    }
   ],
   "source": [
    "nbins = 4\n",
    "score_tbl_with_ukb[\"eer_qbin\"] = pd.qcut(score_tbl_with_ukb.effective_edit_rate, q=nbins)\n",
    "score_tbl_with_ukb[\"nco_qbin\"] = pd.qcut(score_tbl_with_ukb.n_coo, q=nbins)\n",
    "score_tbl_with_ukb[\"ng_qbin\"] = pd.qcut(score_tbl_with_ukb.n_guides, q=nbins)\n",
    "score_tbl_with_ukb[\"cdg_qbin\"] = pd.qcut(score_tbl_with_ukb.n_coo_div_nguides, q=nbins)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22dcf71f",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(1, 4, figsize=(12,4))\n",
    "\n",
    "p = ax[0].hist2d(x=score_tbl_with_ukb.n_coo, y=score_tbl_with_ukb.n_guides, bins= (np.arange(-0.5, score_tbl_with_ukb.n_coo.max()+1.5, 1), np.arange(0.5, score_tbl_with_ukb.n_guides.max()+1.5, 1)))\n",
    "ax[0].set_ylabel(\"# guides\")\n",
    "ax[0].set_xlabel('# co-occurring variants')\n",
    "fig.colorbar(p[3], orientation=\"horizontal\", location=\"top\")\n",
    "\n",
    "measures = [\"n_coo\", \"n_guides\", \"n_coo_div_nguides\",]\n",
    "div = [\"nco_qbin\", \"ng_qbin\", \"cdg_qbin\",]\n",
    "labels = ['# co-occurring variants', '# of variant supporting gRNAs', \"# co-occurring variants\\nnormalized by # supporting gRNAs\"]\n",
    "\n",
    "for i in range(4):\n",
    "    bins = np.arange(-0.5, score_tbl_with_ukb[measures[i]].max()+1.5, 1) if i < 2 else None\n",
    "    ax[1+i].hist(score_tbl_with_ukb[measures[i]], bins= bins)\n",
    "    ax[1+i].set_ylabel(\"Frequency\")\n",
    "    ax[1+i].set_xlabel(labels[i])\n",
    "    for interval in score_tbl_with_ukb[div[i]].unique():\n",
    "        ax[1+i].axvline(interval.right, color='black', linestyle=\"--\")\n",
    "\n",
    "plt.setp(ax, box_aspect=1)\n",
    "plt.tight_layout()\n",
    "plt.suptitle(\"Histogram of variant stats, 76 variants with UKB LDL-C measurements\")\n",
    "fig.savefig(\"variant_stats_4bins.pdf\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff5e1008",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(4, nbins, figsize=(12,12))\n",
    "reverse_qbin=[\"nco_qbin\", \"cdg_qbin\"]\n",
    "div = [\"nco_qbin\", \"ng_qbin\", \"cdg_qbin\", ]\n",
    "label = [\"n_coo\", \"n_guides\", \"n_norm\", ]\n",
    "for i in range(4):\n",
    "    qbins = score_tbl_with_ukb[div[i]].sort_values(ascending=not(div[i] in reverse_qbin))\n",
    "    for j, qbin in enumerate(qbins.unique()):\n",
    "        df = score_tbl_with_ukb.loc[score_tbl_with_ukb[div[i]] == qbin]\n",
    "        sns.scatterplot(data=df, x=\"mean_LDL\", y=\"mu_z_adj\", ax=ax[i, j])\n",
    "        r, pvalue = stats.spearmanr(df[\"mean_LDL\"], df[\"mu_z_adj\"])\n",
    "        #ax[i, j].collections[0].label()\n",
    "        ax[i, j].set_xlabel(\"UKB LDL-C (mg/dL)\")\n",
    "        ax[i, j].set_ylabel(\"BEAN z-score\")\n",
    "        ax[i, j].text(0.98, 0.98, f\"r={r:.2f}\", transform=ax[i, j].transAxes, horizontalalignment='right', verticalalignment='top')\n",
    "        ax[i, j].set_title(f\"{label[i]} in {qbin} \")\n",
    "        ax[i, j].set_box_aspect(1)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"correlation_by_bins_v2.pdf\", bbox_inches=\"tight\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a0302e65",
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(2, 4, figsize=(12, 6))\n",
    "\n",
    "sns.scatterplot(score_tbl, y=\"effective_edit_rate\", x=\"n_coo_div_nguides\", alpha=0.1, ax=ax[0, 0], label = f\"rho={stats.spearmanr(score_tbl.effective_edit_rate, score_tbl.n_coo_div_nguides)[0]:.5f}\")\n",
    "ax[0, 0].set_xlim((0, max(score_tbl.n_coo_div_nguides)*1.05))\n",
    "ax[0, 0].legend()\n",
    "ax[0, 0].set_xlabel(\"n_norm\")\n",
    "ax[0, 0].set_ylabel(\"Effective edit rate\")\n",
    "\n",
    "sns.scatterplot(score_tbl, y=\"mu_sd_adj\", x=\"effective_edit_rate\", alpha=0.1, ax=ax[0, 1], label = f\"rho={stats.spearmanr(score_tbl.mu_sd_adj, score_tbl.effective_edit_rate)[0]:.5f}\")\n",
    "ax[0, 1].legend()\n",
    "ax[0, 1].set_xlabel(\"Effective edit rate\")\n",
    "ax[0, 1].set_ylabel(\"BEAN mu_sd\")\n",
    "\n",
    "sns.scatterplot(score_tbl, y=\"mu_sd_adj\", x=\"n_coo_div_nguides\", alpha=0.1, ax=ax[0, 2], label = f\"rho={stats.spearmanr(score_tbl.mu_sd_adj, score_tbl.n_coo_div_nguides)[0]:.5f}\")\n",
    "ax[0, 2].set_xlim((0, max(score_tbl.n_coo_div_nguides)*1.05))\n",
    "ax[0, 2].legend()\n",
    "ax[0, 2].set_xlabel(\"n_norm\")\n",
    "ax[0, 2].set_ylabel(\"BEAN mu_sd\")\n",
    "\n",
    "ax[0, 3].hist(score_tbl_with_ukb[\"effective_edit_rate\"])\n",
    "ax[0, 3].set_ylabel(\"Frequency\")\n",
    "ax[0, 3].set_xlabel(labels[i])\n",
    "for interval in score_tbl_with_ukb[\"eer_qbin\"].unique():\n",
    "    ax[0, 3].axvline(interval.right, color='black', linestyle=\"--\")\n",
    "    \n",
    "\n",
    "qbins = score_tbl_with_ukb[\"eer_qbin\"].sort_values()\n",
    "for j, qbin in enumerate(qbins.unique()):\n",
    "    df = score_tbl_with_ukb.loc[score_tbl_with_ukb[\"eer_qbin\"] == qbin]\n",
    "    sns.scatterplot(data=df, x=\"mean_LDL\", y=\"mu_z_adj\", ax=ax[1, j])\n",
    "    r, pvalue = stats.spearmanr(df[\"mean_LDL\"], df[\"mu_z_adj\"])\n",
    "    #ax[1, j].collections[0].label()\n",
    "    ax[1, j].set_xlabel(\"UKB LDL-C (mg/dL)\")\n",
    "    ax[1, j].set_ylabel(\"BEAN z-score\")\n",
    "    ax[1, j].text(0.98, 0.98, f\"r={r:.2f}\", transform=ax[1, j].transAxes, horizontalalignment='right', verticalalignment='top')\n",
    "    ax[1, j].set_title(f\"{label[i]} in {qbin} \")\n",
    "    ax[1, j].set_box_aspect(1)\n",
    "\n",
    "plt.setp(ax, box_aspect=1)\n",
    "plt.suptitle(f\"LDLR tiling library variant stats (n={len(score_tbl)})\")\n",
    "plt.tight_layout()\n",
    "fig.savefig(\"FigR2.pdf\", bbox_inches=\"tight\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jy_18loci",
   "language": "python",
   "name": "jy_18loci"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
