{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "297de4b0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/pinello/SHARED_SOFTWARE/anaconda_latest/envs/jy_anbe_py38/lib/python3.8/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.\n",
      "  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,\n",
      "/data/pinello/SHARED_SOFTWARE/anaconda_latest/envs/jy_anbe_py38/lib/python3.8/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Float64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.\n",
      "  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import subprocess\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from tqdm.auto import tqdm\n",
    "from patsy import dmatrices\n",
    "import statsmodels.api as sm\n",
    "import statsmodels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "08faa30a",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bean.mapping._supporting_fn import _get_allele_from_alignment\n",
    "from bean import Allele,Edit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8042cc13",
   "metadata": {},
   "outputs": [],
   "source": [
    "edit_dict = {\"A\":\"G\", \"T\":\"C\"}\n",
    "revcomp_dict = {\"A\":\"T\", \"T\":\"A\", \"C\":\"G\", \"G\":\"C\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "7a4c65c4",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_info_tbl = pd.read_excel(\"../../resources/LDLvar/20221013_LDLvar_simpleZscores_credset.xlsx\", index_col=0, header=1)\n",
    "amplicon_info = pd.read_csv(\"../../resources/atac_seq/030123_ATACseq_info.csv\", ).set_index(\"gRNA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "7846970f",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_ids = amplicon_info.index.tolist()\n",
    "#var_ids.remove(\"rs116734477_Maj_ABE_51\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a9869cce",
   "metadata": {},
   "outputs": [],
   "source": [
    "fraction_dict = {0.5:(1,1), 1:(0,2), 0:(2,0), 0.333: (2,1), 0.667:(1,2)}\n",
    "def get_variant_info(var_id):\n",
    "    var_alleles = var_info_tbl.loc[var_id, [\"A1\", \"A2\", \"HepG2 alternate allele fraction\"]]\n",
    "    print(var_alleles)\n",
    "    if var_id == \"rs771555783_Maj_ABE_309\":\n",
    "        ref_bases = [\"A\"]\n",
    "#     elif var_id == \"rs116734477_Maj_ABE_51\":\n",
    "#         return None\n",
    "    else:\n",
    "        n_A1, n_A2 = fraction_dict[var_alleles[\"HepG2 alternate allele fraction\"]]\n",
    "        ref_bases = []\n",
    "        if n_A1 > 0 and var_alleles.A1:\n",
    "            ref_bases.append(var_alleles.A1)\n",
    "        if n_A2 > 0 and var_alleles.A2:\n",
    "            ref_bases.append(var_alleles.A2)\n",
    "        assert len(ref_bases) > 0, (var_id, var_alleles)\n",
    "    #var_alleles['edited'] = var_alleles[var_id].map(lambda b: edit_dict[b] if b in edit_dict.keys() else np.nan)\n",
    "    #var_alleles = var_alleles.loc[~var_alleles.edited.isnull()]\n",
    "    var_chrom = int(var_info_tbl.loc[var_id, \"CHR\"])\n",
    "    var_pos = int(var_info_tbl.loc[var_id, 'HG19 BP'])\n",
    "    var_alleles.columns=[\"allele\", \"ref\", \"edited\"]\n",
    "    \n",
    "    amplicon_start = amplicon_info.loc[var_id, \"Varpos_fw\"]-4\n",
    "    amplicon_seq = amplicon_info.loc[var_id, \"Amplicon_fw (reference)\"].replace(\"N\", \"\")\n",
    "    amplicon_base = amplicon_seq[amplicon_start]\n",
    "        \n",
    "    return var_alleles, var_chrom, var_pos, ref_bases, amplicon_start, amplicon_base, amplicon_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "06a4f736",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rs4149309_Maj_ABE_206\n",
      "A1                                   A\n",
      "A2                                   T\n",
      "HepG2 alternate allele fraction    0.5\n",
      "Name: rs4149309_Maj_ABE_206, dtype: object\n",
      "rs10107182_Min_ABE_351\n",
      "A1                                   C\n",
      "A2                                   T\n",
      "HepG2 alternate allele fraction    1.0\n",
      "Name: rs10107182_Min_ABE_351, dtype: object\n",
      "rs3767844_Maj_ABE_192\n",
      "A1                                   A\n",
      "A2                                   G\n",
      "HepG2 alternate allele fraction    0.5\n",
      "Name: rs3767844_Maj_ABE_192, dtype: object\n",
      "rs62084210_Maj_ABE_250\n",
      "A1                                     A\n",
      "A2                                     G\n",
      "HepG2 alternate allele fraction    0.333\n",
      "Name: rs62084210_Maj_ABE_250, dtype: object\n",
      "rs115421711_Maj_ABE_45\n",
      "A1                                   A\n",
      "A2                                   G\n",
      "HepG2 alternate allele fraction    0.0\n",
      "Name: rs115421711_Maj_ABE_45, dtype: object\n",
      "rs1434282_Min_ABE_412\n",
      "A1                                   C\n",
      "A2                                   T\n",
      "HepG2 alternate allele fraction    1.0\n",
      "Name: rs1434282_Min_ABE_412, dtype: object\n",
      "rs4719853_Maj_ABE_214\n",
      "A1                                   T\n",
      "A2                                   A\n",
      "HepG2 alternate allele fraction    0.0\n",
      "Name: rs4719853_Maj_ABE_214, dtype: object\n",
      "rs771555783_Maj_ABE_309\n",
      "A1                                 CACCAT\n",
      "A2                                      C\n",
      "HepG2 alternate allele fraction       0.0\n",
      "Name: rs771555783_Maj_ABE_309, dtype: object\n",
      "rs704_Min_ABE_550\n",
      "A1                                   G\n",
      "A2                                   A\n",
      "HepG2 alternate allele fraction    1.0\n",
      "Name: rs704_Min_ABE_550, dtype: object\n",
      "rs11149612_Min_ABE_369\n",
      "A1                                   C\n",
      "A2                                   T\n",
      "HepG2 alternate allele fraction    1.0\n",
      "Name: rs11149612_Min_ABE_369, dtype: object\n",
      "rs116734477_Maj_ABE_51\n",
      "A1                                   C\n",
      "A2                                   T\n",
      "HepG2 alternate allele fraction    0.0\n",
      "Name: rs116734477_Maj_ABE_51, dtype: object\n",
      "rs76895963_Maj_ABE_306\n",
      "A1                                   T\n",
      "A2                                   G\n",
      "HepG2 alternate allele fraction    0.0\n",
      "Name: rs76895963_Maj_ABE_306, dtype: object\n",
      "rs12042481_Maj_ABE_54\n",
      "A1                                   T\n",
      "A2                                   C\n",
      "HepG2 alternate allele fraction    0.0\n",
      "Name: rs12042481_Maj_ABE_54, dtype: object\n",
      "rs58198139_Min_ABE_523\n",
      "A1                                   C\n",
      "A2                                   T\n",
      "HepG2 alternate allele fraction    1.0\n",
      "Name: rs58198139_Min_ABE_523, dtype: object\n",
      "rs2618566_Min_ABE_453\n",
      "A1                                   G\n",
      "A2                                   T\n",
      "HepG2 alternate allele fraction    0.5\n",
      "Name: rs2618566_Min_ABE_453, dtype: object\n",
      "rs35081008_Min_ABE_465\n",
      "A1                                   C\n",
      "A2                                   T\n",
      "HepG2 alternate allele fraction    0.5\n",
      "Name: rs35081008_Min_ABE_465, dtype: object\n",
      "rs4390169_Maj_ABE_208\n",
      "A1                                   A\n",
      "A2                                   G\n",
      "HepG2 alternate allele fraction    0.5\n",
      "Name: rs4390169_Maj_ABE_208, dtype: object\n",
      "rs402348_Maj_ABE_201\n",
      "A1                                   T\n",
      "A2                                   G\n",
      "HepG2 alternate allele fraction    0.5\n",
      "Name: rs402348_Maj_ABE_201, dtype: object\n",
      "rs8126001_Min_ABE_574\n",
      "A1                                   C\n",
      "A2                                   T\n",
      "HepG2 alternate allele fraction    1.0\n",
      "Name: rs8126001_Min_ABE_574, dtype: object\n",
      "rs1250259_Min_ABE_392\n",
      "A1                                     T\n",
      "A2                                     A\n",
      "HepG2 alternate allele fraction    0.667\n",
      "Name: rs1250259_Min_ABE_392, dtype: object\n"
     ]
    }
   ],
   "source": [
    "var_info_dict = {}\n",
    "for var_id in var_ids:\n",
    "    print(var_id)\n",
    "    vdf, vc, vp,var_refs, varpos_amplicon,  amplicon_base, amplicon_seq = get_variant_info(var_id)\n",
    "    vdict = {}\n",
    "    vdict[\"chrom\"] = vc\n",
    "    vdict[\"pos\"] = vp\n",
    "    vdict[\"amplicon_pos\"] = varpos_amplicon\n",
    "    vdict[\"refs\"] = var_refs\n",
    "    vdict[\"amplicon_base\"] = amplicon_base\n",
    "    vdict[\"amplicon_seq\"] = amplicon_seq\n",
    "    vdict[\"vdf\"] = vdf\n",
    "    var_info_dict[var_id] = vdict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d2cc9c95",
   "metadata": {},
   "outputs": [],
   "source": [
    "def _get_allele_from_alignment(\n",
    "    ref_aligned: str,\n",
    "    query_aligned: str,\n",
    "    offset: int,\n",
    "    strand: int,\n",
    "    start_pos: int,\n",
    "    end_pos: int,\n",
    "    positionwise_quality: np.ndarray = None,\n",
    "    quality_thres: float = -1,\n",
    "):\n",
    "    # Include N, no quality filter\n",
    "    assert len(ref_aligned) == len(query_aligned)\n",
    "    allele = Allele()\n",
    "    ref_gaps = 0\n",
    "    alt_gaps = 0\n",
    "    alt_seq_len = len(query_aligned) - query_aligned.count(\"-\")\n",
    "    if positionwise_quality is None:\n",
    "        # alt_position_is_good_quality = np.ones(alt_seq_len, dtype=bool)\n",
    "        alt_position_is_good_quality = np.array(\n",
    "            [c != \"N\" for c in query_aligned.replace(\"-\", \"\")]\n",
    "        )\n",
    "    else:\n",
    "        assert len(positionwise_quality) == alt_seq_len\n",
    "        alt_position_is_good_quality = positionwise_quality > quality_thres\n",
    "    for i in range(len(ref_aligned)):\n",
    "        if ref_aligned[i] == query_aligned[i]:\n",
    "            continue\n",
    "        ref_base = ref_aligned[i]\n",
    "        alt_base = query_aligned[i]\n",
    "        if alt_base != \"-\":\n",
    "            alt_base_is_good_quality = alt_position_is_good_quality[i - alt_gaps]\n",
    "        else:\n",
    "            alt_base_is_good_quality = True\n",
    "        if ref_base == \"-\":\n",
    "            ref_gaps += 1\n",
    "        elif alt_base == \"-\":\n",
    "            alt_gaps += 1\n",
    "        ref_pos = i - ref_gaps\n",
    "        allele.add(\n",
    "            Edit(\n",
    "                rel_pos=ref_pos,\n",
    "                ref_base=ref_base,\n",
    "                alt_base=alt_base,\n",
    "                offset=offset,\n",
    "                strand=strand,\n",
    "            )\n",
    "        )\n",
    "    return allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "172a63f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_edit(row, target_pos):\n",
    "    return _get_allele_from_alignment(row.Reference_Sequence, row.Aligned_Sequence, offset=0, strand=1, start_pos=target_pos-20, end_pos=target_pos+20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3e206ebd",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_snp_pos(var_id):\n",
    "    var_file = \"/PHShome/jr1025/projects/ANBE/data/HepG2_reference/ENCFF713BPG.vcf.gz\"\n",
    "    vdf, vc, vp, varpos_amplicon, amplicon_base, amplicon_seq = get_variant_info(var_id)\n",
    "    pos = f\"{vc}:{vp-varpos_amplicon + 20}-{vp+len(amplicon_seq)-varpos_amplicon-20}\"\n",
    "    command = f\"bcftools query -r {pos} -f '%POS\\t%REF\\t%ALT'\\n {var_file}\"\n",
    "    proc = subprocess.Popen((\"module load bcftools; \"+command).split(\" \"), stdout=subprocess.PIPE, shell=True)\n",
    "    (out, err) = proc.communicate()\n",
    "    return out, err\n",
    "\n",
    "def get_pos_edit_rate(aftbl):\n",
    "    total_counts = aftbl['#Reads'].sum()\n",
    "    aftbl = aftbl.loc[aftbl.be_allele != \"\",:].copy()\n",
    "    aftbl['edits'] = aftbl.be_allele.map(lambda a: a.split(\",\"))\n",
    "    edit_tbl = aftbl.explode(\"edits\")\n",
    "    edit_tbl['pos'] = edit_tbl.edits.map(lambda e: int(e.split(\":\")[0]))\n",
    "    per_pos_edits = edit_tbl.groupby(\"pos\")[\"#Reads\"].sum()\n",
    "    per_pos_edit_rates = per_pos_edits / aftbl['#Reads'].sum()\n",
    "    return np.where(per_pos_edit_rates > 0.5)[0]\n",
    "    \n",
    "def format_allele_freq_table(path, varpos_amplicon):\n",
    "    aftbl = pd.read_csv(path, sep='\\t')\n",
    "    aftbl[\"allele\"] = aftbl.apply(lambda row:get_edit(row, varpos_amplicon),axis=1)\n",
    "    aftbl['be_allele'] = aftbl.allele.map(filter_out_indels)\n",
    "    alt_positions = get_pos_edit_rate(aftbl)\n",
    "    aftbl['err'] = aftbl.be_allele.map(lambda a:filter_allele_by_pos(a, alt_positions, varpos_amplicon)[0])\n",
    "    aftbl['N'] = aftbl.be_allele.map(lambda a:filter_allele_by_pos(a, alt_positions, varpos_amplicon)[1])\n",
    "    aftbl['target_proximal_N'] = aftbl.be_allele.map(lambda a:filter_allele_by_pos(a, alt_positions, varpos_amplicon)[2])\n",
    "    aftbl['pos_allele'] = aftbl.be_allele.map(lambda a:filter_allele_by_pos(a, alt_positions, varpos_amplicon)[3])\n",
    "    aftbl['clean'] = aftbl.err <= 1\n",
    "    return aftbl\n",
    "\n",
    "def get_edit_counts(aftbl, target_pos, amplicon_base):\n",
    "    ref, alt = zip(*aftbl.pos_allele.map(lambda a: get_pos_base(a, target_pos, amplicon_base)))\n",
    "    aftbl['ref'] = ref\n",
    "    aftbl['alt'] = alt\n",
    "    return aftbl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "be1b7c34",
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_allele_by_pos(allele, list_pos, target_pos):\n",
    "    # filters allele by variant included in list_pos or near target_pos\n",
    "    edits = []\n",
    "    n_out_edits = 0\n",
    "    n_lq = 0\n",
    "    n_crit_lq = 0\n",
    "    for edit in allele.split(\",\"):\n",
    "        if edit == \"\": continue\n",
    "        alt = edit.split(\">\")[-1]\n",
    "        pos = int(edit.split(\":\")[0])\n",
    "        if alt == \"N\": \n",
    "            n_lq += 1\n",
    "            if abs(pos-target_pos) < 5:\n",
    "                n_crit_lq += 1\n",
    "        elif abs(pos-target_pos) < 5:\n",
    "            edits.append(edit)\n",
    "        else:\n",
    "            n_out_edits += 1\n",
    "    return n_out_edits, n_lq, n_crit_lq, \",\".join(sorted(edits))\n",
    "\n",
    "def get_pos_base(allele, target_pos, amplicon_base):\n",
    "    edits = []\n",
    "    for edit in allele.split(\",\"):\n",
    "        if edit == \"\": continue\n",
    "        pos = int(edit.split(\":\")[0])\n",
    "        if pos == target_pos:\n",
    "            ref, alt = edit.split(\":\")[-1].split(\">\")\n",
    "            return ref, alt\n",
    "    return amplicon_base, amplicon_base\n",
    "\n",
    "def filter_out_indels(allele):\n",
    "    edits = []\n",
    "    for edit in allele.edits:\n",
    "        if edit.ref_base != \"-\" and edit.alt_base != \"-\":\n",
    "            edits.append(str(edit))\n",
    "    return \",\".join(sorted(edits))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "ba780ffa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rs4149309_Maj_ABE_206\n",
      "rs10107182_Min_ABE_351\n",
      "rs3767844_Maj_ABE_192\n",
      "rs62084210_Maj_ABE_250\n",
      "rs115421711_Maj_ABE_45\n",
      "rs1434282_Min_ABE_412\n",
      "rs4719853_Maj_ABE_214\n",
      "rs771555783_Maj_ABE_309\n",
      "../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_1-Ser_20-locus_GDNA.rs771555783_Maj_ABE_309/rs771555783_Maj_ABE_3.Nucleotide_frequency_table.txt not found\n",
      "../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_2-Ser_20-locus_GDNA.rs771555783_Maj_ABE_309/rs771555783_Maj_ABE_3.Nucleotide_frequency_table.txt not found\n",
      "../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_3-Ser_20-locus_GDNA.rs771555783_Maj_ABE_309/rs771555783_Maj_ABE_3.Nucleotide_frequency_table.txt not found\n",
      "../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_3-SS_20-locus_GDNA.rs771555783_Maj_ABE_309/rs771555783_Maj_ABE_3.Nucleotide_frequency_table.txt not found\n",
      "../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_C-Ser_20-locus_GDNA.rs771555783_Maj_ABE_309/rs771555783_Maj_ABE_3.Nucleotide_frequency_table.txt not found\n",
      "rs704_Min_ABE_550\n",
      "rs11149612_Min_ABE_369\n",
      "rs116734477_Maj_ABE_51\n",
      "rs76895963_Maj_ABE_306\n",
      "rs12042481_Maj_ABE_54\n",
      "rs58198139_Min_ABE_523\n",
      "rs2618566_Min_ABE_453\n",
      "rs35081008_Min_ABE_465\n",
      "rs4390169_Maj_ABE_208\n",
      "rs402348_Maj_ABE_201\n",
      "rs8126001_Min_ABE_574\n",
      "rs1250259_Min_ABE_392\n"
     ]
    }
   ],
   "source": [
    "counts_list = []\n",
    "\n",
    "for i, var_id in enumerate(var_ids):\n",
    "    print(var_id)\n",
    "    varpos_amplicon = var_info_dict[var_id][\"amplicon_pos\"]\n",
    "    amplicon_base = var_info_dict[var_id][\"amplicon_base\"]\n",
    "    for rep in [1,2,3,\"C\"]:\n",
    "        for cond in [\"Ser\", \"SS\"]:\n",
    "            for exp in [\"ATAC_Seq\", \"GDNA\"]:\n",
    "                try:\n",
    "                    nt_freq_path = f\"../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_{rep}-{cond}_20-locus_{exp}.{var_id}/{var_id[:21]}.Nucleotide_frequency_table.txt\"\n",
    "                    nt_freq_tbl = pd.read_csv(nt_freq_path, sep='\\t', header=None, index_col = 0)\n",
    "                except FileNotFoundError:\n",
    "                    print(f\"{nt_freq_path} not found\")\n",
    "                    continue\n",
    "                \n",
    "                refs = nt_freq_tbl.iloc[0,:].tolist()\n",
    "                nt_freq_tbl = nt_freq_tbl.iloc[1:,:]\n",
    "                target_pos_edits = nt_freq_tbl.iloc[:,[varpos_amplicon]].reset_index()\n",
    "                assert refs[varpos_amplicon] == amplicon_base\n",
    "                target_pos_edits.columns=[\"base\", \"count\"]\n",
    "                target_pos_edits[\"amplicon_ref\"] = refs[varpos_amplicon]\n",
    "                target_pos_edits[\"var_id\"] = var_id\n",
    "                target_pos_edits[\"rep\"] = rep\n",
    "                target_pos_edits[\"cond\"] = cond\n",
    "                target_pos_edits[\"exp\"] = exp\n",
    "                counts_list.append(target_pos_edits)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b0351376",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_error_stats(aftbl, axs, label=\"\", title=\"\"):\n",
    "    sns.scatterplot(data=aftbl.groupby(\"err\")[\"#Reads\"].sum().reset_index(), x='err', y='#Reads', ax = axs[0], s=10)\n",
    "    axs[0].set_title(\"Mismatch\")\n",
    "    axs[0].set_ylabel(f\"{title}\\n#Reads\")\n",
    "    sns.scatterplot(data=aftbl.groupby(\"N\")[\"#Reads\"].sum().reset_index(), x='N', y='#Reads', ax = axs[1], s= 10)\n",
    "    axs[1].set_title(\"# Low quality nt\")\n",
    "    sns.scatterplot(data=aftbl.groupby(\"target_proximal_N\")[\"#Reads\"].sum().reset_index(), x='target_proximal_N', y='#Reads', ax = axs[2], label=label, s = 10)\n",
    "    axs[2].set_title(\"# Low quality nt near target\")\n",
    "    axs[2].get_legend().remove()\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "fa4512ed",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_df = pd.concat(counts_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "11735e68",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_valid(ref, alt, var_id):\n",
    "    '''Checks if ref>alt base is valid for var_id. \n",
    "    Returns 0 for ref, 1 for alt, and -1 for invalid'''\n",
    "    var_info = var_info_dict[var_id]\n",
    "    assert ref ==  var_info[\"amplicon_base\"]\n",
    "    if alt == \"N\":\n",
    "        return -1\n",
    "    if ref not in var_info[\"refs\"]:\n",
    "        # ex) \n",
    "        assert len(var_info[\"refs\"]) == 1\n",
    "        ref_base = var_info[\"refs\"][0]\n",
    "        if alt == ref_base: return 0\n",
    "        if alt == edit_dict[ref_base]: return 1\n",
    "        return -1\n",
    "    possible_alts = []\n",
    "    for r in var_info['refs']:\n",
    "        if r in edit_dict.keys():\n",
    "            if r == ref: continue\n",
    "            possible_alts.append(edit_dict[r])\n",
    "            \n",
    "    if ref in edit_dict.keys():\n",
    "        if edit_dict[ref] in var_info[\"refs\"]:\n",
    "            # Ambiguous edited vs other phase's WT\n",
    "            return 0.5\n",
    "        if alt in possible_alts:\n",
    "            # Ambiguous WT vs other phase's edited\n",
    "            return 0.5\n",
    "        if edit_dict[ref] == alt: return 1\n",
    "        if ref == alt: return 0\n",
    "    phases = var_info[\"vdf\"][:2].tolist()\n",
    "    phases.remove(ref)\n",
    "    if ref == alt: return 0\n",
    "    if alt == phases[0]: return 0\n",
    "    if phases[0] in edit_dict.keys() and alt == edit_dict[phases[0]]: return 1\n",
    "    else:\n",
    "        print(ref, alt, var_id)\n",
    "        return -1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "34a7a1f4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A - rs4149309_Maj_ABE_206\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "A C rs115421711_Maj_ABE_45\n",
      "A T rs115421711_Maj_ABE_45\n",
      "A - rs115421711_Maj_ABE_45\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "T - rs4719853_Maj_ABE_214\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "C A rs116734477_Maj_ABE_51\n",
      "C G rs116734477_Maj_ABE_51\n",
      "C - rs116734477_Maj_ABE_51\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs76895963_Maj_ABE_306\n",
      "T - rs76895963_Maj_ABE_306\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "T A rs12042481_Maj_ABE_54\n",
      "T G rs12042481_Maj_ABE_54\n",
      "T - rs12042481_Maj_ABE_54\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "G A rs2618566_Min_ABE_453\n",
      "G - rs2618566_Min_ABE_453\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "C A rs35081008_Min_ABE_465\n",
      "C G rs35081008_Min_ABE_465\n",
      "C - rs35081008_Min_ABE_465\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T A rs402348_Maj_ABE_201\n",
      "T - rs402348_Maj_ABE_201\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n",
      "T - rs1250259_Min_ABE_392\n"
     ]
    }
   ],
   "source": [
    "counts_df[\"valid_mt\"] = counts_df.apply(lambda row: check_valid(row.amplicon_ref, row.base, row.var_id), axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "323c5a02",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "var_counts_dict = {}\n",
    "failed_variants = [\"rs3767844_Maj_ABE_192\", \"rs62084210_Maj_ABE_250\", \"rs116734477_Maj_ABE_51\", \"rs35081008_Min_ABE_465\", \"rs4390169_Maj_ABE_208\"] + [\"rs771555783_Maj_ABE_309\"] # needs phasing, not enough reads\n",
    "pdfs = []\n",
    "edit_assignments = []\n",
    "for var_id in var_ids:\n",
    "    pdf = pd.pivot(counts_df.loc[counts_df.var_id == var_id], index=[\"var_id\", \"exp\", \"cond\", \"rep\",], columns=[\"amplicon_ref\", \"base\",], values='count').astype(float)\n",
    "    if len(pdf) == 0:\n",
    "        failed_variants.append(var_id)\n",
    "        print(f\"ERROR: {var_id} does not have reads.\")\n",
    "    edit_assignment = counts_df.loc[counts_df.var_id == var_id, [\"amplicon_ref\", \"base\", \"valid_mt\"]].drop_duplicates()\n",
    "    if len(edit_assignment.valid_mt.unique()) == 1:\n",
    "        print(f\"WARNING:{var_id} cannot be resolved.\")\n",
    "        if var_id not in failed_variants:\n",
    "            print(f\"ERROR: {var_id} not expected to fail but is not resolved.\")\n",
    "        failed_variants.append(var_id)\n",
    "    pdf_t = pdf.T\n",
    "    pdf_t['valid_mt'] = edit_assignment.set_index([\"amplicon_ref\", \"base\"])\n",
    "    pdf_t = pdf_t.groupby(\"valid_mt\").sum()\n",
    "    pdf = pdf_t.T\n",
    "    pdf.columns = pdf.columns.map({0:\"ref\", 0.5:\"ambiguous\", 1:\"alt\"})\n",
    "    var_counts_dict[var_id] = (pdf, edit_assignment)\n",
    "    pdf['status'] = {True:\"FAILED\", False:\"SUCCESS\"}[len(edit_assignment.valid_mt.unique()) == 1]\n",
    "    pdfs.append(pdf)\n",
    "    edit_assignment['label'] = edit_assignment.valid_mt.map({0:\"ref\", 0.5:\"ambiguous\", 1:\"alt\"})\n",
    "    edit_assignment[\"var_id\"] = var_id\n",
    "    edit_assignment[\"status\"] = {True:\"FAILED\", False:\"SUCCESS\"}[len(edit_assignment.valid_mt.unique()) == 1]\n",
    "    edit_assignments.append(edit_assignment[[\"var_id\", \"amplicon_ref\", \"base\", \"label\", \"status\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a4655139",
   "metadata": {},
   "outputs": [],
   "source": [
    "resolved_counts = pd.concat(pdfs).reset_index()\n",
    "edit_assignment_all = pd.concat(edit_assignments).drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f0bcb4bb",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rs4149309_Maj_ABE_206\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                        9\n",
      "Model Family:                Binomial   Df Model:                            6\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -11.082\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       5.5180\n",
      "Time:                        15:38:37   Pearson chi2:                     5.66\n",
      "No. Iterations:                    27                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -5.5197      0.454    -12.167      0.000      -6.409      -4.631\n",
      "rep[T.2]                       -2.2212      0.757     -2.934      0.003      -3.705      -0.737\n",
      "rep[T.3]                       -2.8474      1.036     -2.750      0.006      -4.877      -0.818\n",
      "rep[T.C]                      -23.5729   1.78e+04     -0.001      0.999   -3.49e+04    3.48e+04\n",
      "cond[T.Ser]                     1.1953      0.562      2.127      0.033       0.094       2.297\n",
      "is_atac[T.True]                 0.7974      0.735      1.085      0.278      -0.643       2.238\n",
      "cond[T.Ser]:is_atac[T.True]    -2.7301      1.290     -2.117      0.034      -5.258      -0.203\n",
      "===============================================================================================\n",
      "rs10107182_Min_ABE_351\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                        9\n",
      "Model Family:                Binomial   Df Model:                            6\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -29.081\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       11.332\n",
      "Time:                        15:38:37   Pearson chi2:                     13.6\n",
      "No. Iterations:                    12                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -7.1577      0.282    -25.347      0.000      -7.711      -6.604\n",
      "rep[T.2]                        0.0552      0.304      0.181      0.856      -0.541       0.651\n",
      "rep[T.3]                       -1.2151      0.403     -3.013      0.003      -2.006      -0.425\n",
      "rep[T.C]                       -1.2373      0.440     -2.815      0.005      -2.099      -0.376\n",
      "cond[T.Ser]                    -0.2440      0.358     -0.682      0.495      -0.945       0.457\n",
      "is_atac[T.True]                 0.1524      0.319      0.478      0.632      -0.472       0.777\n",
      "cond[T.Ser]:is_atac[T.True]    -1.1280      0.590     -1.913      0.056      -2.284       0.028\n",
      "===============================================================================================\n",
      "rs115421711_Maj_ABE_45\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                        9\n",
      "Model Family:                Binomial   Df Model:                            6\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -73.817\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       57.602\n",
      "Time:                        15:38:37   Pearson chi2:                     63.2\n",
      "No. Iterations:                    13                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -4.4849      0.035   -128.236      0.000      -4.553      -4.416\n",
      "rep[T.2]                       -3.4895      0.081    -43.310      0.000      -3.647      -3.332\n",
      "rep[T.3]                       -4.3967      0.130    -33.776      0.000      -4.652      -4.142\n",
      "rep[T.C]                       -4.9754      0.183    -27.125      0.000      -5.335      -4.616\n",
      "cond[T.Ser]                    -0.0459      0.056     -0.822      0.411      -0.155       0.064\n",
      "is_atac[T.True]                -0.0059      0.042     -0.140      0.889      -0.089       0.077\n",
      "cond[T.Ser]:is_atac[T.True]    -0.6761      0.066    -10.225      0.000      -0.806      -0.546\n",
      "===============================================================================================\n",
      "rs1434282_Min_ABE_412\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                        9\n",
      "Model Family:                Binomial   Df Model:                            6\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -8.5571\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       1.0637\n",
      "Time:                        15:38:37   Pearson chi2:                    0.713\n",
      "No. Iterations:                    28                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -6.1011      0.397    -15.359      0.000      -6.880      -5.323\n",
      "rep[T.2]                        0.4963      0.443      1.119      0.263      -0.373       1.365\n",
      "rep[T.3]                      -23.4912   2.69e+04     -0.001      0.999   -5.27e+04    5.27e+04\n",
      "rep[T.C]                      -23.5332   2.44e+04     -0.001      0.999   -4.78e+04    4.78e+04\n",
      "cond[T.Ser]                     0.1899      0.432      0.440      0.660      -0.657       1.036\n",
      "is_atac[T.True]               -22.6561   8.33e+04     -0.000      1.000   -1.63e+05    1.63e+05\n",
      "cond[T.Ser]:is_atac[T.True]    22.1438   8.33e+04      0.000      1.000   -1.63e+05    1.63e+05\n",
      "===============================================================================================\n",
      "rs4719853_Maj_ABE_214\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                        9\n",
      "Model Family:                Binomial   Df Model:                            6\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -27.938\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       10.498\n",
      "Time:                        15:38:37   Pearson chi2:                     9.79\n",
      "No. Iterations:                    13                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                     -10.2629      0.334    -30.698      0.000     -10.918      -9.608\n",
      "rep[T.2]                        0.5358      0.369      1.451      0.147      -0.188       1.259\n",
      "rep[T.3]                        0.4955      0.368      1.346      0.178      -0.226       1.217\n",
      "rep[T.C]                        0.5410      0.390      1.388      0.165      -0.223       1.305\n",
      "cond[T.Ser]                     0.2567      0.287      0.895      0.371      -0.306       0.819\n",
      "is_atac[T.True]                 0.6367      0.410      1.554      0.120      -0.166       1.440\n",
      "cond[T.Ser]:is_atac[T.True]    -0.6387      0.605     -1.056      0.291      -1.824       0.547\n",
      "===============================================================================================\n",
      "rs704_Min_ABE_550\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                        9\n",
      "Model Family:                Binomial   Df Model:                            6\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -307.32\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       477.65\n",
      "Time:                        15:38:37   Pearson chi2:                     478.\n",
      "No. Iterations:                    12                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -4.2007      0.013   -321.625      0.000      -4.226      -4.175\n",
      "rep[T.2]                       -0.0877      0.013     -6.959      0.000      -0.112      -0.063\n",
      "rep[T.3]                       -1.2294      0.017    -72.754      0.000      -1.263      -1.196\n",
      "rep[T.C]                       -4.4165      0.079    -56.143      0.000      -4.571      -4.262\n",
      "cond[T.Ser]                     0.0157      0.017      0.913      0.361      -0.018       0.049\n",
      "is_atac[T.True]                -0.2151      0.016    -13.255      0.000      -0.247      -0.183\n",
      "cond[T.Ser]:is_atac[T.True]     0.2247      0.023      9.672      0.000       0.179       0.270\n",
      "===============================================================================================\n",
      "***Significant differential enrichment of ALT allele in ATAC-seq compared to GDNA: dlogit=-0.21509735003272953***\n",
      "rs11149612_Min_ABE_369\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                        9\n",
      "Model Family:                Binomial   Df Model:                            6\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -787.63\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       1410.9\n",
      "Time:                        15:38:37   Pearson chi2:                 1.44e+03\n",
      "No. Iterations:                    12                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -2.5788      0.004   -580.986      0.000      -2.587      -2.570\n",
      "rep[T.2]                       -1.0333      0.006   -165.925      0.000      -1.046      -1.021\n",
      "rep[T.3]                        0.8306      0.004    197.565      0.000       0.822       0.839\n",
      "rep[T.C]                       -6.0729      0.072    -84.252      0.000      -6.214      -5.932\n",
      "cond[T.Ser]                    -0.0157      0.005     -3.156      0.002      -0.026      -0.006\n",
      "is_atac[T.True]                 0.0091      0.005      1.954      0.051   -2.61e-05       0.018\n",
      "cond[T.Ser]:is_atac[T.True]     0.0961      0.007     13.705      0.000       0.082       0.110\n",
      "===============================================================================================\n",
      "rs76895963_Maj_ABE_306\n",
      "valid_mt                  var_id       exp cond rep   NaN     ref  ambiguous  \\\n",
      "171       rs76895963_Maj_ABE_306  ATAC_Seq   SS   1  24.0  1134.0        NaN   \n",
      "172       rs76895963_Maj_ABE_306  ATAC_Seq   SS   2  10.0   489.0        NaN   \n",
      "173       rs76895963_Maj_ABE_306  ATAC_Seq   SS   3  33.0  2163.0        NaN   \n",
      "174       rs76895963_Maj_ABE_306  ATAC_Seq   SS   C  14.0   799.0        NaN   \n",
      "175       rs76895963_Maj_ABE_306  ATAC_Seq  Ser   1  25.0  1442.0        NaN   \n",
      "176       rs76895963_Maj_ABE_306  ATAC_Seq  Ser   2  16.0  1202.0        NaN   \n",
      "177       rs76895963_Maj_ABE_306  ATAC_Seq  Ser   3  49.0  2596.0        NaN   \n",
      "178       rs76895963_Maj_ABE_306  ATAC_Seq  Ser   C  16.0  1384.0        NaN   \n",
      "179       rs76895963_Maj_ABE_306      GDNA   SS   1  18.0   883.0        NaN   \n",
      "180       rs76895963_Maj_ABE_306      GDNA   SS   2  11.0   956.0        NaN   \n",
      "181       rs76895963_Maj_ABE_306      GDNA   SS   3   7.0   572.0        NaN   \n",
      "182       rs76895963_Maj_ABE_306      GDNA   SS   C  18.0  1141.0        NaN   \n",
      "183       rs76895963_Maj_ABE_306      GDNA  Ser   1  14.0   664.0        NaN   \n",
      "184       rs76895963_Maj_ABE_306      GDNA  Ser   2  15.0   849.0        NaN   \n",
      "185       rs76895963_Maj_ABE_306      GDNA  Ser   3  15.0   770.0        NaN   \n",
      "186       rs76895963_Maj_ABE_306      GDNA  Ser   C  22.0   789.0        NaN   \n",
      "\n",
      "valid_mt  alt   status sample  is_atac  is_trt  \n",
      "171       0.0  SUCCESS   1_SS     True    True  \n",
      "172       1.0  SUCCESS   2_SS     True    True  \n",
      "173       0.0  SUCCESS   3_SS     True    True  \n",
      "174       0.0  SUCCESS   C_SS     True   False  \n",
      "175       0.0  SUCCESS  1_Ser     True    True  \n",
      "176       0.0  SUCCESS  2_Ser     True    True  \n",
      "177       0.0  SUCCESS  3_Ser     True    True  \n",
      "178       0.0  SUCCESS  C_Ser     True   False  \n",
      "179       0.0  SUCCESS   1_SS    False    True  \n",
      "180       0.0  SUCCESS   2_SS    False    True  \n",
      "181       0.0  SUCCESS   3_SS    False    True  \n",
      "182       0.0  SUCCESS   C_SS    False   False  \n",
      "183       0.0  SUCCESS  1_Ser    False    True  \n",
      "184       0.0  SUCCESS  2_Ser    False    True  \n",
      "185       0.0  SUCCESS  3_Ser    False    True  \n",
      "186       0.0  SUCCESS  C_Ser    False   False  \n",
      "rs12042481_Maj_ABE_54\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                        9\n",
      "Model Family:                Binomial   Df Model:                            6\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -61.024\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       26.763\n",
      "Time:                        15:38:37   Pearson chi2:                     28.1\n",
      "No. Iterations:                    12                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -5.2129      0.056    -92.816      0.000      -5.323      -5.103\n",
      "rep[T.2]                        3.0878      0.035     87.571      0.000       3.019       3.157\n",
      "rep[T.3]                       -0.6330      0.060    -10.546      0.000      -0.751      -0.515\n",
      "rep[T.C]                       -3.9046      0.291    -13.427      0.000      -4.475      -3.335\n",
      "cond[T.Ser]                     0.0112      0.063      0.177      0.859      -0.113       0.135\n",
      "is_atac[T.True]                 0.1031      0.048      2.147      0.032       0.009       0.197\n",
      "cond[T.Ser]:is_atac[T.True]    -0.0414      0.066     -0.628      0.530      -0.171       0.088\n",
      "===============================================================================================\n",
      "***Significant differential enrichment of ALT allele in ATAC-seq compared to GDNA: dlogit=0.10305288359211778***\n",
      "rs58198139_Min_ABE_523\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                        9\n",
      "Model Family:                Binomial   Df Model:                            6\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -23.771\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       7.0675\n",
      "Time:                        15:38:37   Pearson chi2:                     6.22\n",
      "No. Iterations:                    11                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -6.4883      0.394    -16.465      0.000      -7.261      -5.716\n",
      "rep[T.2]                        2.7928      0.390      7.170      0.000       2.029       3.556\n",
      "rep[T.3]                        0.2296      0.518      0.443      0.658      -0.787       1.246\n",
      "rep[T.C]                       -2.0004      1.070     -1.869      0.062      -4.098       0.097\n",
      "cond[T.Ser]                    -0.1116      0.215     -0.519      0.603      -0.533       0.309\n",
      "is_atac[T.True]                -1.1797      0.315     -3.744      0.000      -1.797      -0.562\n",
      "cond[T.Ser]:is_atac[T.True]     0.5543      0.393      1.412      0.158      -0.215       1.324\n",
      "===============================================================================================\n",
      "***Significant differential enrichment of ALT allele in ATAC-seq compared to GDNA: dlogit=-1.1796692175172452***\n",
      "rs2618566_Min_ABE_453\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                        9\n",
      "Model Family:                Binomial   Df Model:                            6\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -428.32\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       732.77\n",
      "Time:                        15:38:38   Pearson chi2:                     733.\n",
      "No. Iterations:                    13                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -3.9656      0.020   -201.078      0.000      -4.004      -3.927\n",
      "rep[T.2]                       -0.0380      0.014     -2.707      0.007      -0.066      -0.010\n",
      "rep[T.3]                       -1.0868      0.021    -52.776      0.000      -1.127      -1.046\n",
      "rep[T.C]                       -5.7967      0.209    -27.764      0.000      -6.206      -5.388\n",
      "cond[T.Ser]                     0.0201      0.025      0.811      0.417      -0.028       0.069\n",
      "is_atac[T.True]                -0.3687      0.023    -16.195      0.000      -0.413      -0.324\n",
      "cond[T.Ser]:is_atac[T.True]     0.6528      0.030     22.110      0.000       0.595       0.711\n",
      "===============================================================================================\n",
      "***Significant differential enrichment of ALT allele in ATAC-seq compared to GDNA: dlogit=-0.3686817784602553***\n",
      "rs402348_Maj_ABE_201\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                        9\n",
      "Model Family:                Binomial   Df Model:                            6\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -42.907\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       14.503\n",
      "Time:                        15:38:38   Pearson chi2:                     14.6\n",
      "No. Iterations:                    12                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -4.5271      0.043   -105.122      0.000      -4.612      -4.443\n",
      "rep[T.2]                        1.2180      0.045     26.897      0.000       1.129       1.307\n",
      "rep[T.3]                       -1.4652      0.101    -14.456      0.000      -1.664      -1.267\n",
      "rep[T.C]                       -4.4044      0.319    -13.823      0.000      -5.029      -3.780\n",
      "cond[T.Ser]                    -0.1223      0.040     -3.046      0.002      -0.201      -0.044\n",
      "is_atac[T.True]                -0.0573      0.197     -0.291      0.771      -0.444       0.329\n",
      "cond[T.Ser]:is_atac[T.True]    -0.0987      0.281     -0.351      0.726      -0.650       0.453\n",
      "===============================================================================================\n",
      "rs8126001_Min_ABE_574\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                        9\n",
      "Model Family:                Binomial   Df Model:                            6\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -57.004\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       14.754\n",
      "Time:                        15:38:38   Pearson chi2:                     13.4\n",
      "No. Iterations:                     8                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -4.4586      0.037   -120.440      0.000      -4.531      -4.386\n",
      "rep[T.2]                        0.1655      0.047      3.549      0.000       0.074       0.257\n",
      "rep[T.3]                        0.2798      0.042      6.696      0.000       0.198       0.362\n",
      "rep[T.C]                       -0.3538      0.052     -6.755      0.000      -0.456      -0.251\n",
      "cond[T.Ser]                     0.0118      0.040      0.299      0.765      -0.066       0.089\n",
      "is_atac[T.True]                 0.6242      0.088      7.069      0.000       0.451       0.797\n",
      "cond[T.Ser]:is_atac[T.True]    -0.0034      0.126     -0.027      0.979      -0.250       0.243\n",
      "===============================================================================================\n",
      "***Significant differential enrichment of ALT allele in ATAC-seq compared to GDNA: dlogit=0.6241737454958258***\n",
      "rs1250259_Min_ABE_392\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                        9\n",
      "Model Family:                Binomial   Df Model:                            6\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -43.722\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       21.272\n",
      "Time:                        15:38:38   Pearson chi2:                     19.7\n",
      "No. Iterations:                    13                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -9.8510      0.196    -50.247      0.000     -10.235      -9.467\n",
      "rep[T.2]                        0.1073      0.199      0.538      0.591      -0.284       0.498\n",
      "rep[T.3]                        0.0969      0.208      0.467      0.641      -0.310       0.504\n",
      "rep[T.C]                       -0.1394      0.238     -0.585      0.559      -0.607       0.328\n",
      "cond[T.Ser]                    -0.1279      0.229     -0.557      0.577      -0.578       0.322\n",
      "is_atac[T.True]                -0.0556      0.194     -0.286      0.775      -0.437       0.326\n",
      "cond[T.Ser]:is_atac[T.True]    -0.0611      0.299     -0.204      0.838      -0.647       0.525\n",
      "===============================================================================================\n"
     ]
    }
   ],
   "source": [
    "enrichment_results = []\n",
    "ps = []\n",
    "bs = []\n",
    "ci_high = []\n",
    "ci_low = []\n",
    "interaction_ps = []\n",
    "interaction_bs = []\n",
    "interaction_ci_high = []\n",
    "interaction_ci_low = []\n",
    "for var_id in var_ids:\n",
    "    if var_id in failed_variants: \n",
    "        ps.append(np.nan)\n",
    "        bs.append(np.nan)\n",
    "        ci_high.append(np.nan)\n",
    "        ci_low.append(np.nan)\n",
    "        interaction_ps.append(np.nan)\n",
    "        interaction_bs.append(np.nan)\n",
    "        interaction_ci_high.append(np.nan)\n",
    "        interaction_ci_low.append(np.nan)\n",
    "        continue\n",
    "    print(var_id)\n",
    "    pdf = resolved_counts.loc[resolved_counts.var_id == var_id].copy()\n",
    "    pdf['sample'] = pdf[\"rep\"].astype(str) + \"_\" + pdf[\"cond\"]\n",
    "    pdf[\"is_atac\"] = pdf.exp == \"ATAC_Seq\"\n",
    "    pdf[\"is_trt\"] = pdf.rep != \"C\"\n",
    "    #pdf = pdf.loc[pdf.rep != \"C\"]\n",
    "    y, X = dmatrices('alt + ref ~ rep + cond + is_atac + cond*is_atac', data = pdf, return_type='dataframe')\n",
    "    try:\n",
    "        glm_binom = sm.GLM(y, X, family=sm.families.Binomial())\n",
    "        glm_res = glm_binom.fit()\n",
    "        print(glm_res.summary())\n",
    "    except:\n",
    "        print(pdf)\n",
    "        ps.append(np.nan)\n",
    "        bs.append(np.nan)\n",
    "        ci_high.append(np.nan)\n",
    "        ci_low.append(np.nan)\n",
    "        interaction_ps.append(np.nan)\n",
    "        interaction_bs.append(np.nan)\n",
    "        interaction_ci_high.append(np.nan)\n",
    "        interaction_ci_low.append(np.nan)\n",
    "        continue\n",
    "    if glm_res.pvalues[\"is_atac[T.True]\"] < 0.05: \n",
    "        print(f\"***Significant differential enrichment of ALT allele in ATAC-seq compared to GDNA: dlogit={glm_res.params['is_atac[T.True]']}***\")\n",
    "    df = pd.DataFrame(glm_res.summary().tables[1])\n",
    "    df.columns=df.iloc[0,:]\n",
    "    df = df.iloc[1:,:]\n",
    "    df.insert(0,\"variant\", var_id)\n",
    "    df.columns=[\"variant\", \"coef label\", \"coef\", \"std err\", \"z\", \"P>|z|\", \"[0.025\", \"0.975]\"]\n",
    "    enrichment_results.append(df)\n",
    "    ps.append(glm_res.pvalues[\"is_atac[T.True]\"])\n",
    "    bs.append(glm_res.params['is_atac[T.True]'])\n",
    "    ci_high.append(df.loc[df[\"coef label\"].astype(str) == \"is_atac[T.True]\", \"0.975]\"].item())\n",
    "    ci_low.append(df.loc[df[\"coef label\"].astype(str) == \"is_atac[T.True]\", \"[0.025\"].item())\n",
    "    interaction_ps.append(glm_res.pvalues[\"cond[T.Ser]:is_atac[T.True]\"])\n",
    "    interaction_bs.append(glm_res.params['cond[T.Ser]:is_atac[T.True]'])\n",
    "    interaction_ci_high.append(df.loc[df[\"coef label\"].astype(str) == \"cond[T.Ser]:is_atac[T.True]\", \"0.975]\"].item())\n",
    "    interaction_ci_low.append(df.loc[df[\"coef label\"].astype(str) == \"cond[T.Ser]:is_atac[T.True]\", \"[0.025\"].item())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "4e982d88",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_res_df = pd.concat(enrichment_results)\n",
    "significance_df = pd.DataFrame({\n",
    "    \"var_id\":var_ids, \n",
    "     \"p|atac\":ps, \n",
    "     \"b|atac\":bs, \n",
    "     \"ci_high|atac\":ci_high,\n",
    "     \"ci_low|atac\":ci_low,\n",
    "     \"p|ser*atac\":interaction_ps, \n",
    "     \"b|ser*atac\":interaction_bs,\n",
    "     \"ci_high|ser*atac\":interaction_ci_high,\n",
    "     \"ci_low|ser*atac\":interaction_ci_low,\n",
    "})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "4364c79a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manually fill in the data for variants that needed phasing\n",
    "tbl = pd.read_csv(\"rs35081008_Min_ABE_465_result.csv\", index_col=0\n",
    "                 )\n",
    "significance_df.loc[significance_df.var_id == \"rs35081008_Min_ABE_465\", \n",
    "                    ['p|atac', \n",
    "                     'b|atac', \n",
    "                     'ci_high|atac', \n",
    "                     'ci_low|atac', \n",
    "                     'p|ser*atac', \n",
    "                     'b|ser*atac', \n",
    "                     'ci_high|ser*atac', \n",
    "                     'ci_low|ser*atac']\n",
    "                   ] = tbl.loc[\"is_atac[T.True]\", [\"P>|z|\", \"coef\", \"0.975]\", \"[0.025\"]].tolist() + tbl.loc[\"cond[T.Ser]:is_atac[T.True]\", [\"P>|z|\", \"coef\", \"0.975]\", \"[0.025\"]].tolist() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0c01fdb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "for covariate in ['atac', 'ser*atac']:\n",
    "    multicorrect_res = statsmodels.stats.multitest.multipletests(significance_df[f\"p|{covariate}\"].dropna(), method=\"fdr_bh\", alpha=0.1)\n",
    "    multicorrect_res = statsmodels.stats.multitest.multipletests(significance_df[f\"p|{covariate}\"].dropna(), method=\"fdr_bh\", alpha=0.1)\n",
    "    significance_df.loc[~significance_df[f\"p|{covariate}\"].isnull(), f\"p_adj|{covariate}\"] = multicorrect_res[1]\n",
    "    significance_df.loc[~significance_df[f\"p|{covariate}\"].isnull(), f\"sig(a=0.1)|{covariate}\"] = multicorrect_res[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0a4e524b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rs4149309_Maj_ABE_206\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -24.519\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       32.392\n",
      "Time:                        15:38:38   Pearson chi2:                     39.3\n",
      "No. Iterations:                    27                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept        -28.5661   1.82e+04     -0.002      0.999   -3.56e+04    3.55e+04\n",
      "is_trt[T.True]    22.5819   1.82e+04      0.001      0.999   -3.56e+04    3.56e+04\n",
      "==================================================================================\n",
      "***Editing not enriched in treatment sample***\n",
      "rs10107182_Min_ABE_351\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -41.702\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       36.575\n",
      "Time:                        15:38:38   Pearson chi2:                     37.9\n",
      "No. Iterations:                    11                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept         -8.5805      0.378    -22.700      0.000      -9.321      -7.840\n",
      "is_trt[T.True]     0.8891      0.402      2.211      0.027       0.101       1.677\n",
      "==================================================================================\n",
      "rs115421711_Maj_ABE_45\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -4489.5\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       8888.9\n",
      "Time:                        15:38:38   Pearson chi2:                 9.71e+03\n",
      "No. Iterations:                    13                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept         -9.5974      0.183    -52.565      0.000      -9.955      -9.240\n",
      "is_trt[T.True]     3.7304      0.183     20.365      0.000       3.371       4.089\n",
      "==================================================================================\n",
      "rs1434282_Min_ABE_412\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -18.845\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       21.640\n",
      "Time:                        15:38:38   Pearson chi2:                     14.9\n",
      "No. Iterations:                    27                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept        -28.5661   1.48e+04     -0.002      0.998   -2.91e+04     2.9e+04\n",
      "is_trt[T.True]    22.4175   1.48e+04      0.002      0.999    -2.9e+04    2.91e+04\n",
      "==================================================================================\n",
      "***Editing not enriched in treatment sample***\n",
      "rs4719853_Maj_ABE_214\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -30.769\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       16.160\n",
      "Time:                        15:38:38   Pearson chi2:                     15.4\n",
      "No. Iterations:                    13                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept         -9.5003      0.258    -36.793      0.000     -10.006      -8.994\n",
      "is_trt[T.True]    -0.1876      0.293     -0.640      0.522      -0.762       0.387\n",
      "==================================================================================\n",
      "***Editing not enriched in treatment sample***\n",
      "rs704_Min_ABE_550\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -4429.1\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       8721.2\n",
      "Time:                        15:38:38   Pearson chi2:                 7.94e+03\n",
      "No. Iterations:                    12                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept         -8.6497      0.078   -110.760      0.000      -8.803      -8.497\n",
      "is_trt[T.True]     4.0321      0.078     51.495      0.000       3.879       4.186\n",
      "==================================================================================\n",
      "rs11149612_Min_ABE_369\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -84590.\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                   1.6902e+05\n",
      "Time:                        15:38:38   Pearson chi2:                 1.59e+05\n",
      "No. Iterations:                    12                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept         -8.6396      0.072   -120.014      0.000      -8.781      -8.498\n",
      "is_trt[T.True]     6.2927      0.072     87.389      0.000       6.152       6.434\n",
      "==================================================================================\n",
      "rs76895963_Maj_ABE_306\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -4.3322\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       6.6665\n",
      "Time:                        15:38:38   Pearson chi2:                     27.0\n",
      "No. Iterations:                    27                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept        -28.5661   1.51e+04     -0.002      0.998   -2.96e+04    2.96e+04\n",
      "is_trt[T.True]    19.0395   1.51e+04      0.001      0.999   -2.96e+04    2.96e+04\n",
      "==================================================================================\n",
      "***Editing not enriched in treatment sample***\n",
      "rs12042481_Maj_ABE_54\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -13540.\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       26985.\n",
      "Time:                        15:38:38   Pearson chi2:                 2.82e+04\n",
      "No. Iterations:                    12                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept         -9.0474      0.289    -31.339      0.000      -9.613      -8.482\n",
      "is_trt[T.True]     5.8271      0.289     20.176      0.000       5.261       6.393\n",
      "==================================================================================\n",
      "rs58198139_Min_ABE_523\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -116.51\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       192.54\n",
      "Time:                        15:38:38   Pearson chi2:                     211.\n",
      "No. Iterations:                    11                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept         -8.7012      1.000     -8.701      0.000     -10.661      -6.741\n",
      "is_trt[T.True]     3.6802      1.004      3.666      0.000       1.713       5.648\n",
      "==================================================================================\n",
      "rs2618566_Min_ABE_453\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -3694.0\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       7264.0\n",
      "Time:                        15:38:38   Pearson chi2:                 7.17e+03\n",
      "No. Iterations:                    13                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept         -9.7679      0.209    -46.844      0.000     -10.177      -9.359\n",
      "is_trt[T.True]     5.5752      0.209     26.724      0.000       5.166       5.984\n",
      "==================================================================================\n",
      "rs402348_Maj_ABE_201\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -969.59\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       1867.9\n",
      "Time:                        15:38:38   Pearson chi2:                 1.85e+03\n",
      "No. Iterations:                    12                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept         -8.9726      0.316    -28.372      0.000      -9.592      -8.353\n",
      "is_trt[T.True]     4.8042      0.317     15.162      0.000       4.183       5.425\n",
      "==================================================================================\n",
      "rs8126001_Min_ABE_574\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -124.74\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       150.23\n",
      "Time:                        15:38:38   Pearson chi2:                     176.\n",
      "No. Iterations:                     8                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept         -4.7968      0.024   -197.430      0.000      -4.844      -4.749\n",
      "is_trt[T.True]     0.5544      0.028     20.006      0.000       0.500       0.609\n",
      "==================================================================================\n",
      "rs1250259_Min_ABE_392\n",
      "                 Generalized Linear Model Regression Results                  \n",
      "==============================================================================\n",
      "Dep. Variable:         ['alt', 'ref']   No. Observations:                   16\n",
      "Model:                            GLM   Df Residuals:                       14\n",
      "Model Family:                Binomial   Df Model:                            1\n",
      "Link Function:                  logit   Scale:                          1.0000\n",
      "Method:                          IRLS   Log-Likelihood:                -44.793\n",
      "Date:                Tue, 22 Aug 2023   Deviance:                       23.414\n",
      "Time:                        15:38:38   Pearson chi2:                     22.8\n",
      "No. Iterations:                    13                                         \n",
      "Covariance Type:            nonrobust                                         \n",
      "==================================================================================\n",
      "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
      "----------------------------------------------------------------------------------\n",
      "Intercept        -10.0888      0.177    -57.070      0.000     -10.435      -9.742\n",
      "is_trt[T.True]     0.2027      0.193      1.048      0.295      -0.176       0.582\n",
      "==================================================================================\n",
      "***Editing not enriched in treatment sample***\n"
     ]
    }
   ],
   "source": [
    "enrichment_results = []\n",
    "ps = []\n",
    "bs = []\n",
    "for var_id in var_ids:\n",
    "    if var_id in failed_variants: \n",
    "        ps.append(np.nan)\n",
    "        bs.append(np.nan)\n",
    "        interaction_ps.append(np.nan)\n",
    "        interaction_bs.append(np.nan)\n",
    "        continue\n",
    "    print(var_id)\n",
    "    pdf = resolved_counts.loc[resolved_counts.var_id == var_id].copy()\n",
    "    pdf['sample'] = pdf[\"rep\"].astype(str) + \"_\" + pdf[\"cond\"]\n",
    "    pdf[\"is_atac\"] = pdf.exp == \"ATAC_Seq\"\n",
    "    pdf[\"is_trt\"] = pdf.rep != \"C\"\n",
    "    #pdf = pdf.loc[pdf.rep != \"C\"]\n",
    "    y, X = dmatrices('alt + ref ~ is_trt', data = pdf, return_type='dataframe')\n",
    "    try:\n",
    "        glm_binom = sm.GLM(y, X, family=sm.families.Binomial())\n",
    "        glm_res = glm_binom.fit()\n",
    "        print(glm_res.summary())\n",
    "    except:\n",
    "        print(pdf)\n",
    "        ps.append(np.nan)\n",
    "        bs.append(np.nan)\n",
    "        interaction_ps.append(np.nan)\n",
    "        interaction_bs.append(np.nan)\n",
    "        continue\n",
    "    if glm_res.pvalues[\"is_trt[T.True]\"] >= 0.05: \n",
    "        print(f\"***Editing not enriched in treatment sample***\")\n",
    "    df = pd.DataFrame(glm_res.summary().tables[1])\n",
    "    df.columns=df.iloc[0,:]\n",
    "    df = df.iloc[1:,:]\n",
    "    df.insert(0,\"variant\", var_id)\n",
    "    df.columns=[\"variant\", \"coef label\", \"coef\", \"std err\", \"z\", \"P>|z|\", \"[0.025\", \"0.975]\"]\n",
    "    enrichment_results.append(df)\n",
    "    ps.append(glm_res.pvalues[\"is_trt[T.True]\"])\n",
    "    bs.append(glm_res.params['is_trt[T.True]'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "43b1328c",
   "metadata": {},
   "outputs": [],
   "source": [
    "enrichment_res_df = pd.concat(enrichment_results)\n",
    "ctrl_significance_df = pd.DataFrame({\n",
    "    \"var_id\":var_ids, \n",
    "     \"p|trt\":ps, \n",
    "     \"b|trt\":bs, })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "e0cf70fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "tbl = pd.read_csv(\"rs35081008_Min_ABE_465_ctrl_result.csv\")\n",
    "ctrl_significance_df.loc[ctrl_significance_df.var_id == \"rs35081008_Min_ABE_465\", [\"p|trt\", \"b|trt\"]] = tbl.loc[1, [\"P>|z|\", \"coef\"]].tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "413c474a",
   "metadata": {},
   "outputs": [],
   "source": [
    "for covariate in ['trt']:\n",
    "    multicorrect_res = statsmodels.stats.multitest.multipletests(ctrl_significance_df[f\"p|{covariate}\"].dropna(), method=\"fdr_bh\", alpha=0.1)\n",
    "    ctrl_significance_df.loc[~ctrl_significance_df[f\"p|{covariate}\"].isnull(), f\"p_adj|{covariate}\"] = multicorrect_res[1]\n",
    "    ctrl_significance_df.loc[~ctrl_significance_df[f\"p|{covariate}\"].isnull(), f\"sig(a=0.1)|{covariate}\"] = multicorrect_res[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6b44cd59",
   "metadata": {},
   "source": [
    "## For unphasable unresolvable variants, ignore editing and calculate caQTL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "db39e9da",
   "metadata": {},
   "outputs": [],
   "source": [
    "unphasable_variants = [\"rs3767844_Maj_ABE_192\", \"rs4390169_Maj_ABE_208\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab9b759f",
   "metadata": {},
   "source": [
    "rs116734477_Maj_ABE_51: C/C, not edited."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8509e40f",
   "metadata": {},
   "outputs": [],
   "source": [
    "heterozygous_variants = [var_id for var_id in var_ids if len(var_info_dict[var_id]['refs']) == 2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "328d1272",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "rs4149309_Maj_ABE_206\n",
      "                       Generalized Linear Model Regression Results                       \n",
      "=========================================================================================\n",
      "Dep. Variable:     ['amplicon_ref', 'HepG2_alt']   No. Observations:                   16\n",
      "Model:                                       GLM   Df Residuals:                        9\n",
      "Model Family:                           Binomial   Df Model:                            6\n",
      "Link Function:                             logit   Scale:                          1.0000\n",
      "Method:                                     IRLS   Log-Likelihood:                -68.196\n",
      "Date:                           Tue, 22 Aug 2023   Deviance:                       36.224\n",
      "Time:                                   15:38:38   Pearson chi2:                     36.6\n",
      "No. Iterations:                                5                                         \n",
      "Covariance Type:                       nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                       1.8059      0.064     28.401      0.000       1.681       1.931\n",
      "rep[T.2]                       -0.0289      0.077     -0.378      0.706      -0.179       0.121\n",
      "rep[T.3]                       -0.0355      0.075     -0.471      0.638      -0.183       0.112\n",
      "rep[T.C]                       -0.1795      0.074     -2.432      0.015      -0.324      -0.035\n",
      "cond[T.Ser]                     0.0488      0.068      0.719      0.472      -0.084       0.182\n",
      "is_atac[T.True]                -0.7991      0.076    -10.453      0.000      -0.949      -0.649\n",
      "cond[T.Ser]:is_atac[T.True]    -0.1891      0.110     -1.726      0.084      -0.404       0.026\n",
      "===============================================================================================\n",
      "***Significant differential enrichment of ALT allele in ATAC-seq compared to GDNA: dlogit=-0.7991282442722046***\n",
      "rs3767844_Maj_ABE_192\n",
      "                       Generalized Linear Model Regression Results                       \n",
      "=========================================================================================\n",
      "Dep. Variable:     ['amplicon_ref', 'HepG2_alt']   No. Observations:                   16\n",
      "Model:                                       GLM   Df Residuals:                        9\n",
      "Model Family:                           Binomial   Df Model:                            6\n",
      "Link Function:                             logit   Scale:                          1.0000\n",
      "Method:                                     IRLS   Log-Likelihood:                -44.199\n",
      "Date:                           Tue, 22 Aug 2023   Deviance:                       4.3072\n",
      "Time:                                   15:38:38   Pearson chi2:                     4.29\n",
      "No. Iterations:                                3                                         \n",
      "Covariance Type:                       nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                       0.0985      0.080      1.224      0.221      -0.059       0.256\n",
      "rep[T.2]                       -0.0799      0.100     -0.802      0.423      -0.275       0.115\n",
      "rep[T.3]                       -0.2376      0.103     -2.316      0.021      -0.439      -0.037\n",
      "rep[T.C]                       -0.0919      0.105     -0.877      0.380      -0.297       0.113\n",
      "cond[T.Ser]                    -0.0425      0.079     -0.539      0.590      -0.197       0.112\n",
      "is_atac[T.True]                -0.0565      0.168     -0.337      0.736      -0.385       0.272\n",
      "cond[T.Ser]:is_atac[T.True]     0.1889      0.218      0.865      0.387      -0.239       0.617\n",
      "===============================================================================================\n",
      "rs62084210_Maj_ABE_250\n",
      "                       Generalized Linear Model Regression Results                       \n",
      "=========================================================================================\n",
      "Dep. Variable:     ['amplicon_ref', 'HepG2_alt']   No. Observations:                   16\n",
      "Model:                                       GLM   Df Residuals:                        9\n",
      "Model Family:                           Binomial   Df Model:                            6\n",
      "Link Function:                             logit   Scale:                          1.0000\n",
      "Method:                                     IRLS   Log-Likelihood:                -82.564\n",
      "Date:                           Tue, 22 Aug 2023   Deviance:                       37.730\n",
      "Time:                                   15:38:38   Pearson chi2:                     37.3\n",
      "No. Iterations:                                4                                         \n",
      "Covariance Type:                       nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                       0.6649      0.024     27.153      0.000       0.617       0.713\n",
      "rep[T.2]                        0.0036      0.030      0.119      0.905      -0.056       0.063\n",
      "rep[T.3]                        0.0427      0.029      1.456      0.145      -0.015       0.100\n",
      "rep[T.C]                        0.0035      0.029      0.120      0.904      -0.053       0.060\n",
      "cond[T.Ser]                    -0.0062      0.023     -0.267      0.790      -0.052       0.040\n",
      "is_atac[T.True]                 0.0885      0.037      2.372      0.018       0.015       0.162\n",
      "cond[T.Ser]:is_atac[T.True]    -0.0854      0.049     -1.732      0.083      -0.182       0.011\n",
      "===============================================================================================\n",
      "***Significant differential enrichment of ALT allele in ATAC-seq compared to GDNA: dlogit=0.08848593473640048***\n",
      "rs2618566_Min_ABE_453\n",
      "                       Generalized Linear Model Regression Results                       \n",
      "=========================================================================================\n",
      "Dep. Variable:     ['amplicon_ref', 'HepG2_alt']   No. Observations:                   16\n",
      "Model:                                       GLM   Df Residuals:                        9\n",
      "Model Family:                           Binomial   Df Model:                            6\n",
      "Link Function:                             logit   Scale:                          1.0000\n",
      "Method:                                     IRLS   Log-Likelihood:                -620.80\n",
      "Date:                           Tue, 22 Aug 2023   Deviance:                       1048.4\n",
      "Time:                                   15:38:38   Pearson chi2:                 1.05e+03\n",
      "No. Iterations:                                3                                         \n",
      "Covariance Type:                       nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                       0.0712      0.004     16.136      0.000       0.063       0.080\n",
      "rep[T.2]                        0.0450      0.004     11.612      0.000       0.037       0.053\n",
      "rep[T.3]                       -0.1234      0.004    -30.646      0.000      -0.131      -0.116\n",
      "rep[T.C]                       -0.0479      0.004    -11.178      0.000      -0.056      -0.040\n",
      "cond[T.Ser]                     0.0021      0.005      0.405      0.685      -0.008       0.012\n",
      "is_atac[T.True]                 0.1289      0.004     29.703      0.000       0.120       0.137\n",
      "cond[T.Ser]:is_atac[T.True]     0.0391      0.006      6.355      0.000       0.027       0.051\n",
      "===============================================================================================\n",
      "***Significant differential enrichment of ALT allele in ATAC-seq compared to GDNA: dlogit=0.1288671134026831***\n",
      "rs35081008_Min_ABE_465\n",
      "                       Generalized Linear Model Regression Results                       \n",
      "=========================================================================================\n",
      "Dep. Variable:     ['amplicon_ref', 'HepG2_alt']   No. Observations:                   16\n",
      "Model:                                       GLM   Df Residuals:                        9\n",
      "Model Family:                           Binomial   Df Model:                            6\n",
      "Link Function:                             logit   Scale:                          1.0000\n",
      "Method:                                     IRLS   Log-Likelihood:                -6685.4\n",
      "Date:                           Tue, 22 Aug 2023   Deviance:                       13191.\n",
      "Time:                                   15:38:38   Pearson chi2:                 1.57e+04\n",
      "No. Iterations:                                7                                         \n",
      "Covariance Type:                       nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                       2.5965      0.004    694.390      0.000       2.589       2.604\n",
      "rep[T.2]                       -0.0076      0.005     -1.612      0.107      -0.017       0.002\n",
      "rep[T.3]                        0.3734      0.005     71.115      0.000       0.363       0.384\n",
      "rep[T.C]                       -0.1834      0.005    -38.848      0.000      -0.193      -0.174\n",
      "cond[T.Ser]                     0.0987      0.004     26.657      0.000       0.091       0.106\n",
      "is_atac[T.True]                 1.5177      0.008    180.284      0.000       1.501       1.534\n",
      "cond[T.Ser]:is_atac[T.True]    -0.1165      0.011    -10.241      0.000      -0.139      -0.094\n",
      "===============================================================================================\n",
      "***Significant differential enrichment of ALT allele in ATAC-seq compared to GDNA: dlogit=1.5176709584295336***\n",
      "rs4390169_Maj_ABE_208\n",
      "                       Generalized Linear Model Regression Results                       \n",
      "=========================================================================================\n",
      "Dep. Variable:     ['amplicon_ref', 'HepG2_alt']   No. Observations:                   16\n",
      "Model:                                       GLM   Df Residuals:                        9\n",
      "Model Family:                           Binomial   Df Model:                            6\n",
      "Link Function:                             logit   Scale:                          1.0000\n",
      "Method:                                     IRLS   Log-Likelihood:                -694.29\n",
      "Date:                           Tue, 22 Aug 2023   Deviance:                       1186.4\n",
      "Time:                                   15:38:38   Pearson chi2:                 1.19e+03\n",
      "No. Iterations:                                3                                         \n",
      "Covariance Type:                       nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -0.0171      0.003     -5.477      0.000      -0.023      -0.011\n",
      "rep[T.2]                       -0.0436      0.003    -14.496      0.000      -0.049      -0.038\n",
      "rep[T.3]                        0.0246      0.003      8.159      0.000       0.019       0.031\n",
      "rep[T.C]                       -0.0017      0.003     -0.532      0.595      -0.008       0.005\n",
      "cond[T.Ser]                     0.0184      0.004      4.839      0.000       0.011       0.026\n",
      "is_atac[T.True]                 0.0057      0.003      1.822      0.068      -0.000       0.012\n",
      "cond[T.Ser]:is_atac[T.True]     0.0429      0.005      9.273      0.000       0.034       0.052\n",
      "===============================================================================================\n",
      "rs402348_Maj_ABE_201\n",
      "                       Generalized Linear Model Regression Results                       \n",
      "=========================================================================================\n",
      "Dep. Variable:     ['amplicon_ref', 'HepG2_alt']   No. Observations:                   16\n",
      "Model:                                       GLM   Df Residuals:                        9\n",
      "Model Family:                           Binomial   Df Model:                            6\n",
      "Link Function:                             logit   Scale:                          1.0000\n",
      "Method:                                     IRLS   Log-Likelihood:                -98.274\n",
      "Date:                           Tue, 22 Aug 2023   Deviance:                       57.913\n",
      "Time:                                   15:38:38   Pearson chi2:                     57.9\n",
      "No. Iterations:                                3                                         \n",
      "Covariance Type:                       nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                       0.0259      0.009      2.947      0.003       0.009       0.043\n",
      "rep[T.2]                       -0.0698      0.012     -5.995      0.000      -0.093      -0.047\n",
      "rep[T.3]                        0.0334      0.012      2.796      0.005       0.010       0.057\n",
      "rep[T.C]                        0.0557      0.011      5.260      0.000       0.035       0.076\n",
      "cond[T.Ser]                     0.0072      0.008      0.877      0.381      -0.009       0.023\n",
      "is_atac[T.True]                 0.0111      0.042      0.262      0.793      -0.072       0.094\n",
      "cond[T.Ser]:is_atac[T.True]    -0.0214      0.064     -0.334      0.738      -0.147       0.104\n",
      "===============================================================================================\n",
      "rs1250259_Min_ABE_392\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                       Generalized Linear Model Regression Results                       \n",
      "=========================================================================================\n",
      "Dep. Variable:     ['amplicon_ref', 'HepG2_alt']   No. Observations:                   16\n",
      "Model:                                       GLM   Df Residuals:                        9\n",
      "Model Family:                           Binomial   Df Model:                            6\n",
      "Link Function:                             logit   Scale:                          1.0000\n",
      "Method:                                     IRLS   Log-Likelihood:                -729.34\n",
      "Date:                           Tue, 22 Aug 2023   Deviance:                       1255.9\n",
      "Time:                                   15:38:38   Pearson chi2:                 1.26e+03\n",
      "No. Iterations:                                4                                         \n",
      "Covariance Type:                       nonrobust                                         \n",
      "===============================================================================================\n",
      "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
      "-----------------------------------------------------------------------------------------------\n",
      "Intercept                      -0.6785      0.003   -229.513      0.000      -0.684      -0.673\n",
      "rep[T.2]                        0.0385      0.003     13.017      0.000       0.033       0.044\n",
      "rep[T.3]                       -0.0401      0.003    -12.924      0.000      -0.046      -0.034\n",
      "rep[T.C]                        0.0015      0.003      0.436      0.663      -0.005       0.008\n",
      "cond[T.Ser]                     0.0017      0.003      0.510      0.610      -0.005       0.008\n",
      "is_atac[T.True]                -0.0016      0.003     -0.547      0.584      -0.008       0.004\n",
      "cond[T.Ser]:is_atac[T.True]    -0.0555      0.004    -12.560      0.000      -0.064      -0.047\n",
      "===============================================================================================\n"
     ]
    }
   ],
   "source": [
    "unphasable_counts_list = []\n",
    "bs = []\n",
    "ps = []\n",
    "pdfs = {}\n",
    "ci_high = []\n",
    "ci_low = []\n",
    "interaction_ps = []\n",
    "interaction_bs = []\n",
    "interaction_ci_high = []\n",
    "interaction_ci_low = []\n",
    "enrichment_results = []\n",
    "for var_id in heterozygous_variants:\n",
    "    print(var_id)\n",
    "    ref_base = var_info_dict[var_id]['amplicon_base']\n",
    "    alt_base = [b for b in var_info_dict[var_id]['refs'] if b != ref_base][0]\n",
    "    var_df = counts_df.loc[(counts_df.var_id == var_id) & (counts_df.base.isin([ref_base, alt_base]))]\n",
    "    pdf = pd.pivot_table(var_df, index=[\"var_id\", \"exp\", \"cond\", 'rep'], columns=[\"base\",], values='count').astype(float).reset_index().rename(columns={ref_base:'amplicon_ref', alt_base:'HepG2_alt'})\n",
    "    unphasable_counts_list.append(pdf)\n",
    "    pdf['sample'] = pdf[\"rep\"].astype(str) + \"_\" + pdf[\"cond\"]\n",
    "    pdf[\"is_atac\"] = pdf.exp == \"ATAC_Seq\"\n",
    "    pdf[\"is_trt\"] = pdf.rep != \"C\"\n",
    "    pdfs[var_id] = pdf\n",
    "    y, X = dmatrices('amplicon_ref + HepG2_alt ~ rep + cond + is_atac + cond*is_atac', data = pdf, return_type='dataframe')\n",
    "    try:\n",
    "        glm_binom = sm.GLM(y, X, family=sm.families.Binomial())\n",
    "        glm_res = glm_binom.fit()\n",
    "        print(glm_res.summary())\n",
    "    except:\n",
    "        print(pdf)\n",
    "        ps.append(np.nan)\n",
    "        bs.append(np.nan)\n",
    "        ci_high.append(np.nan)\n",
    "        ci_low.append(np.nan)\n",
    "        interaction_ps.append(np.nan)\n",
    "        interaction_bs.append(np.nan)\n",
    "        interaction_ci_high.append(np.nan)\n",
    "        interaction_ci_low.append(np.nan)\n",
    "        continue\n",
    "    if glm_res.pvalues[\"is_atac[T.True]\"] < 0.05: \n",
    "        print(f\"***Significant differential enrichment of ALT allele in ATAC-seq compared to GDNA: dlogit={glm_res.params['is_atac[T.True]']}***\")\n",
    "    df = pd.DataFrame(glm_res.summary().tables[1])\n",
    "    df.columns=df.iloc[0,:]\n",
    "    df = df.iloc[1:,:]\n",
    "    df.insert(0,\"variant\", var_id)\n",
    "    df.columns=[\"variant\", \"coef label\", \"coef\", \"std err\", \"z\", \"P>|z|\", \"[0.025\", \"0.975]\"]\n",
    "    enrichment_results.append(df)\n",
    "    ps.append(glm_res.pvalues[\"is_atac[T.True]\"])\n",
    "    bs.append(glm_res.params['is_atac[T.True]'])\n",
    "    ci_high.append(df.loc[df[\"coef label\"].astype(str) == \"is_atac[T.True]\", \"0.975]\"].item())\n",
    "    ci_low.append(df.loc[df[\"coef label\"].astype(str) == \"is_atac[T.True]\", \"[0.025\"].item())\n",
    "    interaction_ps.append(glm_res.pvalues[\"cond[T.Ser]:is_atac[T.True]\"])\n",
    "    interaction_bs.append(glm_res.params['cond[T.Ser]:is_atac[T.True]'])\n",
    "    interaction_ci_high.append(df.loc[df[\"coef label\"].astype(str) == \"cond[T.Ser]:is_atac[T.True]\", \"0.975]\"].item())\n",
    "    interaction_ci_low.append(df.loc[df[\"coef label\"].astype(str) == \"cond[T.Ser]:is_atac[T.True]\", \"[0.025\"].item())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "12112033",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>base</th>\n",
       "      <th>count</th>\n",
       "      <th>amplicon_ref</th>\n",
       "      <th>var_id</th>\n",
       "      <th>rep</th>\n",
       "      <th>cond</th>\n",
       "      <th>exp</th>\n",
       "      <th>valid_mt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>305.0</td>\n",
       "      <td>A</td>\n",
       "      <td>rs4149309_Maj_ABE_206</td>\n",
       "      <td>1</td>\n",
       "      <td>Ser</td>\n",
       "      <td>ATAC_Seq</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>0.0</td>\n",
       "      <td>A</td>\n",
       "      <td>rs4149309_Maj_ABE_206</td>\n",
       "      <td>1</td>\n",
       "      <td>Ser</td>\n",
       "      <td>ATAC_Seq</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G</td>\n",
       "      <td>1.0</td>\n",
       "      <td>A</td>\n",
       "      <td>rs4149309_Maj_ABE_206</td>\n",
       "      <td>1</td>\n",
       "      <td>Ser</td>\n",
       "      <td>ATAC_Seq</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>159.0</td>\n",
       "      <td>A</td>\n",
       "      <td>rs4149309_Maj_ABE_206</td>\n",
       "      <td>1</td>\n",
       "      <td>Ser</td>\n",
       "      <td>ATAC_Seq</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>N</td>\n",
       "      <td>7.0</td>\n",
       "      <td>A</td>\n",
       "      <td>rs4149309_Maj_ABE_206</td>\n",
       "      <td>1</td>\n",
       "      <td>Ser</td>\n",
       "      <td>ATAC_Seq</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>C</td>\n",
       "      <td>8.0</td>\n",
       "      <td>T</td>\n",
       "      <td>rs1250259_Min_ABE_392</td>\n",
       "      <td>C</td>\n",
       "      <td>SS</td>\n",
       "      <td>GDNA</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>G</td>\n",
       "      <td>26.0</td>\n",
       "      <td>T</td>\n",
       "      <td>rs1250259_Min_ABE_392</td>\n",
       "      <td>C</td>\n",
       "      <td>SS</td>\n",
       "      <td>GDNA</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>T</td>\n",
       "      <td>76219.0</td>\n",
       "      <td>T</td>\n",
       "      <td>rs1250259_Min_ABE_392</td>\n",
       "      <td>C</td>\n",
       "      <td>SS</td>\n",
       "      <td>GDNA</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>N</td>\n",
       "      <td>5986.0</td>\n",
       "      <td>T</td>\n",
       "      <td>rs1250259_Min_ABE_392</td>\n",
       "      <td>C</td>\n",
       "      <td>SS</td>\n",
       "      <td>GDNA</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>-</td>\n",
       "      <td>2.0</td>\n",
       "      <td>T</td>\n",
       "      <td>rs1250259_Min_ABE_392</td>\n",
       "      <td>C</td>\n",
       "      <td>SS</td>\n",
       "      <td>GDNA</td>\n",
       "      <td>-1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1890 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   base    count amplicon_ref                 var_id rep cond       exp  \\\n",
       "0     A    305.0            A  rs4149309_Maj_ABE_206   1  Ser  ATAC_Seq   \n",
       "1     C      0.0            A  rs4149309_Maj_ABE_206   1  Ser  ATAC_Seq   \n",
       "2     G      1.0            A  rs4149309_Maj_ABE_206   1  Ser  ATAC_Seq   \n",
       "3     T    159.0            A  rs4149309_Maj_ABE_206   1  Ser  ATAC_Seq   \n",
       "4     N      7.0            A  rs4149309_Maj_ABE_206   1  Ser  ATAC_Seq   \n",
       "..  ...      ...          ...                    ...  ..  ...       ...   \n",
       "1     C      8.0            T  rs1250259_Min_ABE_392   C   SS      GDNA   \n",
       "2     G     26.0            T  rs1250259_Min_ABE_392   C   SS      GDNA   \n",
       "3     T  76219.0            T  rs1250259_Min_ABE_392   C   SS      GDNA   \n",
       "4     N   5986.0            T  rs1250259_Min_ABE_392   C   SS      GDNA   \n",
       "5     -      2.0            T  rs1250259_Min_ABE_392   C   SS      GDNA   \n",
       "\n",
       "    valid_mt  \n",
       "0        0.0  \n",
       "1        0.5  \n",
       "2        1.0  \n",
       "3        0.0  \n",
       "4       -1.0  \n",
       "..       ...  \n",
       "1        1.0  \n",
       "2        0.5  \n",
       "3        0.0  \n",
       "4       -1.0  \n",
       "5       -1.0  \n",
       "\n",
       "[1890 rows x 8 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "4897462f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pdfs[\"rs4149309_Maj_ABE_206\"]\n",
    "df['lfc'] = df['HepG2_alt'] / df['amplicon_ref']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "3cc77519",
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_enrichment_res_df = pd.concat(enrichment_results)\n",
    "caqtl_significance_df = pd.DataFrame(\n",
    "    {\"var_id\":heterozygous_variants, \n",
    "     \"p|atac\":ps, \n",
    "     \"b|atac\":bs, \n",
    "     \"ci_high|atac\":ci_high,\n",
    "     \"ci_low|atac\":ci_low,\n",
    "     \"p|ser*atac\":interaction_ps, \n",
    "     \"b|ser*atac\":interaction_bs,\n",
    "     \"ci_high|ser*atac\":interaction_ci_high,\n",
    "     \"ci_low|ser*atac\":interaction_ci_low,\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "4d5f6209",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manually fill in the data for variants that needed phasing\n",
    "tbl = pd.read_csv(\"rs35081008_Min_ABE_465_caqtl_result.csv\", index_col=0\n",
    "                 )\n",
    "caqtl_significance_df.loc[caqtl_significance_df.var_id == \"rs35081008_Min_ABE_465\", \n",
    "                    ['p|atac', \n",
    "                     'b|atac', \n",
    "                     'ci_high|atac', \n",
    "                     'ci_low|atac', \n",
    "                     'p|ser*atac', \n",
    "                     'b|ser*atac', \n",
    "                     'ci_high|ser*atac', \n",
    "                     'ci_low|ser*atac']\n",
    "                   ] = tbl.loc[\"is_atac[T.True]\", [\"P>|z|\", \"coef\", \"0.975]\", \"[0.025\"]].tolist() + tbl.loc[\"cond[T.Ser]:is_atac[T.True]\", [\"P>|z|\", \"coef\", \"0.975]\", \"[0.025\"]].tolist() \n",
    "\n",
    "tbl = pd.read_csv(\"rs62084210_Maj_ABE_250_caqtl_result.csv\", index_col=0\n",
    "                 )\n",
    "caqtl_significance_df.loc[caqtl_significance_df.var_id == \"rs62084210_Maj_ABE_250\", \n",
    "                    ['p|atac', \n",
    "                     'b|atac', \n",
    "                     'ci_high|atac', \n",
    "                     'ci_low|atac', \n",
    "                     'p|ser*atac', \n",
    "                     'b|ser*atac', \n",
    "                     'ci_high|ser*atac', \n",
    "                     'ci_low|ser*atac']\n",
    "                   ] = tbl.loc[\"is_atac[T.True]\", [\"P>|z|\", \"coef\", \"0.975]\", \"[0.025\"]].tolist() + tbl.loc[\"cond[T.Ser]:is_atac[T.True]\", [\"P>|z|\", \"coef\", \"0.975]\", \"[0.025\"]].tolist() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "850d9b13",
   "metadata": {},
   "outputs": [],
   "source": [
    "for covariate in ['atac', 'ser*atac']:\n",
    "    multicorrect_res = statsmodels.stats.multitest.multipletests(caqtl_significance_df[f\"p|{covariate}\"].dropna(), method=\"fdr_bh\", alpha=0.1)\n",
    "    multicorrect_res = statsmodels.stats.multitest.multipletests(caqtl_significance_df[f\"p|{covariate}\"].dropna(), method=\"fdr_bh\", alpha=0.1)\n",
    "    caqtl_significance_df.loc[~caqtl_significance_df[f\"p|{covariate}\"].isnull(), f\"p_adj|{covariate}\"] = multicorrect_res[1]\n",
    "    caqtl_significance_df.loc[~caqtl_significance_df[f\"p|{covariate}\"].isnull(), f\"sig(a=0.1)|{covariate}\"] = multicorrect_res[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "4d62fd7c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>var_id</th>\n",
       "      <th>p|atac</th>\n",
       "      <th>b|atac</th>\n",
       "      <th>ci_high|atac</th>\n",
       "      <th>ci_low|atac</th>\n",
       "      <th>p|ser*atac</th>\n",
       "      <th>b|ser*atac</th>\n",
       "      <th>ci_high|ser*atac</th>\n",
       "      <th>ci_low|ser*atac</th>\n",
       "      <th>p_adj|atac</th>\n",
       "      <th>sig(a=0.1)|atac</th>\n",
       "      <th>p_adj|ser*atac</th>\n",
       "      <th>sig(a=0.1)|ser*atac</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs4149309_Maj_ABE_206</td>\n",
       "      <td>1.421221e-25</td>\n",
       "      <td>-0.799128</td>\n",
       "      <td>-0.649</td>\n",
       "      <td>-0.949</td>\n",
       "      <td>8.429579e-02</td>\n",
       "      <td>-0.189066</td>\n",
       "      <td>0.026</td>\n",
       "      <td>-0.404</td>\n",
       "      <td>3.789923e-25</td>\n",
       "      <td>True</td>\n",
       "      <td>1.348733e-01</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs3767844_Maj_ABE_192</td>\n",
       "      <td>7.358820e-01</td>\n",
       "      <td>-0.056502</td>\n",
       "      <td>0.272</td>\n",
       "      <td>-0.385</td>\n",
       "      <td>3.872154e-01</td>\n",
       "      <td>0.188869</td>\n",
       "      <td>0.617</td>\n",
       "      <td>-0.239</td>\n",
       "      <td>7.934388e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>5.162872e-01</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs62084210_Maj_ABE_250</td>\n",
       "      <td>4.570000e-01</td>\n",
       "      <td>-0.062300</td>\n",
       "      <td>0.102</td>\n",
       "      <td>-0.226</td>\n",
       "      <td>7.660000e-01</td>\n",
       "      <td>0.033000</td>\n",
       "      <td>0.25</td>\n",
       "      <td>-0.184</td>\n",
       "      <td>7.312000e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>7.660000e-01</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs2618566_Min_ABE_453</td>\n",
       "      <td>7.111619e-194</td>\n",
       "      <td>0.128867</td>\n",
       "      <td>0.137</td>\n",
       "      <td>0.120</td>\n",
       "      <td>2.081110e-10</td>\n",
       "      <td>0.039112</td>\n",
       "      <td>0.051</td>\n",
       "      <td>0.027</td>\n",
       "      <td>2.844647e-193</td>\n",
       "      <td>True</td>\n",
       "      <td>4.162221e-10</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs35081008_Min_ABE_465</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>1.555300</td>\n",
       "      <td>1.58</td>\n",
       "      <td>1.531</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>-0.149300</td>\n",
       "      <td>-0.117</td>\n",
       "      <td>-0.182</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rs4390169_Maj_ABE_208</td>\n",
       "      <td>6.847839e-02</td>\n",
       "      <td>0.005717</td>\n",
       "      <td>0.012</td>\n",
       "      <td>-0.000</td>\n",
       "      <td>1.803900e-20</td>\n",
       "      <td>0.042893</td>\n",
       "      <td>0.052</td>\n",
       "      <td>0.034</td>\n",
       "      <td>1.369568e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>4.810400e-20</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>rs402348_Maj_ABE_201</td>\n",
       "      <td>7.934388e-01</td>\n",
       "      <td>0.011121</td>\n",
       "      <td>0.094</td>\n",
       "      <td>-0.072</td>\n",
       "      <td>7.382978e-01</td>\n",
       "      <td>-0.021374</td>\n",
       "      <td>0.104</td>\n",
       "      <td>-0.147</td>\n",
       "      <td>7.934388e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>7.660000e-01</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>rs1250259_Min_ABE_392</td>\n",
       "      <td>5.843289e-01</td>\n",
       "      <td>-0.001645</td>\n",
       "      <td>0.004</td>\n",
       "      <td>-0.008</td>\n",
       "      <td>3.505914e-36</td>\n",
       "      <td>-0.055544</td>\n",
       "      <td>-0.047</td>\n",
       "      <td>-0.064</td>\n",
       "      <td>7.791052e-01</td>\n",
       "      <td>False</td>\n",
       "      <td>1.402366e-35</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   var_id         p|atac    b|atac ci_high|atac ci_low|atac  \\\n",
       "0   rs4149309_Maj_ABE_206   1.421221e-25 -0.799128       -0.649      -0.949   \n",
       "1   rs3767844_Maj_ABE_192   7.358820e-01 -0.056502        0.272      -0.385   \n",
       "2  rs62084210_Maj_ABE_250   4.570000e-01 -0.062300        0.102      -0.226   \n",
       "3   rs2618566_Min_ABE_453  7.111619e-194  0.128867        0.137       0.120   \n",
       "4  rs35081008_Min_ABE_465   0.000000e+00  1.555300         1.58       1.531   \n",
       "5   rs4390169_Maj_ABE_208   6.847839e-02  0.005717        0.012      -0.000   \n",
       "6    rs402348_Maj_ABE_201   7.934388e-01  0.011121        0.094      -0.072   \n",
       "7   rs1250259_Min_ABE_392   5.843289e-01 -0.001645        0.004      -0.008   \n",
       "\n",
       "     p|ser*atac  b|ser*atac ci_high|ser*atac ci_low|ser*atac     p_adj|atac  \\\n",
       "0  8.429579e-02   -0.189066            0.026          -0.404   3.789923e-25   \n",
       "1  3.872154e-01    0.188869            0.617          -0.239   7.934388e-01   \n",
       "2  7.660000e-01    0.033000             0.25          -0.184   7.312000e-01   \n",
       "3  2.081110e-10    0.039112            0.051           0.027  2.844647e-193   \n",
       "4  0.000000e+00   -0.149300           -0.117          -0.182   0.000000e+00   \n",
       "5  1.803900e-20    0.042893            0.052           0.034   1.369568e-01   \n",
       "6  7.382978e-01   -0.021374            0.104          -0.147   7.934388e-01   \n",
       "7  3.505914e-36   -0.055544           -0.047          -0.064   7.791052e-01   \n",
       "\n",
       "  sig(a=0.1)|atac  p_adj|ser*atac sig(a=0.1)|ser*atac  \n",
       "0            True    1.348733e-01               False  \n",
       "1           False    5.162872e-01               False  \n",
       "2           False    7.660000e-01               False  \n",
       "3            True    4.162221e-10                True  \n",
       "4            True    0.000000e+00                True  \n",
       "5           False    4.810400e-20                True  \n",
       "6           False    7.660000e-01               False  \n",
       "7           False    1.402366e-35                True  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "caqtl_significance_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "568f0221",
   "metadata": {},
   "outputs": [],
   "source": [
    "unresolved_counts = pd.concat(unphasable_counts_list).iloc[:,:-2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "96827483",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_counts = pd.concat([resolved_counts,unresolved_counts])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "d1591d21",
   "metadata": {},
   "outputs": [],
   "source": [
    "phased_variants = [\"rs35081008_Min_ABE_465\", \"rs62084210_Maj_ABE_250\"]\n",
    "nonedited_variants = ['rs116734477_Maj_ABE_51']\n",
    "all_counts.loc[all_counts.var_id.isin(phased_variants), 'status'] = 'Ambiguous edited vs. minor reference, phased'\n",
    "all_counts.loc[all_counts.var_id.isin(nonedited_variants), 'status'] = 'Not edited (HepG2 does not have A or T ref)'\n",
    "all_counts.loc[all_counts.var_id.isin(unphasable_variants),'status'] = 'Ambiguous edited vs. minor reference, unphasable'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "f0dad9de",
   "metadata": {},
   "outputs": [],
   "source": [
    "with pd.ExcelWriter('20230822_ATAC_demuxed_analysis_result.xlsx') as writer:  \n",
    "    all_counts.to_excel(writer, sheet_name=\"counts\")\n",
    "    edit_assignment_all.to_excel(writer, sheet_name=\"edit assignment\")\n",
    "    enrichment_res_df.to_excel(writer, sheet_name=\"edit enrichment\")\n",
    "    significance_df.to_excel(writer, sheet_name=\"edit enrichment summary\")\n",
    "    allele_enrichment_res_df.to_excel(writer, sheet_name=\"allele enrichment\")\n",
    "    caqtl_significance_df.to_excel(writer, sheet_name=\"allele enrichment summary\")\n",
    "    ctrl_significance_df.to_excel(writer, sheet_name='ctrl')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jy_anbe_py38",
   "language": "python",
   "name": "jy_anbe_py38"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
