{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2e4b7a3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/pinello/SHARED_SOFTWARE/anaconda_latest/envs/jy_anbe_py38/lib/python3.8/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.\n",
      "  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,\n",
      "/data/pinello/SHARED_SOFTWARE/anaconda_latest/envs/jy_anbe_py38/lib/python3.8/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Float64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.\n",
      "  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from tqdm.auto import tqdm\n",
    "from patsy import dmatrices\n",
    "import statsmodels.api as sm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ed8f124b",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bean.mapping._supporting_fn import _get_allele_from_alignment\n",
    "from bean import Allele,Edit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "0ec2b57e",
   "metadata": {},
   "outputs": [],
   "source": [
    "edit_dict = {\"A\":\"G\", \"T\":\"C\"}\n",
    "revcomp_dict = {\"A\":\"T\", \"T\":\"A\", \"C\":\"G\", \"G\":\"C\"}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "14bc4597",
   "metadata": {},
   "source": [
    "## Load variant info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "691b1edf",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_id = \"rs35081008_Min_ABE_465\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0fb57d54",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_info_tbl = pd.read_excel(\"../../resources/LDLvar/20221013_LDLvar_simpleZscores_credset.xlsx\", index_col=0, header=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f3f1dc05",
   "metadata": {},
   "outputs": [],
   "source": [
    "amplicon_info = pd.read_csv(\"../../resources/atac_seq/030123_ATACseq_info.csv\", ).set_index(\"gRNA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9afc374d",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_alleles = var_info_tbl.loc[var_id, [\"A1\", \"A2\"]].reset_index()\n",
    "var_alleles['edited'] = var_alleles[var_id].map(lambda b: edit_dict[b] if b in edit_dict.keys() else np.nan)\n",
    "var_alleles = var_alleles.loc[~var_alleles.edited.isnull()]\n",
    "var_alleles.columns=[\"allele\", \"ref\", \"edited\"]\n",
    "var_chrom = int(var_info_tbl.loc[var_id, \"CHR\"])\n",
    "var_pos = int(var_info_tbl.loc[var_id, 'position_hg19'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b305fbdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "amplicon_start = amplicon_info.loc[var_id, \"Varpos_fw\"]-4\n",
    "amplicon_seq = amplicon_info.loc[var_id, \"Amplicon_fw (reference)\"].replace(\"N\", \"\")\n",
    "amplicon_len = len(amplicon_seq)\n",
    "amplicon_end = amplicon_start + amplicon_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bbdd9378",
   "metadata": {},
   "outputs": [],
   "source": [
    "phased_snps = pd.read_table(f\"../../results/atac_seq/{var_id}_phasedSNPs.vcf\", comment=\"#\", header=None)\n",
    "phased_snps.columns=[\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"QUAL\", \"FILTER\", \"INFO\", \"FORMAT\", \"HepG2\"]\n",
    "phased_snps[\"GT\"] = phased_snps.HepG2.map(lambda s: s.split(\":\")[0])\n",
    "phased_snps = phased_snps.loc[phased_snps.GT.isin([\"1|0\", \"0|1\"]) & (phased_snps.FILTER == \"PASS\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b194c277",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_allele(row):\n",
    "    ids = row.GT.split(\"|\")\n",
    "    alleles = [row[{\"0\":\"REF\", \"1\":\"ALT\"}[aid]] for aid in ids]\n",
    "    return alleles\n",
    "\n",
    "def _get_allele_from_alignment(\n",
    "    ref_aligned: str,\n",
    "    query_aligned: str,\n",
    "    offset: int,\n",
    "    strand: int,\n",
    "    start_pos: int,\n",
    "    end_pos: int,\n",
    "    positionwise_quality: np.ndarray = None,\n",
    "    quality_thres: float = -1,\n",
    "):\n",
    "    # Include N, no quality filter\n",
    "    assert len(ref_aligned) == len(query_aligned)\n",
    "    allele = Allele()\n",
    "    ref_gaps = 0\n",
    "    alt_gaps = 0\n",
    "    alt_seq_len = len(query_aligned) - query_aligned.count(\"-\")\n",
    "    if positionwise_quality is None:\n",
    "        # alt_position_is_good_quality = np.ones(alt_seq_len, dtype=bool)\n",
    "        alt_position_is_good_quality = np.array(\n",
    "            [c != \"N\" for c in query_aligned.replace(\"-\", \"\")]\n",
    "        )\n",
    "    else:\n",
    "        assert len(positionwise_quality) == alt_seq_len\n",
    "        alt_position_is_good_quality = positionwise_quality > quality_thres\n",
    "    for i in range(len(ref_aligned)):\n",
    "        if ref_aligned[i] == query_aligned[i]:\n",
    "            continue\n",
    "        ref_base = ref_aligned[i]\n",
    "        alt_base = query_aligned[i]\n",
    "        if alt_base != \"-\":\n",
    "            alt_base_is_good_quality = alt_position_is_good_quality[i - alt_gaps]\n",
    "        else:\n",
    "            alt_base_is_good_quality = True\n",
    "        if ref_base == \"-\":\n",
    "            ref_gaps += 1\n",
    "        elif alt_base == \"-\":\n",
    "            alt_gaps += 1\n",
    "        ref_pos = i - ref_gaps\n",
    "        allele.add(\n",
    "            Edit(\n",
    "                rel_pos=ref_pos,\n",
    "                ref_base=ref_base,\n",
    "                alt_base=alt_base,\n",
    "                offset=offset,\n",
    "                strand=strand,\n",
    "            )\n",
    "        )\n",
    "    return allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5a3c62bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(phased_snps) == 1:\n",
    "    raise ValueError(\"Cannot phase variant of interest.\")\n",
    "\n",
    "phased_snps['alleles'] = phased_snps.apply(get_allele, axis=1)\n",
    "phased_snps['amplicon_pos'] = phased_snps.POS - var_pos + amplicon_start\n",
    "phased_snps[\"amplicon_base\"] = phased_snps.amplicon_pos.map(\n",
    "    lambda i: amplicon_seq[i]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "509f8c4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "phased_snps_nonvar = phased_snps.loc[phased_snps.amplicon_pos != amplicon_start]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "a48a4956",
   "metadata": {},
   "outputs": [],
   "source": [
    "editable_phase = np.where(np.array([p in edit_dict.keys() for p in phased_snps.loc[phased_snps.amplicon_pos == amplicon_start, \"alleles\"].item()]))[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "b82ec23d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>HepG2</th>\n",
       "      <th>GT</th>\n",
       "      <th>alleles</th>\n",
       "      <th>amplicon_pos</th>\n",
       "      <th>amplicon_base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58662212</td>\n",
       "      <td>rs78560062</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>736.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AN=2;DB;MUMAP_REF=59.6667;MUMAP_ALT=60;AO=18;R...</td>\n",
       "      <td>GT:BX:PS:PQ:JQ</td>\n",
       "      <td>1|0:ACAAGGGGTGTAAGTA-1_69;ACGATCATCGGGTCCA-1_7...</td>\n",
       "      <td>1|0</td>\n",
       "      <td>[T, C]</td>\n",
       "      <td>105</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58662235</td>\n",
       "      <td>rs35081008</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>920.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AN=2;DB;MUMAP_REF=60;MUMAP_ALT=59.619;AO=21;RO...</td>\n",
       "      <td>GT:BX:PS:PQ:JQ</td>\n",
       "      <td>0|1:TATGCCCTCTGTTGTT-1_69;ATTGTTCTCGTAGGAG-1_6...</td>\n",
       "      <td>0|1</td>\n",
       "      <td>[C, T]</td>\n",
       "      <td>128</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM       POS          ID REF ALT    QUAL FILTER  \\\n",
       "0  chr19  58662212  rs78560062   C   T  736.77   PASS   \n",
       "1  chr19  58662235  rs35081008   C   T  920.77   PASS   \n",
       "\n",
       "                                                INFO          FORMAT  \\\n",
       "0  AN=2;DB;MUMAP_REF=59.6667;MUMAP_ALT=60;AO=18;R...  GT:BX:PS:PQ:JQ   \n",
       "1  AN=2;DB;MUMAP_REF=60;MUMAP_ALT=59.619;AO=21;RO...  GT:BX:PS:PQ:JQ   \n",
       "\n",
       "                                               HepG2   GT alleles  \\\n",
       "0  1|0:ACAAGGGGTGTAAGTA-1_69;ACGATCATCGGGTCCA-1_7...  1|0  [T, C]   \n",
       "1  0|1:TATGCCCTCTGTTGTT-1_69;ATTGTTCTCGTAGGAG-1_6...  0|1  [C, T]   \n",
       "\n",
       "   amplicon_pos amplicon_base  \n",
       "0           105             C  \n",
       "1           128             C  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phased_snps"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "d80e9aca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nphase0    TT\\nphase1    CC\\n'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phases = pd.DataFrame(\n",
    "    phased_snps_nonvar.alleles.tolist(), \n",
    "    columns=[f'phase{i}' for i in range(len(phased_snps_nonvar.alleles.tolist()[0]))]\n",
    ").sum(axis=0)\n",
    "'''\n",
    "phase0    TT\n",
    "phase1    CC\n",
    "'''\n",
    "#pd.DataFrame(df2[\"teams\"].to_list(), columns=['team1', 'team2'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "50c78ac5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "phase0    T\n",
       "phase1    C\n",
       "dtype: object"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5c4401f6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>ID</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>QUAL</th>\n",
       "      <th>FILTER</th>\n",
       "      <th>INFO</th>\n",
       "      <th>FORMAT</th>\n",
       "      <th>HepG2</th>\n",
       "      <th>GT</th>\n",
       "      <th>alleles</th>\n",
       "      <th>amplicon_pos</th>\n",
       "      <th>amplicon_base</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58662212</td>\n",
       "      <td>rs78560062</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>736.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AN=2;DB;MUMAP_REF=59.6667;MUMAP_ALT=60;AO=18;R...</td>\n",
       "      <td>GT:BX:PS:PQ:JQ</td>\n",
       "      <td>1|0:ACAAGGGGTGTAAGTA-1_69;ACGATCATCGGGTCCA-1_7...</td>\n",
       "      <td>1|0</td>\n",
       "      <td>[T, C]</td>\n",
       "      <td>105</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>58662235</td>\n",
       "      <td>rs35081008</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>920.77</td>\n",
       "      <td>PASS</td>\n",
       "      <td>AN=2;DB;MUMAP_REF=60;MUMAP_ALT=59.619;AO=21;RO...</td>\n",
       "      <td>GT:BX:PS:PQ:JQ</td>\n",
       "      <td>0|1:TATGCCCTCTGTTGTT-1_69;ATTGTTCTCGTAGGAG-1_6...</td>\n",
       "      <td>0|1</td>\n",
       "      <td>[C, T]</td>\n",
       "      <td>128</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   CHROM       POS          ID REF ALT    QUAL FILTER  \\\n",
       "0  chr19  58662212  rs78560062   C   T  736.77   PASS   \n",
       "1  chr19  58662235  rs35081008   C   T  920.77   PASS   \n",
       "\n",
       "                                                INFO          FORMAT  \\\n",
       "0  AN=2;DB;MUMAP_REF=59.6667;MUMAP_ALT=60;AO=18;R...  GT:BX:PS:PQ:JQ   \n",
       "1  AN=2;DB;MUMAP_REF=60;MUMAP_ALT=59.619;AO=21;RO...  GT:BX:PS:PQ:JQ   \n",
       "\n",
       "                                               HepG2   GT alleles  \\\n",
       "0  1|0:ACAAGGGGTGTAAGTA-1_69;ACGATCATCGGGTCCA-1_7...  1|0  [T, C]   \n",
       "1  0|1:TATGCCCTCTGTTGTT-1_69;ATTGTTCTCGTAGGAG-1_6...  0|1  [C, T]   \n",
       "\n",
       "   amplicon_pos amplicon_base  \n",
       "0           105             C  \n",
       "1           128             C  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phased_snps"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "094f3bcc",
   "metadata": {},
   "source": [
    "## Phase reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "0cd95020",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_edit(row, target_pos):\n",
    "    return _get_allele_from_alignment(row.Reference_Sequence, row.Aligned_Sequence, offset=0, strand=1, start_pos=target_pos-20, end_pos=target_pos+20)\n",
    "\n",
    "def get_all_edits(row):\n",
    "    return _get_allele_from_alignment(row.Reference_Sequence, row.Aligned_Sequence, offset=0, strand=1, start_pos=phased_snps.amplicon_pos.min()-1, end_pos = phased_snps.amplicon_pos.max()+1)\n",
    "    \n",
    "\n",
    "def filter_allele_by_pos_qual(allele, target_pos):\n",
    "    # filters allele by variant included in list_pos or near target_pos\n",
    "    edits = []\n",
    "    n_out_edits = 0\n",
    "    n_lq = 0\n",
    "    n_crit_lq = 0\n",
    "    for edit in allele.split(\",\"):\n",
    "        if edit == \"\": continue\n",
    "        alt = edit.split(\">\")[-1]\n",
    "        pos = int(edit.split(\":\")[0])\n",
    "        if alt == \"N\": \n",
    "            n_lq += 1\n",
    "            if abs(pos-target_pos) < 5:\n",
    "                n_crit_lq += 1\n",
    "        elif abs(pos-target_pos) < 5:\n",
    "            edits.append(edit)\n",
    "        else:\n",
    "            n_out_edits += 1\n",
    "    return n_out_edits, n_lq, n_crit_lq, \",\".join(sorted(edits))\n",
    "\n",
    "def filter_allele_by_pos(list_pos, allele, target_pos):\n",
    "    edits = []\n",
    "    for edit in allele.split(\",\"):\n",
    "        if edit == \"\": continue\n",
    "        pos = int(edit.split(\":\")[0])\n",
    "        if pos in list_pos or abs(pos-target_pos) < 5:\n",
    "            edits.append(edit)\n",
    "    return \",\".join(sorted(edits))\n",
    "\n",
    "def filter_out_indels(allele):\n",
    "    edits = []\n",
    "    for edit in allele.edits:\n",
    "        if edit.ref_base != \"-\" and edit.alt_base != \"-\":\n",
    "            edits.append(str(edit))\n",
    "    return \",\".join(sorted(edits))\n",
    "\n",
    "\n",
    "\n",
    "def get_phase(allele, phased_snps, var_pos):\n",
    "    phase = []\n",
    "    allele_edits = {}\n",
    "    if allele != \"\":\n",
    "        edits = allele.split(\",\")\n",
    "        for edit in edits:\n",
    "            try:\n",
    "                allele_edits[int(edit.split(\":\")[0])] = edit.split(\":\")[-1]\n",
    "            except ValueError:\n",
    "                print(edit)\n",
    "    for pos in phased_snps.amplicon_pos.tolist():\n",
    "        if pos == var_pos: continue\n",
    "        if pos in allele_edits.keys():\n",
    "            phase.append(allele_edits[pos].split(\">\")[-1])\n",
    "        else:\n",
    "            phase.append(phased_snps[['amplicon_pos', 'amplicon_base']].set_index(\"amplicon_pos\").loc[pos, \"amplicon_base\"])\n",
    "    return phase\n",
    "\n",
    "\n",
    "def mask_seq(seq, mask_pos):\n",
    "    seq_array = np.array(list(seq))\n",
    "    seq_array[mask_pos] = \"N\"\n",
    "    return \"\".join(seq_array.tolist())\n",
    "    \n",
    "def assign_phase(phases, allele_phase):\n",
    "    match= np.where(phases == allele_phase[0])[0]\n",
    "    if len(match) != 1:\n",
    "        return -1\n",
    "    return match[0]\n",
    "def filter_out_indels(allele):\n",
    "    edits = []\n",
    "    for edit in allele.edits:\n",
    "        if edit.ref_base != \"-\" and edit.alt_base != \"-\":\n",
    "            edits.append(str(edit))\n",
    "    return \",\".join(sorted(edits))\n",
    "\n",
    "def get_pos_edit_rate(aftbl):\n",
    "    total_counts = aftbl['#Reads'].sum()\n",
    "    aftbl = aftbl.loc[aftbl.be_allele != \"\",:].copy()\n",
    "    aftbl['edits'] = aftbl.be_allele.map(lambda a: a.split(\",\"))\n",
    "    edit_tbl = aftbl.explode(\"edits\")\n",
    "    edit_tbl['pos'] = edit_tbl.edits.map(lambda e: int(e.split(\":\")[0]))\n",
    "    per_pos_edits = edit_tbl.groupby(\"pos\")[\"#Reads\"].sum()\n",
    "    per_pos_edit_rates = per_pos_edits / aftbl['#Reads'].sum()\n",
    "    return np.where(per_pos_edit_rates > 0.5)[0]\n",
    "\n",
    "def format_allele_freq_table(path, varpos_amplicon):\n",
    "    aftbl = pd.read_csv(path, sep='\\t')\n",
    "    aftbl[\"allele\"] = aftbl.apply(lambda row:get_all_edits(row),axis=1)\n",
    "    aftbl[\"allele_around_target\"] = aftbl.apply(lambda row:get_edit(row, varpos_amplicon),axis=1)\n",
    "    aftbl['be_allele'] = aftbl.allele.map(filter_out_indels)\n",
    "    aftbl['be_allele_around_target'] = aftbl.allele_around_target.map(filter_out_indels)\n",
    "    aftbl['err'] = aftbl.be_allele_around_target.map(lambda a:filter_allele_by_pos_qual(a, varpos_amplicon)[0])\n",
    "    aftbl['N'] = aftbl.be_allele_around_target.map(lambda a:filter_allele_by_pos_qual(a, varpos_amplicon)[1])\n",
    "    aftbl['target_proximal_N'] = aftbl.be_allele_around_target.map(lambda a:filter_allele_by_pos_qual(a, varpos_amplicon)[2])\n",
    "    aftbl['phased_allele'] = aftbl.be_allele.map(lambda a: filter_allele_by_pos(phased_snps.amplicon_pos.tolist(), a, varpos_amplicon))\n",
    "    aftbl = aftbl.loc[(aftbl.err <= 1) & (aftbl.target_proximal_N == 0) & (aftbl.N <= 2)]\n",
    "    aftbl['phase'] = aftbl.phased_allele.map(lambda a: get_phase(a, phased_snps, amplicon_start))\n",
    "    aftbl[\"assigned_phase\"] = aftbl.phase.map(lambda ap: assign_phase(phases, ap))\n",
    "    return aftbl\n",
    "\n",
    "def get_phased_edit_rate(aftbl):\n",
    "    var_amplicon_base = phased_snps.loc[phased_snps.amplicon_pos == amplicon_start, \"amplicon_base\"].item()\n",
    "    aftbl.loc[~aftbl.assigned_phase.isin(editable_phase), \"edited\"] = np.nan\n",
    "    if var_amplicon_base == var_alleles.ref.item():\n",
    "        aftbl.loc[aftbl.assigned_phase.isin(editable_phase),'edited'] = aftbl.loc[aftbl.assigned_phase.isin(editable_phase),:].phased_allele.map(lambda s: f\"{amplicon_start}:{amplicon_start}:+:{var_alleles.ref}>{var_alleles.edited.item()}\" in s)\n",
    "    elif var_amplicon_base == var_alleles.edited.item():\n",
    "        aftbl.loc[aftbl.assigned_phase.isin(editable_phase),'edited'] = aftbl.loc[aftbl.assigned_phase.isin(editable_phase),:].phased_allele.map(lambda s: f\"{amplicon_start}:{amplicon_start}:+:{var_amplicon_base}>{var_alleles.ref.item()}\" not in s)\n",
    "    aftbl['edited_clean'] = aftbl['edited'].map(lambda s: False if np.isnan(s) else s)\n",
    "    return aftbl.groupby(['assigned_phase', 'edited_clean'])['#Reads'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "46be73bc",
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "d3d34ac8c731408496c287939aab5b07",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6cad0d62da54443b80b801609a6f89ec",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6fbe271d7f6f47249d9be31f0c56f7b3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "bae36a6329ee4821a0b0cfae984abe33",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "dc996ecf232f4b3d84f05918dcfeac2b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "results = {}\n",
    "for rep in tqdm([1,2,3, 'C']):\n",
    "    for cond in tqdm([\"Ser\", \"SS\"]):\n",
    "        exp_id = f\"{rep}-{cond}_20-locus\"\n",
    "        atac_path = f\"../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_{rep}-{cond}_20-locus_ATAC_Seq.{var_id}/Alleles_frequency_table.txt\"\n",
    "        gdna_path = f\"../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_{rep}-{cond}_20-locus_GDNA.{var_id}/Alleles_frequency_table.txt\"\n",
    "        \n",
    "        if not os.path.exists(atac_path):\n",
    "            os.system(f\"unzip ../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_{rep}-{cond}_20-locus_ATAC_Seq.{var_id}/Alleles_frequency_table.zip -d ../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_{rep}-{cond}_20-locus_ATAC_Seq.{var_id}/\")\n",
    "        aftbl_atac = format_allele_freq_table(atac_path, amplicon_start)\n",
    "            \n",
    "        if not os.path.exists(gdna_path):\n",
    "            os.system(f\"unzip ../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_{rep}-{cond}_20-locus_GDNA.{var_id}/Alleles_frequency_table.zip -d ../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_{rep}-{cond}_20-locus_GDNA.{var_id}/\")\n",
    "        aftbl_gdna = format_allele_freq_table(gdna_path, amplicon_start)\n",
    "        results[exp_id] = [get_phased_edit_rate(aftbl_atac), get_phased_edit_rate(aftbl_gdna)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c79bf2f0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                               atac    gdna\n",
      "assigned_phase edited_clean                \n",
      "-1             False            202     247\n",
      " 0             False         185752  235731\n",
      " 1             False           2328   19584\n",
      "               True             112     309\n",
      "                               atac    gdna\n",
      "assigned_phase edited_clean                \n",
      "-1             False            117     320\n",
      " 0             False         110293  306440\n",
      " 1             False           1392   20428\n",
      "               True              92     407\n",
      "                               atac    gdna\n",
      "assigned_phase edited_clean                \n",
      "-1             False            171     269\n",
      " 0             False         164381  256479\n",
      " 1             False           3220   17066\n",
      "               True             216     796\n",
      "                              atac    gdna\n",
      "assigned_phase edited_clean               \n",
      "-1             False            42     328\n",
      " 0             False         40050  279510\n",
      " 1             False          2188   19387\n",
      "               True             91     817\n",
      "                               atac    gdna\n",
      "assigned_phase edited_clean                \n",
      "-1             False            123     268\n",
      " 0             False         105297  228521\n",
      " 1             False           1072   10911\n",
      "               True             172    2499\n",
      "                               atac    gdna\n",
      "assigned_phase edited_clean                \n",
      "-1             False            213     279\n",
      " 0             False         209214  257256\n",
      " 1             False           1009   14710\n",
      "               True             217    3533\n",
      "                              atac    gdna\n",
      "assigned_phase edited_clean               \n",
      "-1             False           103     324\n",
      " 0             False         97455  294078\n",
      " 1             False          2270   19939\n",
      "               True             30     340\n",
      "                               atac    gdna\n",
      "assigned_phase edited_clean                \n",
      "-1             False            112     221\n",
      " 0             False         109869  177951\n",
      " 1             False           2379   18832\n",
      "               True              45     310\n"
     ]
    }
   ],
   "source": [
    "res = []\n",
    "for k, dfs in results.items():\n",
    "    catdf = dfs[0].to_frame().join(dfs[1].to_frame(), how='outer', rsuffix='gdna').fillna(0)\n",
    "    catdf.columns = [\"atac\", \"gdna\"]\n",
    "    print(catdf)\n",
    "    if len(catdf) == 1:\n",
    "        catdf.loc[(0,False),:] = (0,0)\n",
    "    if len(catdf) == 2: \n",
    "        cols= catdf.index.get_level_values(0)\n",
    "    if len(catdf) == 3:\n",
    "        catdf.loc[(-1,False),:] = (0,0)\n",
    "        catdf = catdf.loc[[(-1,False),(0,False),(1,False),(1,True)],:]\n",
    "        print(catdf)\n",
    "    if len(catdf) == 4:\n",
    "        cols = [\"Unassigned\", \"Phase0\", \"Phase1_unedited\", \"Phase1_edited\"]\n",
    "    catdf = catdf.T\n",
    "    catdf.columns=cols\n",
    "    catdf['exp'] = k\n",
    "    res.append(catdf)\n",
    "results_df = pd.concat(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7b96f6a7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unassigned</th>\n",
       "      <th>Phase0</th>\n",
       "      <th>Phase1_unedited</th>\n",
       "      <th>Phase1_edited</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>202</td>\n",
       "      <td>185752</td>\n",
       "      <td>2328</td>\n",
       "      <td>112</td>\n",
       "      <td>1-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>247</td>\n",
       "      <td>235731</td>\n",
       "      <td>19584</td>\n",
       "      <td>309</td>\n",
       "      <td>1-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>117</td>\n",
       "      <td>110293</td>\n",
       "      <td>1392</td>\n",
       "      <td>92</td>\n",
       "      <td>1-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>320</td>\n",
       "      <td>306440</td>\n",
       "      <td>20428</td>\n",
       "      <td>407</td>\n",
       "      <td>1-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>171</td>\n",
       "      <td>164381</td>\n",
       "      <td>3220</td>\n",
       "      <td>216</td>\n",
       "      <td>2-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>269</td>\n",
       "      <td>256479</td>\n",
       "      <td>17066</td>\n",
       "      <td>796</td>\n",
       "      <td>2-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>42</td>\n",
       "      <td>40050</td>\n",
       "      <td>2188</td>\n",
       "      <td>91</td>\n",
       "      <td>2-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>328</td>\n",
       "      <td>279510</td>\n",
       "      <td>19387</td>\n",
       "      <td>817</td>\n",
       "      <td>2-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>123</td>\n",
       "      <td>105297</td>\n",
       "      <td>1072</td>\n",
       "      <td>172</td>\n",
       "      <td>3-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>268</td>\n",
       "      <td>228521</td>\n",
       "      <td>10911</td>\n",
       "      <td>2499</td>\n",
       "      <td>3-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>213</td>\n",
       "      <td>209214</td>\n",
       "      <td>1009</td>\n",
       "      <td>217</td>\n",
       "      <td>3-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>279</td>\n",
       "      <td>257256</td>\n",
       "      <td>14710</td>\n",
       "      <td>3533</td>\n",
       "      <td>3-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>103</td>\n",
       "      <td>97455</td>\n",
       "      <td>2270</td>\n",
       "      <td>30</td>\n",
       "      <td>C-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>324</td>\n",
       "      <td>294078</td>\n",
       "      <td>19939</td>\n",
       "      <td>340</td>\n",
       "      <td>C-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>112</td>\n",
       "      <td>109869</td>\n",
       "      <td>2379</td>\n",
       "      <td>45</td>\n",
       "      <td>C-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>221</td>\n",
       "      <td>177951</td>\n",
       "      <td>18832</td>\n",
       "      <td>310</td>\n",
       "      <td>C-SS_20-locus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unassigned  Phase0  Phase1_unedited  Phase1_edited             exp\n",
       "atac         202  185752             2328            112  1-Ser_20-locus\n",
       "gdna         247  235731            19584            309  1-Ser_20-locus\n",
       "atac         117  110293             1392             92   1-SS_20-locus\n",
       "gdna         320  306440            20428            407   1-SS_20-locus\n",
       "atac         171  164381             3220            216  2-Ser_20-locus\n",
       "gdna         269  256479            17066            796  2-Ser_20-locus\n",
       "atac          42   40050             2188             91   2-SS_20-locus\n",
       "gdna         328  279510            19387            817   2-SS_20-locus\n",
       "atac         123  105297             1072            172  3-Ser_20-locus\n",
       "gdna         268  228521            10911           2499  3-Ser_20-locus\n",
       "atac         213  209214             1009            217   3-SS_20-locus\n",
       "gdna         279  257256            14710           3533   3-SS_20-locus\n",
       "atac         103   97455             2270             30  C-Ser_20-locus\n",
       "gdna         324  294078            19939            340  C-Ser_20-locus\n",
       "atac         112  109869             2379             45   C-SS_20-locus\n",
       "gdna         221  177951            18832            310   C-SS_20-locus"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0b390659",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unassigned</th>\n",
       "      <th>Phase0</th>\n",
       "      <th>Phase1_unedited</th>\n",
       "      <th>Phase1_edited</th>\n",
       "      <th>exp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>202</td>\n",
       "      <td>185752</td>\n",
       "      <td>2328</td>\n",
       "      <td>112</td>\n",
       "      <td>1-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>247</td>\n",
       "      <td>235731</td>\n",
       "      <td>19584</td>\n",
       "      <td>309</td>\n",
       "      <td>1-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>117</td>\n",
       "      <td>110293</td>\n",
       "      <td>1392</td>\n",
       "      <td>92</td>\n",
       "      <td>1-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>320</td>\n",
       "      <td>306440</td>\n",
       "      <td>20428</td>\n",
       "      <td>407</td>\n",
       "      <td>1-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>171</td>\n",
       "      <td>164381</td>\n",
       "      <td>3220</td>\n",
       "      <td>216</td>\n",
       "      <td>2-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>269</td>\n",
       "      <td>256479</td>\n",
       "      <td>17066</td>\n",
       "      <td>796</td>\n",
       "      <td>2-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>42</td>\n",
       "      <td>40050</td>\n",
       "      <td>2188</td>\n",
       "      <td>91</td>\n",
       "      <td>2-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>328</td>\n",
       "      <td>279510</td>\n",
       "      <td>19387</td>\n",
       "      <td>817</td>\n",
       "      <td>2-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>123</td>\n",
       "      <td>105297</td>\n",
       "      <td>1072</td>\n",
       "      <td>172</td>\n",
       "      <td>3-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>268</td>\n",
       "      <td>228521</td>\n",
       "      <td>10911</td>\n",
       "      <td>2499</td>\n",
       "      <td>3-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>213</td>\n",
       "      <td>209214</td>\n",
       "      <td>1009</td>\n",
       "      <td>217</td>\n",
       "      <td>3-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>279</td>\n",
       "      <td>257256</td>\n",
       "      <td>14710</td>\n",
       "      <td>3533</td>\n",
       "      <td>3-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>103</td>\n",
       "      <td>97455</td>\n",
       "      <td>2270</td>\n",
       "      <td>30</td>\n",
       "      <td>C-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>324</td>\n",
       "      <td>294078</td>\n",
       "      <td>19939</td>\n",
       "      <td>340</td>\n",
       "      <td>C-Ser_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>112</td>\n",
       "      <td>109869</td>\n",
       "      <td>2379</td>\n",
       "      <td>45</td>\n",
       "      <td>C-SS_20-locus</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>221</td>\n",
       "      <td>177951</td>\n",
       "      <td>18832</td>\n",
       "      <td>310</td>\n",
       "      <td>C-SS_20-locus</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unassigned  Phase0  Phase1_unedited  Phase1_edited             exp\n",
       "atac         202  185752             2328            112  1-Ser_20-locus\n",
       "gdna         247  235731            19584            309  1-Ser_20-locus\n",
       "atac         117  110293             1392             92   1-SS_20-locus\n",
       "gdna         320  306440            20428            407   1-SS_20-locus\n",
       "atac         171  164381             3220            216  2-Ser_20-locus\n",
       "gdna         269  256479            17066            796  2-Ser_20-locus\n",
       "atac          42   40050             2188             91   2-SS_20-locus\n",
       "gdna         328  279510            19387            817   2-SS_20-locus\n",
       "atac         123  105297             1072            172  3-Ser_20-locus\n",
       "gdna         268  228521            10911           2499  3-Ser_20-locus\n",
       "atac         213  209214             1009            217   3-SS_20-locus\n",
       "gdna         279  257256            14710           3533   3-SS_20-locus\n",
       "atac         103   97455             2270             30  C-Ser_20-locus\n",
       "gdna         324  294078            19939            340  C-Ser_20-locus\n",
       "atac         112  109869             2379             45   C-SS_20-locus\n",
       "gdna         221  177951            18832            310   C-SS_20-locus"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a4bc299b",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df['is_atac'] = results_df.index == 'atac'\n",
    "results_df['phase0_frac'] = results_df[\"Phase0\"] / results_df.iloc[:,:4].sum(axis=1)\n",
    "results_df['phase1_frac'] = (results_df[\"Phase1_unedited\"]+results_df[\"Phase1_edited\"]) / results_df.iloc[:,:4].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "47ab5035",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df[\"Phase1\"] = results_df[[\"Phase1_edited\", \"Phase1_unedited\"]].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "b3c729b5",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "results_df['exp'] = results_df.exp.map(lambda x: x.split(\"_\")[0])\n",
    "results_df[['rep', 'cond']] = results_df.exp.str.split(\"-\", expand=True)\n",
    "results_df['is_trt'] = results_df.rep != \"C\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "6238ba9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "y, X = dmatrices('Phase1_edited + Phase1_unedited ~ is_trt', data = results_df, return_type='dataframe')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "6669c7f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_binom = sm.GLM(y, X, family=sm.families.Binomial())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "8e930263",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_res = glm_binom.fit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "a24b397f",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_html = glm_res.summary().tables[1].as_html()\n",
    "glm_res_tbl = pd.read_html(result_html, header=0, index_col=0)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "fe9d1bf1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coef</th>\n",
       "      <th>std err</th>\n",
       "      <th>z</th>\n",
       "      <th>P&gt;|z|</th>\n",
       "      <th>[0.025</th>\n",
       "      <th>0.975]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Intercept</th>\n",
       "      <td>-4.0925</td>\n",
       "      <td>0.037</td>\n",
       "      <td>-109.285</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-4.166</td>\n",
       "      <td>-4.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_trt[T.True]</th>\n",
       "      <td>1.5883</td>\n",
       "      <td>0.039</td>\n",
       "      <td>40.751</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.512</td>\n",
       "      <td>1.665</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  coef  std err        z  P>|z|  [0.025  0.975]\n",
       "Intercept      -4.0925    0.037 -109.285    0.0  -4.166  -4.019\n",
       "is_trt[T.True]  1.5883    0.039   40.751    0.0   1.512   1.665"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glm_res_tbl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "b3f1e660",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_res_tbl.to_csv(\"rs35081008_Min_ABE_465_ctrl_result.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "5b70b17f",
   "metadata": {},
   "outputs": [],
   "source": [
    "y, X = dmatrices('Phase0 + Phase1 ~ rep + cond + is_atac + cond*is_atac', data = results_df, return_type='dataframe')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "ba944516",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_binom = sm.GLM(y, X, family=sm.families.Binomial())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "c4c45508",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_res = glm_binom.fit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "33e1b219",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_html = glm_res.summary().tables[1].as_html()\n",
    "glm_res_tbl = pd.read_html(result_html, header=0, index_col=0)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "76235d43",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coef</th>\n",
       "      <th>std err</th>\n",
       "      <th>z</th>\n",
       "      <th>P&gt;|z|</th>\n",
       "      <th>[0.025</th>\n",
       "      <th>0.975]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Intercept</th>\n",
       "      <td>2.5674</td>\n",
       "      <td>0.005</td>\n",
       "      <td>468.633</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.557</td>\n",
       "      <td>2.578</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rep[T.2]</th>\n",
       "      <td>-0.0331</td>\n",
       "      <td>0.007</td>\n",
       "      <td>-4.755</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.047</td>\n",
       "      <td>-0.019</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rep[T.3]</th>\n",
       "      <td>0.1765</td>\n",
       "      <td>0.007</td>\n",
       "      <td>23.754</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.162</td>\n",
       "      <td>0.191</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rep[T.C]</th>\n",
       "      <td>-0.1695</td>\n",
       "      <td>0.007</td>\n",
       "      <td>-24.170</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.183</td>\n",
       "      <td>-0.156</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cond[T.Ser]</th>\n",
       "      <td>0.1108</td>\n",
       "      <td>0.005</td>\n",
       "      <td>20.483</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.100</td>\n",
       "      <td>0.121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_atac[T.True]</th>\n",
       "      <td>1.5553</td>\n",
       "      <td>0.012</td>\n",
       "      <td>125.836</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.531</td>\n",
       "      <td>1.580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cond[T.Ser]:is_atac[T.True]</th>\n",
       "      <td>-0.1493</td>\n",
       "      <td>0.017</td>\n",
       "      <td>-8.955</td>\n",
       "      <td>0.0</td>\n",
       "      <td>-0.182</td>\n",
       "      <td>-0.117</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               coef  std err        z  P>|z|  [0.025  0.975]\n",
       "Intercept                    2.5674    0.005  468.633    0.0   2.557   2.578\n",
       "rep[T.2]                    -0.0331    0.007   -4.755    0.0  -0.047  -0.019\n",
       "rep[T.3]                     0.1765    0.007   23.754    0.0   0.162   0.191\n",
       "rep[T.C]                    -0.1695    0.007  -24.170    0.0  -0.183  -0.156\n",
       "cond[T.Ser]                  0.1108    0.005   20.483    0.0   0.100   0.121\n",
       "is_atac[T.True]              1.5553    0.012  125.836    0.0   1.531   1.580\n",
       "cond[T.Ser]:is_atac[T.True] -0.1493    0.017   -8.955    0.0  -0.182  -0.117"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glm_res_tbl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "a122bed7",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_res_tbl.to_csv(\"rs35081008_Min_ABE_465_caqtl_result.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cf93d8c1",
   "metadata": {},
   "source": [
    "Phase 0 is enriched in ATAC-seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "a798e51f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unassigned</th>\n",
       "      <th>Phase0</th>\n",
       "      <th>Phase1_unedited</th>\n",
       "      <th>Phase1_edited</th>\n",
       "      <th>exp</th>\n",
       "      <th>is_atac</th>\n",
       "      <th>phase0_frac</th>\n",
       "      <th>phase1_frac</th>\n",
       "      <th>Phase1</th>\n",
       "      <th>rep</th>\n",
       "      <th>cond</th>\n",
       "      <th>is_trt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>202</td>\n",
       "      <td>185752</td>\n",
       "      <td>2328</td>\n",
       "      <td>112</td>\n",
       "      <td>1-Ser</td>\n",
       "      <td>True</td>\n",
       "      <td>0.985976</td>\n",
       "      <td>0.012952</td>\n",
       "      <td>2440</td>\n",
       "      <td>1</td>\n",
       "      <td>Ser</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>247</td>\n",
       "      <td>235731</td>\n",
       "      <td>19584</td>\n",
       "      <td>309</td>\n",
       "      <td>1-Ser</td>\n",
       "      <td>False</td>\n",
       "      <td>0.921288</td>\n",
       "      <td>0.077746</td>\n",
       "      <td>19893</td>\n",
       "      <td>1</td>\n",
       "      <td>Ser</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>117</td>\n",
       "      <td>110293</td>\n",
       "      <td>1392</td>\n",
       "      <td>92</td>\n",
       "      <td>1-SS</td>\n",
       "      <td>True</td>\n",
       "      <td>0.985692</td>\n",
       "      <td>0.013263</td>\n",
       "      <td>1484</td>\n",
       "      <td>1</td>\n",
       "      <td>SS</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>320</td>\n",
       "      <td>306440</td>\n",
       "      <td>20428</td>\n",
       "      <td>407</td>\n",
       "      <td>1-SS</td>\n",
       "      <td>False</td>\n",
       "      <td>0.935423</td>\n",
       "      <td>0.063600</td>\n",
       "      <td>20835</td>\n",
       "      <td>1</td>\n",
       "      <td>SS</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>171</td>\n",
       "      <td>164381</td>\n",
       "      <td>3220</td>\n",
       "      <td>216</td>\n",
       "      <td>2-Ser</td>\n",
       "      <td>True</td>\n",
       "      <td>0.978528</td>\n",
       "      <td>0.020454</td>\n",
       "      <td>3436</td>\n",
       "      <td>2</td>\n",
       "      <td>Ser</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>269</td>\n",
       "      <td>256479</td>\n",
       "      <td>17066</td>\n",
       "      <td>796</td>\n",
       "      <td>2-Ser</td>\n",
       "      <td>False</td>\n",
       "      <td>0.933975</td>\n",
       "      <td>0.065045</td>\n",
       "      <td>17862</td>\n",
       "      <td>2</td>\n",
       "      <td>Ser</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>42</td>\n",
       "      <td>40050</td>\n",
       "      <td>2188</td>\n",
       "      <td>91</td>\n",
       "      <td>2-SS</td>\n",
       "      <td>True</td>\n",
       "      <td>0.945222</td>\n",
       "      <td>0.053787</td>\n",
       "      <td>2279</td>\n",
       "      <td>2</td>\n",
       "      <td>SS</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>328</td>\n",
       "      <td>279510</td>\n",
       "      <td>19387</td>\n",
       "      <td>817</td>\n",
       "      <td>2-SS</td>\n",
       "      <td>False</td>\n",
       "      <td>0.931570</td>\n",
       "      <td>0.067337</td>\n",
       "      <td>20204</td>\n",
       "      <td>2</td>\n",
       "      <td>SS</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>123</td>\n",
       "      <td>105297</td>\n",
       "      <td>1072</td>\n",
       "      <td>172</td>\n",
       "      <td>3-Ser</td>\n",
       "      <td>True</td>\n",
       "      <td>0.987184</td>\n",
       "      <td>0.011663</td>\n",
       "      <td>1244</td>\n",
       "      <td>3</td>\n",
       "      <td>Ser</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>268</td>\n",
       "      <td>228521</td>\n",
       "      <td>10911</td>\n",
       "      <td>2499</td>\n",
       "      <td>3-Ser</td>\n",
       "      <td>False</td>\n",
       "      <td>0.943526</td>\n",
       "      <td>0.055368</td>\n",
       "      <td>13410</td>\n",
       "      <td>3</td>\n",
       "      <td>Ser</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>213</td>\n",
       "      <td>209214</td>\n",
       "      <td>1009</td>\n",
       "      <td>217</td>\n",
       "      <td>3-SS</td>\n",
       "      <td>True</td>\n",
       "      <td>0.993169</td>\n",
       "      <td>0.005820</td>\n",
       "      <td>1226</td>\n",
       "      <td>3</td>\n",
       "      <td>SS</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>279</td>\n",
       "      <td>257256</td>\n",
       "      <td>14710</td>\n",
       "      <td>3533</td>\n",
       "      <td>3-SS</td>\n",
       "      <td>False</td>\n",
       "      <td>0.932837</td>\n",
       "      <td>0.066151</td>\n",
       "      <td>18243</td>\n",
       "      <td>3</td>\n",
       "      <td>SS</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>103</td>\n",
       "      <td>97455</td>\n",
       "      <td>2270</td>\n",
       "      <td>30</td>\n",
       "      <td>C-Ser</td>\n",
       "      <td>True</td>\n",
       "      <td>0.975936</td>\n",
       "      <td>0.023033</td>\n",
       "      <td>2300</td>\n",
       "      <td>C</td>\n",
       "      <td>Ser</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>324</td>\n",
       "      <td>294078</td>\n",
       "      <td>19939</td>\n",
       "      <td>340</td>\n",
       "      <td>C-Ser</td>\n",
       "      <td>False</td>\n",
       "      <td>0.934527</td>\n",
       "      <td>0.064443</td>\n",
       "      <td>20279</td>\n",
       "      <td>C</td>\n",
       "      <td>Ser</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>112</td>\n",
       "      <td>109869</td>\n",
       "      <td>2379</td>\n",
       "      <td>45</td>\n",
       "      <td>C-SS</td>\n",
       "      <td>True</td>\n",
       "      <td>0.977439</td>\n",
       "      <td>0.021565</td>\n",
       "      <td>2424</td>\n",
       "      <td>C</td>\n",
       "      <td>SS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>221</td>\n",
       "      <td>177951</td>\n",
       "      <td>18832</td>\n",
       "      <td>310</td>\n",
       "      <td>C-SS</td>\n",
       "      <td>False</td>\n",
       "      <td>0.901867</td>\n",
       "      <td>0.097013</td>\n",
       "      <td>19142</td>\n",
       "      <td>C</td>\n",
       "      <td>SS</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unassigned  Phase0  Phase1_unedited  Phase1_edited    exp  is_atac  \\\n",
       "atac         202  185752             2328            112  1-Ser     True   \n",
       "gdna         247  235731            19584            309  1-Ser    False   \n",
       "atac         117  110293             1392             92   1-SS     True   \n",
       "gdna         320  306440            20428            407   1-SS    False   \n",
       "atac         171  164381             3220            216  2-Ser     True   \n",
       "gdna         269  256479            17066            796  2-Ser    False   \n",
       "atac          42   40050             2188             91   2-SS     True   \n",
       "gdna         328  279510            19387            817   2-SS    False   \n",
       "atac         123  105297             1072            172  3-Ser     True   \n",
       "gdna         268  228521            10911           2499  3-Ser    False   \n",
       "atac         213  209214             1009            217   3-SS     True   \n",
       "gdna         279  257256            14710           3533   3-SS    False   \n",
       "atac         103   97455             2270             30  C-Ser     True   \n",
       "gdna         324  294078            19939            340  C-Ser    False   \n",
       "atac         112  109869             2379             45   C-SS     True   \n",
       "gdna         221  177951            18832            310   C-SS    False   \n",
       "\n",
       "      phase0_frac  phase1_frac  Phase1 rep cond  is_trt  \n",
       "atac     0.985976     0.012952    2440   1  Ser    True  \n",
       "gdna     0.921288     0.077746   19893   1  Ser    True  \n",
       "atac     0.985692     0.013263    1484   1   SS    True  \n",
       "gdna     0.935423     0.063600   20835   1   SS    True  \n",
       "atac     0.978528     0.020454    3436   2  Ser    True  \n",
       "gdna     0.933975     0.065045   17862   2  Ser    True  \n",
       "atac     0.945222     0.053787    2279   2   SS    True  \n",
       "gdna     0.931570     0.067337   20204   2   SS    True  \n",
       "atac     0.987184     0.011663    1244   3  Ser    True  \n",
       "gdna     0.943526     0.055368   13410   3  Ser    True  \n",
       "atac     0.993169     0.005820    1226   3   SS    True  \n",
       "gdna     0.932837     0.066151   18243   3   SS    True  \n",
       "atac     0.975936     0.023033    2300   C  Ser   False  \n",
       "gdna     0.934527     0.064443   20279   C  Ser   False  \n",
       "atac     0.977439     0.021565    2424   C   SS   False  \n",
       "gdna     0.901867     0.097013   19142   C   SS   False  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6dab62a3",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df.columns=results_df.columns.map(lambda c: c.replace(\" \", \"_\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ee9f2c5a",
   "metadata": {},
   "source": [
    "### Edits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "7db10033",
   "metadata": {},
   "outputs": [],
   "source": [
    "y, X = dmatrices('Phase1_edited + Phase1_unedited ~ is_trt', data = results_df, return_type='dataframe')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "705d8d57",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_binom = sm.GLM(y, X, family=sm.families.Binomial())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "4bfda437",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_res = glm_binom.fit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "b414e568",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"simpletable\">\n",
       "<caption>Generalized Linear Model Regression Results</caption>\n",
       "<tr>\n",
       "  <th>Dep. Variable:</th>   <td>['Phase1_edited', 'Phase1_unedited']</td> <th>  No. Observations:  </th>  <td>    16</td> \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Model:</th>                            <td>GLM</td>                 <th>  Df Residuals:      </th>  <td>    14</td> \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Model Family:</th>                  <td>Binomial</td>               <th>  Df Model:          </th>  <td>     1</td> \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Link Function:</th>                   <td>logit</td>                <th>  Scale:             </th> <td>  1.0000</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Method:</th>                          <td>IRLS</td>                 <th>  Log-Likelihood:    </th> <td> -4131.8</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Date:</th>                      <td>Wed, 14 Jun 2023</td>           <th>  Deviance:          </th> <td>  8145.9</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Time:</th>                          <td>13:49:49</td>               <th>  Pearson chi2:      </th> <td>8.90e+03</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>No. Iterations:</th>                    <td>7</td>                  <th>                     </th>     <td> </td>   \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Covariance Type:</th>               <td>nonrobust</td>              <th>                     </th>     <td> </td>   \n",
       "</tr>\n",
       "</table>\n",
       "<table class=\"simpletable\">\n",
       "<tr>\n",
       "         <td></td>           <th>coef</th>     <th>std err</th>      <th>z</th>      <th>P>|z|</th>  <th>[0.025</th>    <th>0.975]</th>  \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Intercept</th>      <td>   -4.0925</td> <td>    0.037</td> <td> -109.285</td> <td> 0.000</td> <td>   -4.166</td> <td>   -4.019</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>is_trt[T.True]</th> <td>    1.5883</td> <td>    0.039</td> <td>   40.751</td> <td> 0.000</td> <td>    1.512</td> <td>    1.665</td>\n",
       "</tr>\n",
       "</table>"
      ],
      "text/plain": [
       "<class 'statsmodels.iolib.summary.Summary'>\n",
       "\"\"\"\n",
       "                          Generalized Linear Model Regression Results                           \n",
       "================================================================================================\n",
       "Dep. Variable:     ['Phase1_edited', 'Phase1_unedited']   No. Observations:                   16\n",
       "Model:                                              GLM   Df Residuals:                       14\n",
       "Model Family:                                  Binomial   Df Model:                            1\n",
       "Link Function:                                    logit   Scale:                          1.0000\n",
       "Method:                                            IRLS   Log-Likelihood:                -4131.8\n",
       "Date:                                  Wed, 14 Jun 2023   Deviance:                       8145.9\n",
       "Time:                                          13:49:49   Pearson chi2:                 8.90e+03\n",
       "No. Iterations:                                       7                                         \n",
       "Covariance Type:                              nonrobust                                         \n",
       "==================================================================================\n",
       "                     coef    std err          z      P>|z|      [0.025      0.975]\n",
       "----------------------------------------------------------------------------------\n",
       "Intercept         -4.0925      0.037   -109.285      0.000      -4.166      -4.019\n",
       "is_trt[T.True]     1.5883      0.039     40.751      0.000       1.512       1.665\n",
       "==================================================================================\n",
       "\"\"\""
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glm_res.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "ff6564b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "y, X = dmatrices('Phase1_edited + Phase1_unedited ~ rep + cond + is_atac + cond*is_atac', data = results_df.loc[results_df.rep != \"C\",:], return_type='dataframe')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "8c849bc6",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_binom = sm.GLM(y, X, family=sm.families.Binomial())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "bfd14dcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_res = glm_binom.fit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "2e12f437",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"simpletable\">\n",
       "<caption>Generalized Linear Model Regression Results</caption>\n",
       "<tr>\n",
       "  <th>Dep. Variable:</th>   <td>['Phase1_edited', 'Phase1_unedited']</td> <th>  No. Observations:  </th>  <td>    12</td> \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Model:</th>                            <td>GLM</td>                 <th>  Df Residuals:      </th>  <td>     6</td> \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Model Family:</th>                  <td>Binomial</td>               <th>  Df Model:          </th>  <td>     5</td> \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Link Function:</th>                   <td>logit</td>                <th>  Scale:             </th> <td>  1.0000</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Method:</th>                          <td>IRLS</td>                 <th>  Log-Likelihood:    </th> <td> -150.59</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Date:</th>                      <td>Wed, 14 Jun 2023</td>           <th>  Deviance:          </th> <td>  209.45</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Time:</th>                          <td>13:49:49</td>               <th>  Pearson chi2:      </th>  <td>  244.</td> \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>No. Iterations:</th>                    <td>7</td>                  <th>                     </th>     <td> </td>   \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Covariance Type:</th>               <td>nonrobust</td>              <th>                     </th>     <td> </td>   \n",
       "</tr>\n",
       "</table>\n",
       "<table class=\"simpletable\">\n",
       "<tr>\n",
       "               <td></td>                  <th>coef</th>     <th>std err</th>      <th>z</th>      <th>P>|z|</th>  <th>[0.025</th>    <th>0.975]</th>  \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Intercept</th>                   <td>   -3.8585</td> <td>    0.035</td> <td> -108.984</td> <td> 0.000</td> <td>   -3.928</td> <td>   -3.789</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>rep[T.2]</th>                    <td>    0.7717</td> <td>    0.041</td> <td>   18.952</td> <td> 0.000</td> <td>    0.692</td> <td>    0.852</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>rep[T.3]</th>                    <td>    2.4007</td> <td>    0.036</td> <td>   66.459</td> <td> 0.000</td> <td>    2.330</td> <td>    2.472</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>cond[T.Ser]</th>                 <td>   -0.0376</td> <td>    0.024</td> <td>   -1.577</td> <td> 0.115</td> <td>   -0.084</td> <td>    0.009</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>is_atac[T.True]</th>             <td>    0.1174</td> <td>    0.056</td> <td>    2.079</td> <td> 0.038</td> <td>    0.007</td> <td>    0.228</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>cond[T.Ser]:is_atac[T.True]</th> <td>    0.0790</td> <td>    0.076</td> <td>    1.036</td> <td> 0.300</td> <td>   -0.070</td> <td>    0.228</td>\n",
       "</tr>\n",
       "</table>"
      ],
      "text/plain": [
       "<class 'statsmodels.iolib.summary.Summary'>\n",
       "\"\"\"\n",
       "                          Generalized Linear Model Regression Results                           \n",
       "================================================================================================\n",
       "Dep. Variable:     ['Phase1_edited', 'Phase1_unedited']   No. Observations:                   12\n",
       "Model:                                              GLM   Df Residuals:                        6\n",
       "Model Family:                                  Binomial   Df Model:                            5\n",
       "Link Function:                                    logit   Scale:                          1.0000\n",
       "Method:                                            IRLS   Log-Likelihood:                -150.59\n",
       "Date:                                  Wed, 14 Jun 2023   Deviance:                       209.45\n",
       "Time:                                          13:49:49   Pearson chi2:                     244.\n",
       "No. Iterations:                                       7                                         \n",
       "Covariance Type:                              nonrobust                                         \n",
       "===============================================================================================\n",
       "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
       "-----------------------------------------------------------------------------------------------\n",
       "Intercept                      -3.8585      0.035   -108.984      0.000      -3.928      -3.789\n",
       "rep[T.2]                        0.7717      0.041     18.952      0.000       0.692       0.852\n",
       "rep[T.3]                        2.4007      0.036     66.459      0.000       2.330       2.472\n",
       "cond[T.Ser]                    -0.0376      0.024     -1.577      0.115      -0.084       0.009\n",
       "is_atac[T.True]                 0.1174      0.056      2.079      0.038       0.007       0.228\n",
       "cond[T.Ser]:is_atac[T.True]     0.0790      0.076      1.036      0.300      -0.070       0.228\n",
       "===============================================================================================\n",
       "\"\"\""
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glm_res.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "0a6199a9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Intercept                      0.000000e+00\n",
       "rep[T.2]                       4.240342e-80\n",
       "rep[T.3]                       0.000000e+00\n",
       "cond[T.Ser]                    1.147559e-01\n",
       "is_atac[T.True]                3.759523e-02\n",
       "cond[T.Ser]:is_atac[T.True]    3.003828e-01\n",
       "dtype: float64"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glm_res.pvalues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "7ebaff53",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_html = glm_res.summary().tables[1].as_html()\n",
    "glm_res_tbl = pd.read_html(result_html, header=0, index_col=0)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "a7c08a7e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coef</th>\n",
       "      <th>std err</th>\n",
       "      <th>z</th>\n",
       "      <th>P&gt;|z|</th>\n",
       "      <th>[0.025</th>\n",
       "      <th>0.975]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Intercept</th>\n",
       "      <td>-3.8585</td>\n",
       "      <td>0.035</td>\n",
       "      <td>-108.984</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-3.928</td>\n",
       "      <td>-3.789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rep[T.2]</th>\n",
       "      <td>0.7717</td>\n",
       "      <td>0.041</td>\n",
       "      <td>18.952</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.692</td>\n",
       "      <td>0.852</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rep[T.3]</th>\n",
       "      <td>2.4007</td>\n",
       "      <td>0.036</td>\n",
       "      <td>66.459</td>\n",
       "      <td>0.000</td>\n",
       "      <td>2.330</td>\n",
       "      <td>2.472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cond[T.Ser]</th>\n",
       "      <td>-0.0376</td>\n",
       "      <td>0.024</td>\n",
       "      <td>-1.577</td>\n",
       "      <td>0.115</td>\n",
       "      <td>-0.084</td>\n",
       "      <td>0.009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_atac[T.True]</th>\n",
       "      <td>0.1174</td>\n",
       "      <td>0.056</td>\n",
       "      <td>2.079</td>\n",
       "      <td>0.038</td>\n",
       "      <td>0.007</td>\n",
       "      <td>0.228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cond[T.Ser]:is_atac[T.True]</th>\n",
       "      <td>0.0790</td>\n",
       "      <td>0.076</td>\n",
       "      <td>1.036</td>\n",
       "      <td>0.300</td>\n",
       "      <td>-0.070</td>\n",
       "      <td>0.228</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               coef  std err        z  P>|z|  [0.025  0.975]\n",
       "Intercept                   -3.8585    0.035 -108.984  0.000  -3.928  -3.789\n",
       "rep[T.2]                     0.7717    0.041   18.952  0.000   0.692   0.852\n",
       "rep[T.3]                     2.4007    0.036   66.459  0.000   2.330   2.472\n",
       "cond[T.Ser]                 -0.0376    0.024   -1.577  0.115  -0.084   0.009\n",
       "is_atac[T.True]              0.1174    0.056    2.079  0.038   0.007   0.228\n",
       "cond[T.Ser]:is_atac[T.True]  0.0790    0.076    1.036  0.300  -0.070   0.228"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glm_res_tbl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "b46eec78",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_res_tbl.to_csv(\"rs35081008_Min_ABE_465_result.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0358bfb2",
   "metadata": {},
   "source": [
    "Edited Phase 1 is enriched in ATAC-seq"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jy_anbe_py38",
   "language": "python",
   "name": "jy_anbe_py38"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
