{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2e4b7a3a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/pinello/SHARED_SOFTWARE/anaconda_latest/envs/jy_anbe_py38/lib/python3.8/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.\n",
      "  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,\n",
      "/data/pinello/SHARED_SOFTWARE/anaconda_latest/envs/jy_anbe_py38/lib/python3.8/site-packages/statsmodels/tsa/base/tsa_model.py:7: FutureWarning: pandas.Float64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.\n",
      "  from pandas import (to_datetime, Int64Index, DatetimeIndex, Period,\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from tqdm.auto import tqdm\n",
    "from patsy import dmatrices\n",
    "import statsmodels.api as sm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "75fc6aeb",
   "metadata": {},
   "outputs": [],
   "source": [
    "from bean.mapping._supporting_fn import _get_allele_from_alignment\n",
    "from bean import Allele,Edit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "57b0e602",
   "metadata": {},
   "outputs": [],
   "source": [
    "edit_dict = {\"A\":\"G\", \"T\":\"C\"}\n",
    "revcomp_dict = {\"A\":\"T\", \"T\":\"A\", \"C\":\"G\", \"G\":\"C\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "d32d7706",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_id = \"rs62084210_Maj_ABE_250\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "386e79cf",
   "metadata": {},
   "source": [
    "### Load variant & amplicon info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0fb57d54",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_info_tbl = pd.read_excel(\"../../resources/LDLvar/20221013_LDLvar_simpleZscores_credset.xlsx\", index_col=0, header=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "b80b4063",
   "metadata": {},
   "outputs": [],
   "source": [
    "amplicon_info = pd.read_csv(\"../../resources/atac_seq/030123_ATACseq_info.csv\", ).set_index(\"gRNA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9afc374d",
   "metadata": {},
   "outputs": [],
   "source": [
    "var_alleles = var_info_tbl.loc[var_id, [\"A1\", \"A2\"]].reset_index()\n",
    "var_alleles['edited'] = var_alleles[var_id].map(lambda b: edit_dict[b] if b in edit_dict.keys() else np.nan)\n",
    "var_alleles = var_alleles.loc[~var_alleles.edited.isnull()]\n",
    "var_alleles.columns=[\"allele\", \"ref\", \"edited\"]\n",
    "var_chrom = int(var_info_tbl.loc[var_id, \"CHR\"])\n",
    "var_pos = int(var_info_tbl.loc[var_id, 'position_hg19'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b305fbdd",
   "metadata": {},
   "outputs": [],
   "source": [
    "amplicon_start = amplicon_info.loc[var_id, \"Varpos_fw\"]-4\n",
    "amplicon_seq = amplicon_info.loc[var_id, \"Amplicon_fw (reference)\"].replace(\"N\", \"\")\n",
    "amplicon_len = len(amplicon_seq)\n",
    "amplicon_end = amplicon_start + amplicon_len"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bbdd9378",
   "metadata": {},
   "outputs": [],
   "source": [
    "phased_snps = pd.read_table(f\"../../results/atac_seq/{var_id}_phasedSNPs.vcf\", comment=\"#\", header=None)\n",
    "phased_snps.columns=[\"CHROM\", \"POS\", \"ID\", \"REF\", \"ALT\", \"QUAL\", \"FILTER\", \"INFO\", \"FORMAT\", \"HepG2\"]\n",
    "phased_snps[\"GT\"] = phased_snps.HepG2.map(lambda s: s.split(\":\")[0])\n",
    "phased_snps = phased_snps.loc[phased_snps.GT.isin([\"1|0\", \"0|1\"]) & (phased_snps.FILTER == \"PASS\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "7220127d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_allele(row):\n",
    "    ids = row.GT.split(\"|\")\n",
    "    alleles = [row[{\"0\":\"REF\", \"1\":\"ALT\"}[aid]] for aid in ids]\n",
    "    return alleles\n",
    "\n",
    "def _get_allele_from_alignment(\n",
    "    ref_aligned: str,\n",
    "    query_aligned: str,\n",
    "    offset: int,\n",
    "    strand: int,\n",
    "    start_pos: int,\n",
    "    end_pos: int,\n",
    "    positionwise_quality: np.ndarray = None,\n",
    "    quality_thres: float = -1,\n",
    "):\n",
    "    # Include N, no quality filter\n",
    "    assert len(ref_aligned) == len(query_aligned)\n",
    "    allele = Allele()\n",
    "    ref_gaps = 0\n",
    "    alt_gaps = 0\n",
    "    alt_seq_len = len(query_aligned) - query_aligned.count(\"-\")\n",
    "    if positionwise_quality is None:\n",
    "        # alt_position_is_good_quality = np.ones(alt_seq_len, dtype=bool)\n",
    "        alt_position_is_good_quality = np.array(\n",
    "            [c != \"N\" for c in query_aligned.replace(\"-\", \"\")]\n",
    "        )\n",
    "    else:\n",
    "        assert len(positionwise_quality) == alt_seq_len\n",
    "        alt_position_is_good_quality = positionwise_quality > quality_thres\n",
    "    for i in range(len(ref_aligned)):\n",
    "        if ref_aligned[i] == query_aligned[i]:\n",
    "            continue\n",
    "        ref_base = ref_aligned[i]\n",
    "        alt_base = query_aligned[i]\n",
    "        if alt_base != \"-\":\n",
    "            alt_base_is_good_quality = alt_position_is_good_quality[i - alt_gaps]\n",
    "        else:\n",
    "            alt_base_is_good_quality = True\n",
    "        if ref_base == \"-\":\n",
    "            ref_gaps += 1\n",
    "        elif alt_base == \"-\":\n",
    "            alt_gaps += 1\n",
    "        ref_pos = i - ref_gaps\n",
    "        allele.add(\n",
    "            Edit(\n",
    "                rel_pos=ref_pos,\n",
    "                ref_base=ref_base,\n",
    "                alt_base=alt_base,\n",
    "                offset=offset,\n",
    "                strand=strand,\n",
    "            )\n",
    "        )\n",
    "    return allele"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5a3c62bb",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(phased_snps) == 1:\n",
    "    raise ValueError(\"Cannot phase variant of interest.\")\n",
    "\n",
    "phased_snps['alleles'] = phased_snps.apply(get_allele, axis=1)\n",
    "phased_snps['amplicon_pos'] = phased_snps.POS - var_pos + amplicon_start\n",
    "phased_snps[\"amplicon_base\"] = phased_snps.amplicon_pos.map(\n",
    "    lambda i: amplicon_seq[i]\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9aa6cc13",
   "metadata": {},
   "outputs": [],
   "source": [
    "phased_snps_nonvar = phased_snps.loc[phased_snps.amplicon_pos != amplicon_start]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "426007db",
   "metadata": {},
   "outputs": [],
   "source": [
    "editable_phase = np.where(np.array([p in edit_dict.keys() for p in phased_snps.loc[phased_snps.amplicon_pos == amplicon_start, \"alleles\"].item()]))[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "73abeff9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nphase0    TT\\nphase1    CC\\n'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "phases = pd.DataFrame(\n",
    "    phased_snps_nonvar.alleles.tolist(), \n",
    "    columns=[f'phase{i}' for i in range(len(phased_snps_nonvar.alleles.tolist()[0]))]\n",
    ").sum(axis=0)\n",
    "'''\n",
    "phase0    TT\n",
    "phase1    CC\n",
    "'''\n",
    "#pd.DataFrame(df2[\"teams\"].to_list(), columns=['team1', 'team2'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07db6b61",
   "metadata": {},
   "source": [
    "## Phase reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "06484bdb",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_edit(row, target_pos):\n",
    "    return _get_allele_from_alignment(row.Reference_Sequence, row.Aligned_Sequence, offset=0, strand=1, start_pos=target_pos-20, end_pos=target_pos+20)\n",
    "\n",
    "def get_all_edits(row):\n",
    "    return _get_allele_from_alignment(row.Reference_Sequence, row.Aligned_Sequence, offset=0, strand=1, start_pos=phased_snps.amplicon_pos.min()-1, end_pos = phased_snps.amplicon_pos.max()+1)\n",
    "    \n",
    "\n",
    "def filter_allele_by_pos_qual(allele, target_pos):\n",
    "    # filters allele by variant included in list_pos or near target_pos\n",
    "    edits = []\n",
    "    n_out_edits = 0\n",
    "    n_lq = 0\n",
    "    n_crit_lq = 0\n",
    "    for edit in allele.split(\",\"):\n",
    "        if edit == \"\": continue\n",
    "        alt = edit.split(\">\")[-1]\n",
    "        pos = int(edit.split(\":\")[0])\n",
    "        if alt == \"N\": \n",
    "            n_lq += 1\n",
    "            if abs(pos-target_pos) < 5:\n",
    "                n_crit_lq += 1\n",
    "        elif abs(pos-target_pos) < 5:\n",
    "            edits.append(edit)\n",
    "        else:\n",
    "            n_out_edits += 1\n",
    "    return n_out_edits, n_lq, n_crit_lq, \",\".join(sorted(edits))\n",
    "\n",
    "def filter_allele_by_pos(list_pos, allele, target_pos):\n",
    "    edits = []\n",
    "    for edit in allele.split(\",\"):\n",
    "        if edit == \"\": continue\n",
    "        pos = int(edit.split(\":\")[0])\n",
    "        if pos in list_pos or abs(pos-target_pos) < 5:\n",
    "            edits.append(edit)\n",
    "    return \",\".join(sorted(edits))\n",
    "\n",
    "def filter_out_indels(allele):\n",
    "    edits = []\n",
    "    for edit in allele.edits:\n",
    "        if edit.ref_base != \"-\" and edit.alt_base != \"-\":\n",
    "            edits.append(str(edit))\n",
    "    return \",\".join(sorted(edits))\n",
    "\n",
    "\n",
    "\n",
    "def get_phase(allele, phased_snps, var_pos):\n",
    "    phase = []\n",
    "    allele_edits = {}\n",
    "    if allele != \"\":\n",
    "        edits = allele.split(\",\")\n",
    "        for edit in edits:\n",
    "            try:\n",
    "                allele_edits[int(edit.split(\":\")[0])] = edit.split(\":\")[-1]\n",
    "            except ValueError:\n",
    "                print(edit)\n",
    "    for pos in phased_snps.amplicon_pos.tolist():\n",
    "        if pos == var_pos: continue\n",
    "        if pos in allele_edits.keys():\n",
    "            phase.append(allele_edits[pos].split(\">\")[-1])\n",
    "        else:\n",
    "            phase.append(phased_snps[['amplicon_pos', 'amplicon_base']].set_index(\"amplicon_pos\").loc[pos, \"amplicon_base\"])\n",
    "    return phase\n",
    "\n",
    "\n",
    "def mask_seq(seq, mask_pos):\n",
    "    seq_array = np.array(list(seq))\n",
    "    seq_array[mask_pos] = \"N\"\n",
    "    return \"\".join(seq_array.tolist())\n",
    "    \n",
    "def assign_phase(phases, allele_phase):\n",
    "    match= np.where(phases == allele_phase[0])[0]\n",
    "    if len(match) != 1:\n",
    "        return -1\n",
    "    return match[0]\n",
    "def filter_out_indels(allele):\n",
    "    edits = []\n",
    "    for edit in allele.edits:\n",
    "        if edit.ref_base != \"-\" and edit.alt_base != \"-\":\n",
    "            edits.append(str(edit))\n",
    "    return \",\".join(sorted(edits))\n",
    "\n",
    "def get_pos_edit_rate(aftbl):\n",
    "    total_counts = aftbl['#Reads'].sum()\n",
    "    aftbl = aftbl.loc[aftbl.be_allele != \"\",:].copy()\n",
    "    aftbl['edits'] = aftbl.be_allele.map(lambda a: a.split(\",\"))\n",
    "    edit_tbl = aftbl.explode(\"edits\")\n",
    "    edit_tbl['pos'] = edit_tbl.edits.map(lambda e: int(e.split(\":\")[0]))\n",
    "    per_pos_edits = edit_tbl.groupby(\"pos\")[\"#Reads\"].sum()\n",
    "    per_pos_edit_rates = per_pos_edits / aftbl['#Reads'].sum()\n",
    "    return np.where(per_pos_edit_rates > 0.5)[0]\n",
    "\n",
    "def format_allele_freq_table(path, varpos_amplicon):\n",
    "    aftbl = pd.read_csv(path, sep='\\t')\n",
    "    aftbl[\"allele\"] = aftbl.apply(lambda row:get_all_edits(row),axis=1)\n",
    "    aftbl[\"allele_around_target\"] = aftbl.apply(lambda row:get_edit(row, varpos_amplicon),axis=1)\n",
    "    aftbl['be_allele'] = aftbl.allele.map(filter_out_indels)\n",
    "    aftbl['be_allele_around_target'] = aftbl.allele_around_target.map(filter_out_indels)\n",
    "    aftbl['err'] = aftbl.be_allele_around_target.map(lambda a:filter_allele_by_pos_qual(a, varpos_amplicon)[0])\n",
    "    aftbl['N'] = aftbl.be_allele_around_target.map(lambda a:filter_allele_by_pos_qual(a, varpos_amplicon)[1])\n",
    "    aftbl['target_proximal_N'] = aftbl.be_allele_around_target.map(lambda a:filter_allele_by_pos_qual(a, varpos_amplicon)[2])\n",
    "    aftbl['phased_allele'] = aftbl.be_allele.map(lambda a: filter_allele_by_pos(phased_snps.amplicon_pos.tolist(), a, varpos_amplicon))\n",
    "    aftbl = aftbl.loc[(aftbl.err <= 1) & (aftbl.target_proximal_N == 0) & (aftbl.N <= 2)]\n",
    "    aftbl['phase'] = aftbl.phased_allele.map(lambda a: get_phase(a, phased_snps, amplicon_start))\n",
    "    aftbl[\"assigned_phase\"] = aftbl.phase.map(lambda ap: assign_phase(phases, ap))\n",
    "    return aftbl\n",
    "\n",
    "def get_phased_edit_rate(aftbl):\n",
    "    var_amplicon_base = phased_snps.loc[phased_snps.amplicon_pos == amplicon_start, \"amplicon_base\"].item()\n",
    "    aftbl.loc[~aftbl.assigned_phase.isin(editable_phase), \"edited\"] = np.nan\n",
    "    if var_amplicon_base == var_alleles.ref.item():\n",
    "        aftbl.loc[aftbl.assigned_phase.isin(editable_phase),'edited'] = aftbl.loc[aftbl.assigned_phase.isin(editable_phase),:].phased_allele.map(lambda s: f\"{amplicon_start}:{amplicon_start}:+:{var_alleles.ref}>{var_alleles.edited.item()}\" in s)\n",
    "    elif var_amplicon_base == var_alleles.edited.item():\n",
    "        aftbl.loc[aftbl.assigned_phase.isin(editable_phase),'edited'] = aftbl.loc[aftbl.assigned_phase.isin(editable_phase),:].phased_allele.map(lambda s: f\"{amplicon_start}:{amplicon_start}:+:{var_amplicon_base}>{var_alleles.ref.item()}\" not in s)\n",
    "    aftbl['edited_clean'] = aftbl['edited'].map(lambda s: False if np.isnan(s) else s)\n",
    "    return aftbl.groupby(['assigned_phase', 'edited_clean'])['#Reads'].sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "573031aa",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "08740c5f3599400fb41c614c881c8305",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/4 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "ef3cab2aeb774f15a484c7b24cb20691",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "179e302f8e8940f78cb73cb747d046d3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "769df67d48e0466287a67efff08d622b",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "49e7b39e8964405fabaa9c5ec4c6b211",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "  0%|          | 0/2 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "results = {}\n",
    "for rep in tqdm([1,2,3, \"C\"]):\n",
    "    for cond in tqdm([\"Ser\", \"SS\"]):\n",
    "        exp_id = f\"{rep}-{cond}_20-locus\"\n",
    "        atac_path = f\"../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_{rep}-{cond}_20-locus_ATAC_Seq.{var_id}/Alleles_frequency_table.txt\"\n",
    "        gdna_path = f\"../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_{rep}-{cond}_20-locus_GDNA.{var_id}/Alleles_frequency_table.txt\"\n",
    "        \n",
    "        if not os.path.exists(atac_path):\n",
    "            os.system(f\"unzip ../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_{rep}-{cond}_20-locus_ATAC_Seq.{var_id}/Alleles_frequency_table.zip -d ../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_{rep}-{cond}_20-locus_ATAC_Seq.{var_id}/\")\n",
    "        aftbl_atac = format_allele_freq_table(atac_path, amplicon_start)\n",
    "            \n",
    "        if not os.path.exists(gdna_path):\n",
    "            os.system(f\"unzip ../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_{rep}-{cond}_20-locus_GDNA.{var_id}/Alleles_frequency_table.zip -d ../../results/atac_seq/crispresso_runs_indiv_demuxed/CRISPResso_on_{rep}-{cond}_20-locus_GDNA.{var_id}/\")\n",
    "        aftbl_gdna = format_allele_freq_table(gdna_path, amplicon_start)\n",
    "        results[exp_id] = [get_phased_edit_rate(aftbl_atac), get_phased_edit_rate(aftbl_gdna)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4130c773",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                             #Reads  #Readsgdna\n",
      "assigned_phase edited_clean                    \n",
      "-1             False            0.0           1\n",
      " 0             False           82.0         305\n",
      " 1             False          117.0         360\n",
      "                              atac  gdna\n",
      "assigned_phase edited_clean             \n",
      "-1             False           0.0     1\n",
      " 0             False          82.0   305\n",
      " 1             False         117.0   360\n",
      "                             #Reads  #Readsgdna\n",
      "assigned_phase edited_clean                    \n",
      "0              False             57         313\n",
      "1              False             57         327\n",
      "                             atac  gdna\n",
      "assigned_phase edited_clean            \n",
      "0              False           57   313\n",
      "1              False           57   327\n",
      "                             #Reads  #Readsgdna\n",
      "assigned_phase edited_clean                    \n",
      "0              False             80         258\n",
      "1              False             75         324\n",
      "                             atac  gdna\n",
      "assigned_phase edited_clean            \n",
      "0              False           80   258\n",
      "1              False           75   324\n",
      "                             #Reads  #Readsgdna\n",
      "assigned_phase edited_clean                    \n",
      "-1             False            0.0           2\n",
      " 0             False           30.0         337\n",
      " 1             False           49.0         349\n",
      "                             atac  gdna\n",
      "assigned_phase edited_clean            \n",
      "-1             False          0.0     2\n",
      " 0             False         30.0   337\n",
      " 1             False         49.0   349\n",
      "                             #Reads  #Readsgdna\n",
      "assigned_phase edited_clean                    \n",
      "0              False            204         280\n",
      "1              False            210         258\n",
      "                             atac  gdna\n",
      "assigned_phase edited_clean            \n",
      "0              False          204   280\n",
      "1              False          210   258\n",
      "                             #Reads  #Readsgdna\n",
      "assigned_phase edited_clean                    \n",
      "-1             False              1         0.0\n",
      " 0             False            256       286.0\n",
      " 1             False            259       284.0\n",
      "                             atac   gdna\n",
      "assigned_phase edited_clean             \n",
      "-1             False            1    0.0\n",
      " 0             False          256  286.0\n",
      " 1             False          259  284.0\n",
      "                             #Reads  #Readsgdna\n",
      "assigned_phase edited_clean                    \n",
      "0              False            109         356\n",
      "1              False            123         378\n",
      "                             atac  gdna\n",
      "assigned_phase edited_clean            \n",
      "0              False          109   356\n",
      "1              False          123   378\n",
      "                             #Reads  #Readsgdna\n",
      "assigned_phase edited_clean                    \n",
      "0              False             56         354\n",
      "1              False             49         363\n",
      "                             atac  gdna\n",
      "assigned_phase edited_clean            \n",
      "0              False           56   354\n",
      "1              False           49   363\n"
     ]
    }
   ],
   "source": [
    "res = []\n",
    "for k, dfs in results.items():\n",
    "    catdf = dfs[0].to_frame().join(dfs[1].to_frame(), how='outer', rsuffix='gdna').fillna(0)\n",
    "    print(catdf)\n",
    "    catdf.columns = [\"atac\", \"gdna\"]\n",
    "    print(catdf)\n",
    "    cols = catdf.index.get_level_values(0)\n",
    "    catdf = catdf.T\n",
    "    catdf.columns=cols\n",
    "    catdf['exp'] = k\n",
    "    res.append(catdf)\n",
    "results_df = pd.concat(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a4bc299b",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df.columns = [\"Unassigned\", \"Phase0\", \"Phase1\", \"exp\"]\n",
    "results_df['is_atac'] = results_df.index == 'atac'\n",
    "results_df['phase0_frac'] = results_df.Phase0 / results_df.iloc[:,1:3].sum(axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "090ae6b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "results_df['exp'] = results_df.exp.map(lambda x: x.split(\"_\")[0])\n",
    "results_df[['rep', 'cond']] = results_df.exp.str.split(\"-\", expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "37744bf6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unassigned</th>\n",
       "      <th>Phase0</th>\n",
       "      <th>Phase1</th>\n",
       "      <th>exp</th>\n",
       "      <th>is_atac</th>\n",
       "      <th>phase0_frac</th>\n",
       "      <th>rep</th>\n",
       "      <th>cond</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>0.0</td>\n",
       "      <td>82.0</td>\n",
       "      <td>117.0</td>\n",
       "      <td>1-Ser</td>\n",
       "      <td>True</td>\n",
       "      <td>0.412060</td>\n",
       "      <td>1</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>1.0</td>\n",
       "      <td>305.0</td>\n",
       "      <td>360.0</td>\n",
       "      <td>1-Ser</td>\n",
       "      <td>False</td>\n",
       "      <td>0.458647</td>\n",
       "      <td>1</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>NaN</td>\n",
       "      <td>57.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>1-SS</td>\n",
       "      <td>True</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>1</td>\n",
       "      <td>SS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>NaN</td>\n",
       "      <td>313.0</td>\n",
       "      <td>327.0</td>\n",
       "      <td>1-SS</td>\n",
       "      <td>False</td>\n",
       "      <td>0.489063</td>\n",
       "      <td>1</td>\n",
       "      <td>SS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>2-Ser</td>\n",
       "      <td>True</td>\n",
       "      <td>0.516129</td>\n",
       "      <td>2</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>NaN</td>\n",
       "      <td>258.0</td>\n",
       "      <td>324.0</td>\n",
       "      <td>2-Ser</td>\n",
       "      <td>False</td>\n",
       "      <td>0.443299</td>\n",
       "      <td>2</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>0.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>2-SS</td>\n",
       "      <td>True</td>\n",
       "      <td>0.379747</td>\n",
       "      <td>2</td>\n",
       "      <td>SS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>2.0</td>\n",
       "      <td>337.0</td>\n",
       "      <td>349.0</td>\n",
       "      <td>2-SS</td>\n",
       "      <td>False</td>\n",
       "      <td>0.491254</td>\n",
       "      <td>2</td>\n",
       "      <td>SS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>NaN</td>\n",
       "      <td>204.0</td>\n",
       "      <td>210.0</td>\n",
       "      <td>3-Ser</td>\n",
       "      <td>True</td>\n",
       "      <td>0.492754</td>\n",
       "      <td>3</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>NaN</td>\n",
       "      <td>280.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>3-Ser</td>\n",
       "      <td>False</td>\n",
       "      <td>0.520446</td>\n",
       "      <td>3</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>1.0</td>\n",
       "      <td>256.0</td>\n",
       "      <td>259.0</td>\n",
       "      <td>3-SS</td>\n",
       "      <td>True</td>\n",
       "      <td>0.497087</td>\n",
       "      <td>3</td>\n",
       "      <td>SS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>0.0</td>\n",
       "      <td>286.0</td>\n",
       "      <td>284.0</td>\n",
       "      <td>3-SS</td>\n",
       "      <td>False</td>\n",
       "      <td>0.501754</td>\n",
       "      <td>3</td>\n",
       "      <td>SS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>NaN</td>\n",
       "      <td>109.0</td>\n",
       "      <td>123.0</td>\n",
       "      <td>C-Ser</td>\n",
       "      <td>True</td>\n",
       "      <td>0.469828</td>\n",
       "      <td>C</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>NaN</td>\n",
       "      <td>356.0</td>\n",
       "      <td>378.0</td>\n",
       "      <td>C-Ser</td>\n",
       "      <td>False</td>\n",
       "      <td>0.485014</td>\n",
       "      <td>C</td>\n",
       "      <td>Ser</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>atac</th>\n",
       "      <td>NaN</td>\n",
       "      <td>56.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>C-SS</td>\n",
       "      <td>True</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>C</td>\n",
       "      <td>SS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gdna</th>\n",
       "      <td>NaN</td>\n",
       "      <td>354.0</td>\n",
       "      <td>363.0</td>\n",
       "      <td>C-SS</td>\n",
       "      <td>False</td>\n",
       "      <td>0.493724</td>\n",
       "      <td>C</td>\n",
       "      <td>SS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      Unassigned  Phase0  Phase1    exp  is_atac  phase0_frac rep cond\n",
       "atac         0.0    82.0   117.0  1-Ser     True     0.412060   1  Ser\n",
       "gdna         1.0   305.0   360.0  1-Ser    False     0.458647   1  Ser\n",
       "atac         NaN    57.0    57.0   1-SS     True     0.500000   1   SS\n",
       "gdna         NaN   313.0   327.0   1-SS    False     0.489063   1   SS\n",
       "atac         NaN    80.0    75.0  2-Ser     True     0.516129   2  Ser\n",
       "gdna         NaN   258.0   324.0  2-Ser    False     0.443299   2  Ser\n",
       "atac         0.0    30.0    49.0   2-SS     True     0.379747   2   SS\n",
       "gdna         2.0   337.0   349.0   2-SS    False     0.491254   2   SS\n",
       "atac         NaN   204.0   210.0  3-Ser     True     0.492754   3  Ser\n",
       "gdna         NaN   280.0   258.0  3-Ser    False     0.520446   3  Ser\n",
       "atac         1.0   256.0   259.0   3-SS     True     0.497087   3   SS\n",
       "gdna         0.0   286.0   284.0   3-SS    False     0.501754   3   SS\n",
       "atac         NaN   109.0   123.0  C-Ser     True     0.469828   C  Ser\n",
       "gdna         NaN   356.0   378.0  C-Ser    False     0.485014   C  Ser\n",
       "atac         NaN    56.0    49.0   C-SS     True     0.533333   C   SS\n",
       "gdna         NaN   354.0   363.0   C-SS    False     0.493724   C   SS"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5b70b17f",
   "metadata": {},
   "outputs": [],
   "source": [
    "y, X = dmatrices('Phase0 + Phase1 ~ rep + cond + is_atac + cond*is_atac', data = results_df, return_type='dataframe')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "ba944516",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_binom = sm.GLM(y, X, family=sm.families.Binomial())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "c4c45508",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_res = glm_binom.fit()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ff327155",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"simpletable\">\n",
       "<caption>Generalized Linear Model Regression Results</caption>\n",
       "<tr>\n",
       "  <th>Dep. Variable:</th>   <td>['Phase0', 'Phase1']</td> <th>  No. Observations:  </th>  <td>    16</td> \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Model:</th>                    <td>GLM</td>         <th>  Df Residuals:      </th>  <td>     9</td> \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Model Family:</th>          <td>Binomial</td>       <th>  Df Model:          </th>  <td>     6</td> \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Link Function:</th>           <td>logit</td>        <th>  Scale:             </th> <td>  1.0000</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Method:</th>                  <td>IRLS</td>         <th>  Log-Likelihood:    </th> <td> -55.723</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Date:</th>              <td>Wed, 14 Jun 2023</td>   <th>  Deviance:          </th> <td>  10.834</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Time:</th>                  <td>13:13:09</td>       <th>  Pearson chi2:      </th>  <td>  10.8</td> \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>No. Iterations:</th>            <td>3</td>          <th>                     </th>     <td> </td>   \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Covariance Type:</th>       <td>nonrobust</td>      <th>                     </th>     <td> </td>   \n",
       "</tr>\n",
       "</table>\n",
       "<table class=\"simpletable\">\n",
       "<tr>\n",
       "               <td></td>                  <th>coef</th>     <th>std err</th>      <th>z</th>      <th>P>|z|</th>  <th>[0.025</th>    <th>0.975]</th>  \n",
       "</tr>\n",
       "<tr>\n",
       "  <th>Intercept</th>                   <td>   -0.0825</td> <td>    0.058</td> <td>   -1.414</td> <td> 0.157</td> <td>   -0.197</td> <td>    0.032</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>rep[T.2]</th>                    <td>    0.0013</td> <td>    0.072</td> <td>    0.018</td> <td> 0.986</td> <td>   -0.140</td> <td>    0.142</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>rep[T.3]</th>                    <td>    0.1524</td> <td>    0.069</td> <td>    2.219</td> <td> 0.026</td> <td>    0.018</td> <td>    0.287</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>rep[T.C]</th>                    <td>    0.0862</td> <td>    0.069</td> <td>    1.254</td> <td> 0.210</td> <td>   -0.049</td> <td>    0.221</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>cond[T.Ser]</th>                 <td>   -0.0717</td> <td>    0.056</td> <td>   -1.281</td> <td> 0.200</td> <td>   -0.181</td> <td>    0.038</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>is_atac[T.True]</th>             <td>   -0.0623</td> <td>    0.084</td> <td>   -0.744</td> <td> 0.457</td> <td>   -0.226</td> <td>    0.102</td>\n",
       "</tr>\n",
       "<tr>\n",
       "  <th>cond[T.Ser]:is_atac[T.True]</th> <td>    0.0330</td> <td>    0.111</td> <td>    0.298</td> <td> 0.766</td> <td>   -0.184</td> <td>    0.250</td>\n",
       "</tr>\n",
       "</table>"
      ],
      "text/plain": [
       "<class 'statsmodels.iolib.summary.Summary'>\n",
       "\"\"\"\n",
       "                  Generalized Linear Model Regression Results                   \n",
       "================================================================================\n",
       "Dep. Variable:     ['Phase0', 'Phase1']   No. Observations:                   16\n",
       "Model:                              GLM   Df Residuals:                        9\n",
       "Model Family:                  Binomial   Df Model:                            6\n",
       "Link Function:                    logit   Scale:                          1.0000\n",
       "Method:                            IRLS   Log-Likelihood:                -55.723\n",
       "Date:                  Wed, 14 Jun 2023   Deviance:                       10.834\n",
       "Time:                          13:13:09   Pearson chi2:                     10.8\n",
       "No. Iterations:                       3                                         \n",
       "Covariance Type:              nonrobust                                         \n",
       "===============================================================================================\n",
       "                                  coef    std err          z      P>|z|      [0.025      0.975]\n",
       "-----------------------------------------------------------------------------------------------\n",
       "Intercept                      -0.0825      0.058     -1.414      0.157      -0.197       0.032\n",
       "rep[T.2]                        0.0013      0.072      0.018      0.986      -0.140       0.142\n",
       "rep[T.3]                        0.1524      0.069      2.219      0.026       0.018       0.287\n",
       "rep[T.C]                        0.0862      0.069      1.254      0.210      -0.049       0.221\n",
       "cond[T.Ser]                    -0.0717      0.056     -1.281      0.200      -0.181       0.038\n",
       "is_atac[T.True]                -0.0623      0.084     -0.744      0.457      -0.226       0.102\n",
       "cond[T.Ser]:is_atac[T.True]     0.0330      0.111      0.298      0.766      -0.184       0.250\n",
       "===============================================================================================\n",
       "\"\"\""
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glm_res.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "7ea46e50",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_html = glm_res.summary().tables[1].as_html()\n",
    "glm_res_tbl = pd.read_html(result_html, header=0, index_col=0)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ff44499f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>coef</th>\n",
       "      <th>std err</th>\n",
       "      <th>z</th>\n",
       "      <th>P&gt;|z|</th>\n",
       "      <th>[0.025</th>\n",
       "      <th>0.975]</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Intercept</th>\n",
       "      <td>-0.0825</td>\n",
       "      <td>0.058</td>\n",
       "      <td>-1.414</td>\n",
       "      <td>0.157</td>\n",
       "      <td>-0.197</td>\n",
       "      <td>0.032</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rep[T.2]</th>\n",
       "      <td>0.0013</td>\n",
       "      <td>0.072</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.986</td>\n",
       "      <td>-0.140</td>\n",
       "      <td>0.142</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rep[T.3]</th>\n",
       "      <td>0.1524</td>\n",
       "      <td>0.069</td>\n",
       "      <td>2.219</td>\n",
       "      <td>0.026</td>\n",
       "      <td>0.018</td>\n",
       "      <td>0.287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>rep[T.C]</th>\n",
       "      <td>0.0862</td>\n",
       "      <td>0.069</td>\n",
       "      <td>1.254</td>\n",
       "      <td>0.210</td>\n",
       "      <td>-0.049</td>\n",
       "      <td>0.221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cond[T.Ser]</th>\n",
       "      <td>-0.0717</td>\n",
       "      <td>0.056</td>\n",
       "      <td>-1.281</td>\n",
       "      <td>0.200</td>\n",
       "      <td>-0.181</td>\n",
       "      <td>0.038</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>is_atac[T.True]</th>\n",
       "      <td>-0.0623</td>\n",
       "      <td>0.084</td>\n",
       "      <td>-0.744</td>\n",
       "      <td>0.457</td>\n",
       "      <td>-0.226</td>\n",
       "      <td>0.102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cond[T.Ser]:is_atac[T.True]</th>\n",
       "      <td>0.0330</td>\n",
       "      <td>0.111</td>\n",
       "      <td>0.298</td>\n",
       "      <td>0.766</td>\n",
       "      <td>-0.184</td>\n",
       "      <td>0.250</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                               coef  std err      z  P>|z|  [0.025  0.975]\n",
       "Intercept                   -0.0825    0.058 -1.414  0.157  -0.197   0.032\n",
       "rep[T.2]                     0.0013    0.072  0.018  0.986  -0.140   0.142\n",
       "rep[T.3]                     0.1524    0.069  2.219  0.026   0.018   0.287\n",
       "rep[T.C]                     0.0862    0.069  1.254  0.210  -0.049   0.221\n",
       "cond[T.Ser]                 -0.0717    0.056 -1.281  0.200  -0.181   0.038\n",
       "is_atac[T.True]             -0.0623    0.084 -0.744  0.457  -0.226   0.102\n",
       "cond[T.Ser]:is_atac[T.True]  0.0330    0.111  0.298  0.766  -0.184   0.250"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glm_res_tbl"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "0d688088",
   "metadata": {},
   "outputs": [],
   "source": [
    "glm_res_tbl.to_csv(\"rs62084210_Maj_ABE_250_caqtl_result.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4f90b19a",
   "metadata": {},
   "source": [
    "ns"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jy_anbe_py38",
   "language": "python",
   "name": "jy_anbe_py38"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
